---
title: "analysis"
output: html_document
date: "2023-08-15"
---

```{r setup, include=FALSE}
library(ggplot2)
library(ggrepel)
library(EMMREML)
library(reshape2)
library(cobs)
library(cowplot)
library(stringr)
library(viridis)

#bioc packages:
library(edgeR)
library(qvalue)

library(parallel)
library(GGally)

library(UpSetR)
library(mashr)

##cibersort
library(e1071)
library(preprocessCore)
library(edgeR)
library(gridExtra)




```

```{r scripts and dirs}

workDir <- gitDir <- "~/Documents/giterdone/"

dataDir <- paste0(workDir,"data/")
figDir <- paste0(workDir,"figures/")
outDir <- paste0(workDir,"output/")

dir.create(dataDir)
dir.create(figDir)
dir.create(outDir)

source(paste0(gitDir,"meerkatPaper/rScripts/gatherData.R"))
source(paste0(gitDir,"meerkatPaper/rScripts/generateResiduals.R"))
source(paste0(gitDir,"meerkatPaper/rScripts/runEmmreml.R"))
source(paste0(gitDir,"meerkatPaper/rScripts/runResponseEmreml.R"))
source(paste0(gitDir,"meerkatPaper/rScripts/perm.fdr.R"))
source(paste0(gitDir,"meerkatPaper/rScripts/compLMMBeta.R"))
source(paste0(gitDir,"meerkatPaper/rScripts/gsea.R"))


```


gather data for a given model:
Condition (LPS, Dex, Gard, Ctrl), Treatment (LPS-Ctrl, Dex-Ctrl, Gard-Ctrl)
Sex (F, M)
Batch (All)
minGenes per sample (10000)
logCPM (>5, >7, sig in status)
```{r using gatherData}
batchList <- "allBatches"
minGenes <- 10000

for (focalSex in c("M","F")) {
  for (minLogCPM in c(5,7,"status")) {
    for (singTrt in c(T,F)) {
      if (singTrt) {
        for (focalTreatment in c("LPS","Dex","Gard","Ctrl")) {
          gatherData(workDir = workDir, 
                       batchList = batchList,
                       focalSex = focalSex,
                       focalTreatment = focalTreatment,
                       minGenes = minGenes,
                       minLogCPM = minLogCPM,
                       singleTrt = singTrt)
          }
        } else {
          for (focalTreatment in c("LPS","Dex","Gard")) {
            gatherData(workDir = workDir, 
                       batchList = batchList,
                       focalSex = focalSex,
                       focalTreatment = focalTreatment,
                       minGenes = minGenes,
                       minLogCPM = minLogCPM,
                       singleTrt = singTrt)
        }
      } 
    }
  }
}


for (focalSex in c("F","M")) {
  for (minLogCPM in c(5,7,"status")) {
    print(paste0(focalSex," ",minLogCPM))
    gatherData(workDir =  workDir, batchList = "allBatches", focalSex = focalSex, 
               focalTreatment = "allTrt", minGenes = 10000, minLogCPM = minLogCPM, 
               singleTrt = FALSE, combineTrt = TRUE)
  }
}
```

generate residuals
```{r using genResids}
### and now generate residuals

# workDir <- workDir
# batchList <- c("allBatches")
# focalSex <- "F"
# focalTreatment <- "LPS"
# minLogCPM <- 5
# minGenes <- 10000
# singleTrt <- FALSE
# combineTrt <- FALSE

### basic (ctrl for "none"), F (ctrl, lps) / G (dex, gard)
## then weight, preg, status

#P: F + Preg
#Q: G + Preg
#R: F + Status
#S: G + Status
#T: F + Weight
#U: G + Weight



minGenes <- 10000
 

for (focalSex in c("M","F")) {
    for (minLogCPM in c(5,7,"status")) {
      for (singTrt in c(F,T)) {
        if (singTrt) {
          for (focalTreatment in c("LPS","Dex","Gard","Ctrl")) {
            gatherSubset <- paste0(batchList,"_",focalSex,"_",focalTreatment,"only_",minGenes,"genes_",minLogCPM,"lCPM")
            ## run some genResids for TRTonly by Model
            if (focalTreatment %in% c("LPS","Ctrl")) {
              for (modelName in c("F","P","R","T")) {
                if (!(modelName == "P" & focalSex == "M")) {
                  genResids(workDir = workDir, gatherSubset = gatherSubset, model = modelName)
                }
              }
            } else {
              for (modelName in c("G","Q","S","U")) {
                if (!(modelName == "Q" & focalSex == "M")) {
                  genResids(workDir = workDir, gatherSubset = gatherSubset, model = modelName)
                }
              }
            }
          }
        } else if (combineTrt){
          gatherSubset <- paste0(batchList,"_",focalSex,"_allTrt_",minGenes,"genes_",minLogCPM,"lCPM")
          ## run some genResids for combined treatment
          for (modelName in c("G","Q","S","U")) {
            if (!(modelName == "S" & focalSex == "M")) {
              genResids(workDir = workDir, gatherSubset = gatherSubset, model = modelName)
            }
          }
        } else {
          for (focalTreatment in c("LPS","Dex","Gard")) {
            gatherSubset <- paste0(batchList,"_",focalSex,"_",focalTreatment,"_",minGenes,"genes_",minLogCPM,"lCPM")
            ### run some genResids for TRT-Ctrl
            if (focalTreatment %in% c("LPS")) {
              for (modelName in c("F","P","R","T")) {
                if (!(modelName == "P" & focalSex == "M")) {
                  genResids(workDir = workDir, gatherSubset = gatherSubset, model = modelName)
                }
              }
            } else {
              for (modelName in c("G","Q","S","U")) {
                if (!(modelName == "Q" & focalSex == "M")) {
                  genResids(workDir = workDir, gatherSubset = gatherSubset, model = modelName)
                }
              }
            }
          }
        }
      }
    }
  }


for (focalSex in c("F","M")) {
  for (minLogCPM in c(5,7,"status")) {
    print(paste0(focalSex," ",minLogCPM))
    gatherSubset <- paste0(batchList,"_",focalSex,"_allTrt_",minGenes,"genes_",minLogCPM,"lCPM")
    for (modelName in c("G","Q","S","U")) {
      if (!(modelName == "Q" & focalSex == "M")) {
        genResids(workDir = workDir, gatherSubset = gatherSubset, model = modelName)
      }
    }
  }
}


### add age controls

#R: H + Age
#S: I + Age
#T: X + Age-Status Centered
#U: Y + Age-Status Centered

for (focalSex in c("F")) {
    for (minLogCPM in c(5)) {
      for (singTrt in c(T)) {
          for (focalTreatment in c("LPS","Dex","Gard","Ctrl")) {
            gatherSubset <- paste0(batchList,"_",focalSex,"_",focalTreatment,"only_",minGenes,"genes_",minLogCPM,"lCPM")
            ## run some genResids for TRTonly by Model
            if (focalTreatment %in% c("LPS","Ctrl")) {
              for (modelName in c("H","X")) {
                genResids(workDir = workDir, gatherSubset = gatherSubset, model = modelName)
              }
            } else {
              for (modelName in c("I","Y")) {
                genResids(workDir = workDir, gatherSubset = gatherSubset, model = modelName)
              }
            }
          }
      }
    }
}

for (focalSex in c("F")) {
  for (minLogCPM in c(5)) {
    print(paste0(focalSex," ",minLogCPM))
    gatherSubset <- paste0(batchList,"_",focalSex,"_allTrt_",minGenes,"genes_",minLogCPM,"lCPM")
    for (modelName in c("I","X")) {
        genResids(workDir = workDir, gatherSubset = gatherSubset, model = modelName)
    }
  }
}
              
```

run emmreml models
basic | "basic"
basic + preg | ""basic+Preg"
basic + weight | "basic+Weight"
basic + weight + preg | "basic+Preg+Weight"

```{r run emmreml}
batchList <- "allBatches"
minGenes <- 10000
finishedModels <- vector()

### testing single
for (focalSex in c("F","M")) {
  for (focalTreatment in c("LPS","Dex","Gard")) {
    for (minLogCPM in c(7,"status")) {
    #for (minLogCPM in c(5)) {
      print(paste0(focalSex," - ",focalTreatment," - ",minLogCPM))
      gatherSubset <- paste0(batchList,"_",focalSex,"_",focalTreatment,"_",minGenes,"genes_",minLogCPM,"lCPM")
      if (focalTreatment == "LPS") {
        modelList <- c("F","P","R","T")
        noMaleModel <- "P"
      } else {
        modelList <- c("G","Q","S","U")
        noMaleModel <- "Q"
      }
      
      for (modelName in modelList) {
        #if this hasn't been run before
        if (!paste0(gatherSubset,"_",modelName) %in% finishedModels) {
        if (!(modelName == noMaleModel & focalSex == "M")) {
          if (focalSex == "F") {
            #runEmreml(workDir = workDir, gatherSubset = gatherSubset, model = modelName, LM = "basic", Kmat = TRUE, Kfile = "~/Dropbox (Personal)/meerkats/KMPsamples/KMPrelatedness_matrix.RDS")
            #runEmreml(workDir = workDir, gatherSubset = gatherSubset, model = modelName, LM = "basic+Preg", Kmat = TRUE, Kfile = "~/Dropbox (Personal)/meerkats/KMPsamples/KMPrelatedness_matrix.RDS")
            #runEmreml(workDir = workDir, gatherSubset = gatherSubset, model = modelName, LM = "basic+Weight", Kmat = TRUE, Kfile = "~/Dropbox (Personal)/meerkats/KMPsamples/KMPrelatedness_matrix.RDS")
            runEmreml(workDir = workDir, gatherSubset = gatherSubset, model = modelName, LM = "basic+Preg+Weight", Kmat = TRUE, Kfile = "~/Dropbox (Personal)/meerkats/KMPsamples/KMPrelatedness_matrix.RDS")
          } else {
            #runEmreml(workDir = workDir, gatherSubset = gatherSubset, model = modelName, LM = "basic", Kmat = TRUE, Kfile = "~/Dropbox (Personal)/meerkats/KMPsamples/KMPrelatedness_matrix.RDS")
            #add weight
            #runEmreml(workDir = workDir, gatherSubset = gatherSubset, model = modelName, LM = "basic+Weight", Kmat = TRUE, Kfile = "~/Dropbox (Personal)/meerkats/KMPsamples/KMPrelatedness_matrix.RDS")
          }
        }
        finishedModels <- c(finishedModels,paste0(gatherSubset,"_",modelName))
        }
      }
    }
  }
}

### treatment only models

batchList <- "allBatches"
minGenes <- 10000
finishedModels <- vector()

### testing single
for (focalSex in c("F","M")) {
  for (focalTreatment in c("LPS","Dex","Gard","Ctrl")) {
    for (minLogCPM in c(7,"status")) {
    #for (minLogCPM in c(5)) {
      print(paste0(focalSex," - ",focalTreatment," - ",minLogCPM))
      gatherSubset <- paste0(batchList,"_",focalSex,"_",focalTreatment,"only_",minGenes,"genes_",minLogCPM,"lCPM")
      if (focalTreatment %in% c("LPS","Ctrl")) {
        modelList <- c("F","P","R","T")
        noMaleModel <- "P"
      } else {
        modelList <- c("G","Q","S","U")
        noMaleModel <- "Q"
      }
      
      for (modelName in modelList) {
        #if this hasn't been run before
        if (!paste0(gatherSubset,"_",modelName) %in% finishedModels) {
        if (!(modelName == noMaleModel & focalSex == "M")) {
          if (focalSex == "F") {
            #runEmreml(workDir = workDir, gatherSubset = gatherSubset, model = modelName, LM = "basic", Kmat = TRUE, Kfile = "~/Dropbox (Personal)/meerkats/KMPsamples/KMPrelatedness_matrix.RDS")
            #runEmreml(workDir = workDir, gatherSubset = gatherSubset, model = modelName, LM = "basic+Preg", Kmat = TRUE, Kfile = "~/Dropbox (Personal)/meerkats/KMPsamples/KMPrelatedness_matrix.RDS")
            #runEmreml(workDir = workDir, gatherSubset = gatherSubset, model = modelName, LM = "basic+Weight", Kmat = TRUE, Kfile = "~/Dropbox (Personal)/meerkats/KMPsamples/KMPrelatedness_matrix.RDS")
            runEmreml(workDir = workDir, gatherSubset = gatherSubset, model = modelName, LM = "basic+Preg+Weight", Kmat = TRUE, Kfile = "~/Dropbox (Personal)/meerkats/KMPsamples/KMPrelatedness_matrix.RDS")
          } else {
            #runEmreml(workDir = workDir, gatherSubset = gatherSubset, model = modelName, LM = "basic", Kmat = TRUE, Kfile = "~/Dropbox (Personal)/meerkats/KMPsamples/KMPrelatedness_matrix.RDS")
            #add weight
            #runEmreml(workDir = workDir, gatherSubset = gatherSubset, model = modelName, LM = "basic+Weight", Kmat = TRUE, Kfile = "~/Dropbox (Personal)/meerkats/KMPsamples/KMPrelatedness_matrix.RDS")
          }
        }
        finishedModels <- c(finishedModels,paste0(gatherSubset,"_",modelName))
        }
      }
    }
  }
}


#all treatment samples together
#run some emreml

for (focalSex in c("F","M")) {
  for (minLogCPM in c(5,7,"status")) {
    print(paste0(focalSex," ",minLogCPM))
    gatherSubset <- paste0(batchList,"_",focalSex,"_allTrt_",minGenes,"genes_",minLogCPM,"lCPM")
    if (focalSex == "F") {
      for (modelName in c("G","Q","S","U")) {
        print(paste0(focalSex," ",minLogCPM," ",modelName))
        #runEmreml(workDir =  workDir, gatherSubset = gatherSubset, model = modelName, LM = "allTrtF", Kmat = TRUE, Kfile = "~/Dropbox (Personal)/meerkats/KMPsamples/KMPrelatedness_matrix.RDS")
        #runEmreml(workDir =  workDir, gatherSubset = gatherSubset, model = modelName, LM = "allTrtF+Weight", Kmat = TRUE, Kfile = "~/Dropbox (Personal)/meerkats/KMPsamples/KMPrelatedness_matrix.RDS")
        #runEmreml(workDir =  workDir, gatherSubset = gatherSubset, model = modelName, LM = "allTrtF+Preg", Kmat = TRUE, Kfile = "~/Dropbox (Personal)/meerkats/KMPsamples/KMPrelatedness_matrix.RDS")
        #runEmreml(workDir =  workDir, gatherSubset = gatherSubset, model = modelName, LM = "allTrtF+Preg+Weight", Kmat = TRUE, Kfile = "~/Dropbox (Personal)/meerkats/KMPsamples/KMPrelatedness_matrix.RDS")
        }
      } else {
        for (modelName in c("G","S","U")) {
        print(paste0(focalSex," ",minLogCPM," ",modelName))
        #runEmreml(workDir =  workDir, gatherSubset = gatherSubset, model = modelName, LM = "allTrtM", Kmat = TRUE, Kfile = "~/Dropbox (Personal)/meerkats/KMPsamples/KMPrelatedness_matrix.RDS")
        #runEmreml(workDir =  workDir, gatherSubset = gatherSubset, model = modelName, LM = "allTrtM+Weight", Kmat = TRUE, Kfile = "~/Dropbox (Personal)/meerkats/KMPsamples/KMPrelatedness_matrix.RDS")
      }
    }
  }
}
```

```{r age control}
### treatment only models

batchList <- "allBatches"
minGenes <- 10000
finishedModels <- vector()

### testing single
for (focalSex in c("F")) {
  for (focalTreatment in c("LPS","Dex","Gard","Ctrl")) {
    for (minLogCPM in c(5)) {
    #for (minLogCPM in c(5)) {
      print(paste0(focalSex," - ",focalTreatment," - ",minLogCPM))
      gatherSubset <- paste0(batchList,"_",focalSex,"_",focalTreatment,"only_",minGenes,"genes_",minLogCPM,"lCPM")
      if (focalTreatment %in% c("LPS","Ctrl")) {
        modelList <- c("H","X")
      } else {
        modelList <- c("I","Y")
      }
      
      for (modelName in modelList) {
        #if this hasn't been run before
        if (!paste0(gatherSubset,"_",modelName) %in% finishedModels) {
          print(modelName)
            #runEmreml(workDir = workDir, gatherSubset = gatherSubset, model = modelName, LM = "basic", Kmat = TRUE, Kfile = "~/Dropbox (Personal)/meerkats/KMPsamples/KMPrelatedness_matrix.RDS")
            #runEmreml(workDir = workDir, gatherSubset = gatherSubset, model = modelName, LM = "basic+Preg", Kmat = TRUE, Kfile = "~/Dropbox (Personal)/meerkats/KMPsamples/KMPrelatedness_matrix.RDS")
            #runEmreml(workDir = workDir, gatherSubset = gatherSubset, model = modelName, LM = "basic+Weight", Kmat = TRUE, Kfile = "~/Dropbox (Personal)/meerkats/KMPsamples/KMPrelatedness_matrix.RDS")
            runEmreml(workDir = workDir, gatherSubset = gatherSubset, model = modelName, LM = "basic+Preg+Weight", Kmat = TRUE, Kfile = "~/Dropbox (Personal)/meerkats/KMPsamples/KMPrelatedness_matrix.RDS")
          } else {
            #runEmreml(workDir = workDir, gatherSubset = gatherSubset, model = modelName, LM = "basic", Kmat = TRUE, Kfile = "~/Dropbox (Personal)/meerkats/KMPsamples/KMPrelatedness_matrix.RDS")
            #add weight
            #runEmreml(workDir = workDir, gatherSubset = gatherSubset, model = modelName, LM = "basic+Weight", Kmat = TRUE, Kfile = "~/Dropbox (Personal)/meerkats/KMPsamples/KMPrelatedness_matrix.RDS")
          }
        
        finishedModels <- c(finishedModels,paste0(gatherSubset,"_",modelName))
        }
      
    }
  }
}

gatherSubset <- paste0(batchList,"_",focalSex,"_allTrt_",minGenes,"genes_",minLogCPM,"lCPM")

for (modelName in c("I","Y")) {
  print(modelName)
  runEmreml(workDir =  workDir, gatherSubset = gatherSubset, model = modelName, LM = "allTrtF+Preg+Weight", Kmat = TRUE,
            Kfile = "~/Dropbox (Personal)/meerkats/KMPsamples/KMPrelatedness_matrix.RDS")
}

```





run nest emmreml models in Females
nested + weight + preg | "nested+Preg+Weight"

```{r nested models}
batchList <- "allBatches"
minGenes <- 10000
finishedModels <- vector()

### testing single
for (focalTreatment in c("LPS","Dex","Gard")) {
  for (minLogCPM in c(5,7,"status")) {
    print(paste0(focalSex," - ",focalTreatment," - ",minLogCPM))
    gatherSubset <- paste0(batchList,"_",focalSex,"_",focalTreatment,"_",minGenes,"genes_",minLogCPM,"lCPM")
    if (focalTreatment == "LPS") {
      modelList <- c("F","P","R","T")
      noMaleModel <- "P"
    } else {
      modelList <- c("G","Q","S","U")
      noMaleModel <- "Q"
    }
    
    for (modelName in modelList) {
      #if this hasn't been run before
      if (!paste0(gatherSubset,"_",modelName) %in% finishedModels) {
      if (!(modelName == noMaleModel & focalSex == "M")) {
        if (focalSex == "F") {
          #runEmreml(workDir = workDir, gatherSubset = gatherSubset, model = modelName, LM = "basic", Kmat = TRUE, Kfile = "~/Dropbox (Personal)/meerkats/KMPsamples/KMPrelatedness_matrix.RDS")
          #runEmreml(workDir = workDir, gatherSubset = gatherSubset, model = modelName, LM = "basic+Preg", Kmat = TRUE, Kfile = "~/Dropbox (Personal)/meerkats/KMPsamples/KMPrelatedness_matrix.RDS")
          #runEmreml(workDir = workDir, gatherSubset = gatherSubset, model = modelName, LM = "basic+Weight", Kmat = TRUE, Kfile = "~/Dropbox (Personal)/meerkats/KMPsamples/KMPrelatedness_matrix.RDS")
          runEmreml(workDir = workDir, gatherSubset = gatherSubset, model = modelName, LM = "nested+Preg+Weight", Kmat = TRUE, Kfile = "~/Dropbox (Personal)/meerkats/KMPsamples/KMPrelatedness_matrix.RDS")
        } else {
          #runEmreml(workDir = workDir, gatherSubset = gatherSubset, model = modelName, LM = "basic", Kmat = TRUE, Kfile = "~/Dropbox (Personal)/meerkats/KMPsamples/KMPrelatedness_matrix.RDS")
          #add weight
          #runEmreml(workDir = workDir, gatherSubset = gatherSubset, model = modelName, LM = "basic+Weight", Kmat = TRUE, Kfile = "~/Dropbox (Personal)/meerkats/KMPsamples/KMPrelatedness_matrix.RDS")
        }
      }
      finishedModels <- c(finishedModels,paste0(gatherSubset,"_",modelName))
      }
    }
  }
}


```


```{r male and female samples for treatment effect}
#gather data
batchList <- "allBatches"
minGenes <- 10000
minLogCPM <- 5
singTrt <- FALSE
focalSex <- "MF"

for (focalTreatment in c("LPS","Dex","Gard","allTrt")) {
  
  gatherData(workDir = workDir, 
             batchList = batchList,
             focalSex = focalSex,
             focalTreatment = focalTreatment,
             minGenes = minGenes,
             minLogCPM = minLogCPM,
             singleTrt = singTrt)
}

#generate residuals

for (focalTreatment in c("LPS","Dex","Gard","allTrt")) {
  
  gatherSubset <- paste0(batchList,"_",focalSex,"_",focalTreatment,"_",minGenes,"genes_",minLogCPM,"lCPM")
  
  if (focalTreatment %in% c("LPS")) {
    modelName <- "F"
  } else {
    modelName <- "G"
  }
  
  genResids(workDir = workDir, gatherSubset = gatherSubset, model = modelName)
}

#emmreml
#for (focalTreatment in c("LPS","Dex","Gard")) {
for (focalTreatment in c("Gard")) {
  
  print(paste0(focalSex," - ",focalTreatment," - ",minLogCPM))
  gatherSubset <- paste0(batchList,"_",focalSex,"_",focalTreatment,"_",minGenes,"genes_",minLogCPM,"lCPM")
  if (focalTreatment == "LPS") {
        modelName <- "F"
      } else {
        modelName <- "G"
      }
  if (focalTreatment == "allTrt") {
    runEmreml(workDir = workDir, gatherSubset = gatherSubset, model = modelName, LM = "allTrtMF+Weight",
              Kmat = TRUE, Kfile = "~/Dropbox (Personal)/meerkats/KMPsamples/KMPrelatedness_matrix.RDS")
  } else {
    runEmreml(workDir = workDir, gatherSubset = gatherSubset, model = modelName, LM = "basicMF+Weight",
              Kmat = TRUE, Kfile = "~/Dropbox (Personal)/meerkats/KMPsamples/KMPrelatedness_matrix.RDS")
  }
}


#permutations
batchList <- "allBatches"
focalSex <- "MF"
#focalTreatment <- "LPS"
minGenes <- 10000
minLogCPM <- 5
modelName <- "F"
singleTrt <- FALSE
permVersion <- 1
# permute treatment in LPS/Dex/Gard

outputFiles <- list.files(outDir)

permVar <- "treatment"
for (focalTreatment in c("LPS","Dex","Gard")) {
  
  if (focalTreatment %in% c("LPS","Ctrl")) {
    modelName <- "F"
  } else {
    modelName <- "G"
  }
  
  if (focalTreatment %in% c("LPS","Gard","Dex")) {
    LM <- "basicMF+Weight"
  }
  
  gatherSubset <- paste0(batchList,"_",focalSex,"_",focalTreatment,"_",minGenes,"genes_",minLogCPM,"lCPM")
  
  #modelFemreml <- read.table("~/Dropbox (Personal)/meerkats/RNA/fullAnalysis/output/allBatches_F_LPSonly_10000genes_5lCPM_modelF_basic_emmreml_status_perm007")
  
  for (permNum in 1:100) {
    print(paste0(focalTreatment,"_",permNum))
    print(date())
    if (!paste0("allBatches_",focalSex,"_",focalTreatment,"_10000genes_5lCPM_model",modelName,"_",LM,
              "wKmatrix_emmreml_",permVar,"_v",permVersion,"_perm",formatC(permNum, width=3, flag="0")) %in% outputFiles) {
    runEmreml(workDir = workDir, gatherSubset = gatherSubset, model = modelName, LM = LM,
              perms = TRUE, permVar = permVar, permNum = permNum, permVersion = 1, Kmat = TRUE,
              Kfile <- "~/Dropbox (Personal)/meerkats/KMPsamples/KMPrelatedness_matrix.RDS")
    }
  }
}

```



```{r permutation for treatment only}
permVersion <- 1
nPerms <- 100
permVarList <- c("status","weight","preg")

#focalTreatment <- "LPS"
#permVar <- "status"

for (permVar in permVarList) {
  
  for (focalTreatment in c("LPS","Ctrl","Dex","Gard")) {
    
    batchList <- "allBatches"
    focalSex <- "F"
    minGenes <- 10000
    minLogCPM <- 5
    modelName <- "F"
    singleTrt <- TRUE
    
    if (focalTreatment %in% c("LPS","Ctrl")) {
      modelName <- "F"
    } else {
      modelName <- "G"
    }
    
    gatherSubset <- paste0(batchList,"_",focalSex,"_",focalTreatment,"only_",minGenes,"genes_",minLogCPM,"lCPM")
    
    #modelFemreml <- read.table("~/Dropbox (Personal)/meerkats/RNA/fullAnalysis/output/allBatches_F_LPSonly_10000genes_5lCPM_modelF_basic_emmreml_status_perm007")
    
    outputFiles <- list.files(outDir)
    #permNum <- 8
    
    for (permNum in 1:nPerms) {
      print(paste0(focalTreatment,"_",permNum))
      print(date())
      if (!paste0(gatherSubset,"_model",modelName,"_",LM,
              "wKmatrix_emmreml_",permVar,"_v",permVersion,"_perm",formatC(permNum, width=3, flag="0")) %in% outputFiles) {
        runEmreml(workDir = workDir, gatherSubset = gatherSubset, model = modelName, LM = "basic+Preg+Weight",
                  perms = TRUE, permVar = permVar, permNum = permNum, permVersion = 1, Kmat <- TRUE,
                  Kfile <- "~/Dropbox (Personal)/meerkats/KMPsamples/KMPrelatedness_matrix.RDS")
      }
    }
  }
}
```

```{r permutation for all treatment}

batchList <- "allBatches"
focalSex <- "F"
focalTreatment <- "allTrt"
minGenes <- 10000
minLogCPM <- 5
modelName <- "G"
singleTrt <- FALSE
combineTrt <- TRUE

gatherSubset <- paste0(batchList,"_",focalSex,"_allTrt_",minGenes,"genes_",minLogCPM,"lCPM")

#modelFemreml <- read.table("~/Dropbox (Personal)/meerkats/RNA/fullAnalysis/output/allBatches_F_LPSonly_10000genes_5lCPM_modelF_basic_emmreml_status_perm007")

#runEmreml(workDir =  workDir, gatherSubset = gatherSubset, model = modelName, LM = "allTrtF+Preg+Weight", Kmat = TRUE, Kfile = "~/Dropbox (Personal)/meerkats/KMPsamples/KMPrelatedness_matrix.RDS")
nPerms <- 100
LM <- "allTrtF+Preg+Weight"

permVersion <- 1
permVar <- "status"

outputFiles <- list.files(outDir)

for (permNum in 1:nPerms) {
  print(permNum)
  print(date())
  if (!paste0("allBatches_F_",focalTreatment,"_10000genes_5lCPM_model",modelName,"_",LM,
              "wKmatrix_emmreml_",permVar,"_v",permVersion,"_perm",formatC(permNum, width=3, flag="0")) %in% outputFiles) {
    runEmreml(workDir = workDir, gatherSubset = gatherSubset, model = modelName, LM = "allTrtF+Preg+Weight",
              perms = TRUE, permVar = permVar, permNum = permNum, permVersion = permVersion, Kmat = TRUE,
              Kfile <- "~/Dropbox (Personal)/meerkats/KMPsamples/KMPrelatedness_matrix.RDS")
  }
}

```

```{r permutation for male and female treatment}
batchList <- "allBatches"
focalTreatment <- "LPS"
minGenes <- 10000
minLogCPM <- 5

Kmat <- TRUE
Kfile <- "~/Dropbox (Personal)/meerkats/KMPsamples/KMPrelatedness_matrix.RDS"

singleTrt <- FALSE
combineTrt <- FALSE

 if (singleTrt) {
   gatherSubset <- paste0(batchList,"_",focalSex,"_",focalTreatment,"only_",minGenes,"genes_",minLogCPM,"lCPM")
 } else if (combineTrt)  {
   gatherSubset <- paste0(batchList,"_",focalSex,"_allTrt_",minGenes,"genes_",minLogCPM,"lCPM")
 } else {
   gatherSubset <- paste0(batchList,"_",focalSex,"_",focalTreatment,"_",minGenes,"genes_",minLogCPM,"lCPM")
 }

perms <- TRUE
permVersion <- 1
permVar <- "treatment"

for (focalSex in c("F","M")) {
  for (focalTreatment in c("LPS","Dex","Gard")) {

    if (focalTreatment == "LPS") {
      modelName <- "F"
    } else {
      modelName <- "G"
    }
    
    gatherSubset <- paste0(batchList,"_",focalSex,"_",focalTreatment,"_",minGenes,"genes_",minLogCPM,"lCPM")
    
    nPerms <- 100
    if (focalSex == "F") {
      LM <- "basic+Preg+Weight"
    } else {
      LM <- "basic+Weight"
    }
    
    outputFiles <- list.files(outDir)
    #read.table("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_M_LPS_10000genes_5lCPM_modelF_basic+WeightwKmatrixwKmatrix_emmreml_treatment_v1_perm001")
    
    for (permNum in 1:nPerms) {
      print(permNum)
      print(date())
      if (!paste0("allBatches_",focalSex,"_",focalTreatment,"_10000genes_5lCPM_model",modelName,"_",LM,
                  "wKmatrix_emmreml_",permVar,"_v",permVersion,"_perm",formatC(permNum, width=3, flag="0")) %in% outputFiles) {
        runEmreml(workDir = workDir, gatherSubset = gatherSubset, model = modelName, LM = LM,
                  perms = TRUE, permVar = permVar, permNum = permNum, permVersion = permVersion, Kmat = TRUE,
                  Kfile <- "~/Dropbox (Personal)/meerkats/KMPsamples/KMPrelatedness_matrix.RDS")
      }
    }
  }
}
```

```{r plot treatment permutation p-values}
permVersion <- 1
permVar <- "treatment"

#focalTreatment <- "LPS"
for (focalTreatment in c("LPS","Dex","Gard")) {
  
  for (focalSex in c("F","M","MF")) {
    
    if (focalSex == "MF") {
      LM <- "basicMF+Weight"
    } else if (focalSex == "F") {
      LM <- "basic+Preg+Weight"
    } else {
      LM <- "basic+Weight"
    }
  
    #focalSex <- "MF"
    minGenes <- 10000
    minLogCPM <- 5
    if (focalTreatment %in% c("Dex","Gard")) {
      modelName <- "G"
    } else {
      modelName <- "F"
    }
    
    gatherSubset <- paste0(batchList,"_",focalSex,"_",focalTreatment,"_",minGenes,"genes_",minLogCPM,"lCPM")
    
    modelDir <- paste0(figDir,gatherSubset,"/model",modelName)
    

    if (permVersion == 1) {
      LMvarList <- permVar
    } else {
      LMvarList <- c("status","age","weight","preg")
    }
    
    #LMvar <- "treatment"
    for (LMvar in LMvarList) {
      
      #permVar is always "status" - labeled the original as such
      #permVar <- "status"
      #LMvar <- "treatment"
      
      totalPerms <- 100
      #get Emreml
      #"~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_MF_LPS_10000genes_5lCPM_modelF_basicMF+WeightwKmatrix_emmreml"
      emremlOut <- read.table(paste0(outDir,gatherSubset,"_model",modelName,"_",LM,"wKmatrix_emmreml"))
      
      #plot standard p-value
      ggplot(data = emremlOut, aes_string(x = paste0("pval_",LMvar))) +
        geom_histogram(bins = 100)
      
      #allBatches_F_LPS_10000genes_4lCPM_modelF_nested_emmreml_treatment_perm001
      permRun <- 8
      
      permPvals <- vector()
      
      for (permRun in 1:totalPerms) {
        #get permutations
        #"~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_LPSonly_10000genes_5lCPM_modelF_basic+Preg+WeightwKmatrix_emmreml_status_perm008"
        #"~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_LPSonly_10000genes_5lCPM_modelF_basic+Preg+WeightwKmatrix_emmreml_status_v1_perm008"
        permEmreml <- read.table(paste0(outDir,gatherSubset,"_model",modelName,"_",LM,"wKmatrix_emmreml_",
                                        permVar,"_v",permVersion,"_perm",formatC(permRun,width = 3, flag = "0")))
        
        permCol <- which(colnames(permEmreml) == paste0("pval_",LMvar))
        
        permPvals <- c(permPvals,permEmreml[,permCol])
      }
      
      #permutation pvals divided by number of runs should match emreml
      
      ( length(permPvals) / totalPerms ) == dim(emremlOut)[1]
      
      emremlOut <- perm.fdr(input_df = emremlOut,
                            perm_df = permPvals,
                            Pvals_col_name = paste0("pval_",LMvar),
                            name = LMvar)
      
      
      
      ggplot() +
        geom_histogram(aes_string(x = permPvals), bins = 100, fill = "tomato", alpha = .5) +
        xlab(paste0("permuted p-value - ",LMvar)) + 
        ggtitle(paste0(focalTreatment,"only - ",LM," | ",permVar," permutations"), 
                subtitle = paste0(totalPerms," permutation runs, model", modelName))
      ggsave(paste0(modelDir,"/perms/",LM,"_pval_",LMvar,"_v",permVersion,"_wPerms.png"), width = 6, height = 4)
      
      fdrCol <- which(colnames(emremlOut) == paste0("fdr_",LMvar))
      
      ggplot() +
        geom_histogram(aes(x = emremlOut[,permCol], y = ..density..), fill = "dodgerblue2", bins = 100, alpha = .5) +
        geom_density(aes(x = permPvals, y = ..density..), adjust = .25, fill = "tomato", color = "tomato", alpha = .25) +
        annotate("text", x = Inf, y = Inf, vjust = 1, hjust = 1,
                 label = paste0("# Genes < FDR\n",
                                sum(emremlOut[,fdrCol] < .01)," | 0.01\n",
                                sum(emremlOut[,fdrCol] < .10)," | 0.10\n",
                                sum(emremlOut[,fdrCol] < .20)," | 0.20")) +
        xlab(paste0("p-value - ",LMvar)) + 
        ggtitle(paste0(focalTreatment,"only - ",LM," | ",permVar," permutations"), 
                subtitle = paste0(totalPerms," permutation runs, model", modelName,"permuted p-value density plot in red"))
      ggsave(paste0(modelDir,"/",LM,"_pval_",LMvar,"_v",permVersion,"_wPerms.png"), width = 6, height = 4)
      
      write.table(x = emremlOut, file = paste0(outDir,gatherSubset,"_model",modelName,"_",LM,
                                               "wKmatrix_emmreml_wPermFDR_v",permVersion,"_",permVar))
      #read.table(paste0(outDir,gatherSubset,"_model",modelName,"_",LM,"wKmatrix_emmreml"))
    }
  
  }
  
}

```


```{r plot treatment permutation p-values}
permVersion <- 1
permVar <- "treatment"

#focalTreatment <- "LPS"
for (focalSex in c("M","F")) {
  for (focalTreatment in c("LPS","Dex","Gard")) {
    
    #focalSex <- "MF"
    minGenes <- 10000
    minLogCPM <- 5
    if (focalTreatment %in% c("Dex","Gard")) {
      modelName <- "G"
    } else {
      modelName <- "F"
    }
    
    gatherSubset <- paste0(batchList,"_",focalSex,"_",focalTreatment,"_",minGenes,"genes_",minLogCPM,"lCPM")
    
    modelDir <- paste0(figDir,gatherSubset,"/model",modelName)
    
    if (focalSex == "F") {
      LM <- "basic+Preg+Weight"
    } else {
      LM <- "basic+Weight"
    }
    
    if (permVersion == 1) {
      LMvarList <- permVar
    } else {
      LMvarList <- c("status","age","weight","preg")
    }
    
    #LMvar <- "treatment"
    for (LMvar in LMvarList) {
      
      #permVar is always "status" - labeled the original as such
      #permVar <- "status"
      #LMvar <- "treatment"
      
      totalPerms <- 100
      #get Emreml
      #"~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_MF_LPS_10000genes_5lCPM_modelF_basicMF+WeightwKmatrix_emmreml"
      emremlOut <- read.table(paste0(outDir,gatherSubset,"_model",modelName,"_",LM,"wKmatrix_emmreml"))
      
      #plot standard p-value
      ggplot(data = emremlOut, aes_string(x = paste0("pval_",LMvar))) +
        geom_histogram(bins = 100)
      
      #allBatches_F_LPS_10000genes_4lCPM_modelF_nested_emmreml_treatment_perm001
      permRun <- 8
      
      permPvals <- vector()
      
      for (permRun in 1:totalPerms) {
        #get permutations
        #"~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_LPSonly_10000genes_5lCPM_modelF_basic+Preg+WeightwKmatrix_emmreml_status_perm008"
        #"~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_LPSonly_10000genes_5lCPM_modelF_basic+Preg+WeightwKmatrix_emmreml_status_v1_perm008"
        permEmreml <- read.table(paste0(outDir,gatherSubset,"_model",modelName,"_",LM,"wKmatrix_emmreml_",
                                        permVar,"_v",permVersion,"_perm",formatC(permRun,width = 3, flag = "0")))
        
        permCol <- which(colnames(permEmreml) == paste0("pval_",LMvar))
        
        permPvals <- c(permPvals,permEmreml[,permCol])
      }
      
      #permutation pvals divided by number of runs should match emreml
      
      ( length(permPvals) / totalPerms ) == dim(emremlOut)[1]
      
      emremlOut <- perm.fdr(input_df = emremlOut,
                            perm_df = permPvals,
                            Pvals_col_name = paste0("pval_",LMvar),
                            name = LMvar)
      
      
      
      ggplot() +
        geom_histogram(aes_string(x = permPvals), bins = 100, fill = "tomato", alpha = .5) +
        xlab(paste0("permuted p-value - ",LMvar)) + 
        ggtitle(paste0(focalTreatment,"only - ",LM," | ",permVar," permutations"), 
                subtitle = paste0(totalPerms," permutation runs, model", modelName))
      ggsave(paste0(modelDir,"/perms/",LM,"_pval_",LMvar,"_v",permVersion,"_wPerms.png"), width = 6, height = 4)
      
      fdrCol <- which(colnames(emremlOut) == paste0("fdr_",LMvar))
      
      ggplot() +
        geom_histogram(aes(x = emremlOut[,permCol], y = ..density..), fill = "dodgerblue2", bins = 100, alpha = .5) +
        geom_density(aes(x = permPvals, y = ..density..), adjust = .25, fill = "tomato", color = "tomato", alpha = .25) +
        annotate("text", x = Inf, y = Inf, vjust = 1, hjust = 1,
                 label = paste0("# Genes < FDR\n",
                                sum(emremlOut[,fdrCol] < .01)," | 0.01\n",
                                sum(emremlOut[,fdrCol] < .10)," | 0.10\n",
                                sum(emremlOut[,fdrCol] < .20)," | 0.20")) +
        xlab(paste0("p-value - ",LMvar)) + 
        ggtitle(paste0(focalTreatment,"only - ",LM," | ",permVar," permutations"), 
                subtitle = paste0(totalPerms," permutation runs, model", modelName,"permuted p-value density plot in red"))
      ggsave(paste0(modelDir,"/",LM,"_pval_",LMvar,"_v",permVersion,"_wPerms.png"), width = 6, height = 4)
      
      write.table(x = emremlOut, file = paste0(outDir,gatherSubset,"_model",modelName,"_",LM,
                                               "wKmatrix_emmreml_wPermFDR_v",permVersion,"_",permVar))
      #read.table(paste0(outDir,gatherSubset,"_model",modelName,"_",LM,"wKmatrix_emmreml"))
    }
    
  }
}

```

```{r plot permutation p-values}
permVersion <- 1
permVar <- "status"

for (permVar in c("status","preg","weight")) {
  for (focalTreatment in c("Ctrl","LPS","Dex","Gard")) {
    batchList <- "allBatches"
    focalSex <- "F"
    #focalTreatment <- "LPS"
    minGenes <- 10000
    minLogCPM <- 5
    if (focalTreatment %in% c("Dex","Gard")) {
      modelName <- "G"
    } else {
      modelName <- "F"
    }
    
    gatherSubset <- paste0(batchList,"_",focalSex,"_",focalTreatment,"only_",minGenes,"genes_",minLogCPM,"lCPM")
    #permVersion <- 3
    
    modelDir <- paste0(figDir,gatherSubset,"/model",modelName)
    
    LM <- "basic+Preg+Weight"
    if (permVersion == 1) {
      LMvarList <- permVar
    } else {
      LMvarList <- c("status","age","weight","preg")
    }
    
    for (LMvar in LMvarList) {
      
      #permVar is always "status" - labeled the original as such
      #permVar <- "status"
      #LMvar <- "status"
      
      totalPerms <- 100
      #get Emreml
      emremlOut <- read.table(paste0(outDir,gatherSubset,"_model",modelName,"_",LM,"wKmatrix_emmreml"))
      
      #plot standard p-value
      ggplot(data = emremlOut, aes_string(x = paste0("pval_",LMvar))) +
        geom_histogram(bins = 100)
      
      #allBatches_F_LPS_10000genes_4lCPM_modelF_nested_emmreml_treatment_perm001
      permRun <- 8
      
      permPvals <- vector()
      
      for (permRun in 1:totalPerms) {
        #get permutations
        #"~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_LPSonly_10000genes_5lCPM_modelF_basic+Preg+WeightwKmatrix_emmreml_status_perm008"
        #"~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_LPSonly_10000genes_5lCPM_modelF_basic+Preg+WeightwKmatrix_emmreml_status_v1_perm008"
        permEmreml <- read.table(paste0(outDir,gatherSubset,"_model",modelName,"_",LM,"wKmatrix_emmreml_",
                                        permVar,"_v",permVersion,"_perm",formatC(permRun,width = 3, flag = "0")))
        
        permCol <- which(colnames(permEmreml) == paste0("pval_",LMvar))
        
        permPvals <- c(permPvals,permEmreml[,permCol])
      }
      
      #permutation pvals divided by number of runs should match emreml
      
      ( length(permPvals) / totalPerms ) == dim(emremlOut)[1]
      
      emremlOut <- perm.fdr(input_df = emremlOut,
                            perm_df = permPvals,
                            Pvals_col_name = paste0("pval_",LMvar),
                            name = LMvar)
      
      
      
      ggplot() +
        geom_histogram(aes_string(x = permPvals), bins = 100, fill = "tomato", alpha = .5) +
        xlab(paste0("permuted p-value - ",LMvar)) + 
        ggtitle(paste0(focalTreatment,"only - ",LM," | ",permVar," permutations"), 
                subtitle = paste0(totalPerms," permutation runs, model", modelName))
      ggsave(paste0(modelDir,"/perms/",LM,"_pval_",LMvar,"_v",permVersion,"_wPerms.png"), width = 6, height = 4)
      
      fdrCol <- which(colnames(emremlOut) == paste0("fdr_",LMvar))
      
      ggplot() +
        geom_histogram(aes(x = emremlOut[,permCol], y = ..density..), fill = "dodgerblue2", bins = 100, alpha = .5) +
        geom_density(aes(x = permPvals, y = ..density..), adjust = .25, fill = "tomato", color = "tomato", alpha = .25) +
        annotate("text", x = Inf, y = Inf, vjust = 1, hjust = 1,
                 label = paste0("# Genes < FDR\n",
                                sum(emremlOut[,fdrCol] < .01)," | 0.01\n",
                                sum(emremlOut[,fdrCol] < .10)," | 0.10\n",
                                sum(emremlOut[,fdrCol] < .20)," | 0.20")) +
        xlab(paste0("p-value - ",LMvar)) + 
        ggtitle(paste0(focalTreatment,"only - ",LM," | ",permVar," permutations"), 
                subtitle = paste0(totalPerms," permutation runs, model", modelName,"permuted p-value density plot in red"))
      ggsave(paste0(modelDir,"/",LM,"_pval_",LMvar,"_v",permVersion,"_wPerms.png"), width = 6, height = 4)
      
      write.table(x = emremlOut, file = paste0(outDir,gatherSubset,"_model",modelName,"_",LM,
                                               "wKmatrix_emmreml_wPermFDR_v",permVersion,"_",permVar))
      #read.table(paste0(outDir,gatherSubset,"_model",modelName,"_",LM,"wKmatrix_emmreml"))
    }
  }
}

### check the weight-status beta correlations & weight-status actual correlations in the permutations, use LPS
focalTreatment <- "LPS"

batchList <- "allBatches"
focalSex <- "F"
#focalTreatment <- "LPS"
minGenes <- 10000
minLogCPM <- 5
if (focalTreatment %in% c("Dex","Gard")) {
  modelName <- "G"
} else {
  modelName <- "F"
}

gatherSubset <- paste0(batchList,"_",focalSex,"_",focalTreatment,"only_",minGenes,"genes_",minLogCPM,"lCPM")
permVersion <- 3

modelDir <- paste0(figDir,gatherSubset,"/model",modelName)

LM <- "basic+Preg+Weight"

permVar <- "status"
LMvar <- "status"

totalPerms <- 100
#get Emreml
emremlOut <- read.table(paste0(outDir,gatherSubset,"_model",modelName,"_",LM,"wKmatrix_emmreml"))

trueWghtStatR <- cor(emremlOut$beta_weight,emremlOut$beta_status)

trueMetadata <- readRDS("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_LPSonly_10000genes_5lCPM_metadata.RDS")

#match emmreml code
trueMetadata$weight <- trueMetadata$avgWeight6mo
trueMetadata$weightUnScaled <- trueMetadata$weight
animalWeights <- unique(trueMetadata$weight)
scaledWeights <- scale(animalWeights, scale = F, center = T)
for (n in 1:dim(trueMetadata)[1]) {
  matchRow <- which(animalWeights == trueMetadata$weightUnScaled[n])
  trueMetadata$weight[n] <- scaledWeights[matchRow]
}



trueMetaWghtStatR <- cor(ifelse(trueMetadata$status_dom_sub == "S", 0, 1),trueMetadata$weight)


#plot standard p-value
ggplot(data = emremlOut, aes_string(x = paste0("pval_",LMvar))) +
  geom_histogram(bins = 100)

#allBatches_F_LPS_10000genes_4lCPM_modelF_nested_emmreml_treatment_perm001
permRun <- 8

permWghtStatR <- vector()
metaWghtStatR <- vector()

for (permRun in 1:totalPerms) {
  #get permutations
  #"~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_LPSonly_10000genes_5lCPM_modelF_basic+Preg+WeightwKmatrix_emmreml_status_perm008"
  permEmreml <- read.table(paste0(outDir,gatherSubset,"_model",modelName,"_",LM,"wKmatrix_emmreml_",
                                  permVar,"_perm",formatC(permRun,width = 3, flag = "0")))
  
  #get permuted metadata
  #"~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_LPSonly_10000genes_5lCPM_metadata_status_perm008.RDS"
  permMeta <- readRDS(paste0(outDir,gatherSubset,"_metadata_",permVar,"_perm",formatC(permRun,width = 3, flag = "0"),".RDS"))
  
  metaWghtStatR <- c(metaWghtStatR,cor(as.numeric(permMeta$status), permMeta$weight))
  
  statBetaCol <- which(colnames(permEmreml) == paste0("beta_status"))
  weightBetaCol <- which(colnames(permEmreml) == paste0("beta_weight"))
  
  permWghtStatR <- c(permWghtStatR,cor(permEmreml[,weightBetaCol],permEmreml[,statBetaCol]))
}

ggplot() +
  geom_histogram(aes(x = permWghtStatR), fill = "tomato", bins = 20, alpha = .5) +
  geom_vline(xintercept = mean(permWghtStatR), col = "tomato", lwd = 2) +
  geom_vline(xintercept = trueWghtStatR, col = "dodgerblue3", lwd = 2) +
  geom_vline(xintercept = 0, col = "grey50", lwd = .5) +
  xlim(c(-.66,.66)) +
  xlab(paste0("Beta_Status ~ Beta_Weight Correlations"))
ggsave(paste0(modelDir,"/",LM,"_status_weight_betaCorr_wPerms.png"), width = 6, height = 4)

ggplot() +
  geom_histogram(aes(x = metaWghtStatR), fill = "tomato", bins = 10, alpha = .5) +
  geom_vline(xintercept = mean(metaWghtStatR), col = "tomato", lwd = 2) +
  geom_vline(xintercept = trueMetaWghtStatR, col = "dodgerblue3", lwd = 2) +
  geom_vline(xintercept = 0, col = "grey50", lwd = .5) +
  xlab(paste0("Status ~ Weight Correlations"))
ggsave(paste0(modelDir,"/",LM,"_status_weight_rawCorr_wPerms.png"), width = 6, height = 4)
```

```{r plot all treatment perm}

batchList <- "allBatches"
focalSex <- "F"
focalTreatment <- "allTrt"
minGenes <- 10000
minLogCPM <- 5
modelName <- "G"

gatherSubset <- paste0(batchList,"_",focalSex,"_",focalTreatment,"_",minGenes,"genes_",minLogCPM,"lCPM")
permVersion <- 1

modelDir <- paste0(figDir,gatherSubset,"/model",modelName)

LM <- "allTrtF+Preg+Weight"
for (LMvar in c("status")) {
  
  #permVar is always "status" - labeled the original as such
  permVar <- "status"
  #LMvar <- "status"
  
  totalPerms <- 100
  #get Emreml
  emremlOut <- read.table(paste0(outDir,gatherSubset,"_model",modelName,"_",LM,"wKmatrix_emmreml"))
  
  trueWghtStatR <- cor(emremlOut$beta_weight,emremlOut$beta_status)
  
  if (LMvar == "status") {
    trueMetadata <- readRDS("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_allTrt_10000genes_5lCPM_metadata.RDS")
    
    #match emmreml code
    trueMetadata$weight <- trueMetadata$avgWeight6mo
    trueMetadata$weightUnScaled <- trueMetadata$weight
    animalWeights <- unique(trueMetadata$weight)
    scaledWeights <- scale(animalWeights, scale = F, center = T)
    for (n in 1:dim(trueMetadata)[1]) {
      matchRow <- which(animalWeights == trueMetadata$weightUnScaled[n])
      trueMetadata$weight[n] <- scaledWeights[matchRow]
    }
    
     permWghtStatR <- vector()
     metaWghtStatR <- vector()
  }

  trueMetaWghtStatR <- cor(ifelse(trueMetadata$status_dom_sub == "S", 0, 1),trueMetadata$weight)
  
  #plot standard p-value
  ggplot(data = emremlOut, aes_string(x = paste0("pval_",LMvar))) +
    geom_histogram(bins = 100)
  
  #allBatches_F_LPS_10000genes_4lCPM_modelF_nested_emmreml_treatment_perm001
  permRun <- 3
  
  permPvals <- vector()
  
 

  
  for (permRun in 1:totalPerms) {
    #get permutations
    #"~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_LPSonly_10000genes_5lCPM_modelF_basic+Preg+WeightwKmatrix_emmreml_status_perm008"
    
    permEmreml <- read.table(paste0(outDir,gatherSubset,"_model",modelName,"_",LM,"wKmatrix_emmreml_",
                                    permVar,"_v",permVersion,"_perm",formatC(permRun,width = 3, flag = "0")))
    
    permCol <- which(colnames(permEmreml) == paste0("pval_",LMvar))
    
    permPvals <- c(permPvals,permEmreml[,permCol])
    
    if (LMvar == "status") {
      #get permuted metadata
      #"~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_LPSonly_10000genes_5lCPM_metadata_status_perm008.RDS"
      permMeta <- readRDS(paste0(outDir,gatherSubset,"_metadata_",permVar,"_v",permVersion,"_perm",formatC(permRun,width = 3, flag = "0"),".RDS"))
      
      metaWghtStatR <- c(metaWghtStatR,cor(as.numeric(permMeta$status), permMeta$weight))
      
      statBetaCol <- which(colnames(permEmreml) == paste0("beta_status"))
      weightBetaCol <- which(colnames(permEmreml) == paste0("beta_weight"))
      
      permWghtStatR <- c(permWghtStatR,cor(permEmreml[,weightBetaCol],permEmreml[,statBetaCol]))
    }
  }
  
  #permutation pvals divided by number of runs should match emreml
  
  ( length(permPvals) / totalPerms ) == dim(emremlOut)[1]
  
  emremlOut <- perm.fdr(input_df = emremlOut,
                        perm_df = permPvals,
                        Pvals_col_name = paste0("pval_",LMvar),
                        name = LMvar)
  
  
  
  ggplot() +
    geom_histogram(aes_string(x = permPvals), bins = 100, fill = "tomato", alpha = .5) +
    xlab(paste0("permuted p-value - ",LMvar)) + 
    ggtitle(paste0(focalTreatment,"only - ",LM," | ",permVar," permutations"), 
            subtitle = paste0(totalPerms," permutation runs, model", modelName))
  ggsave(paste0(modelDir,"/perms/",LM,"_pval_",LMvar,"_wPerms.png"), width = 6, height = 4)
  
  fdrCol <- which(colnames(emremlOut) == paste0("fdr_",LMvar))
  
  ggplot() +
    geom_histogram(aes(x = emremlOut[,permCol], y = ..density..), fill = "dodgerblue2", bins = 100, alpha = .5) +
    geom_density(aes(x = permPvals, y = ..density..), adjust = .25, fill = "tomato", color = "tomato", alpha = .25) +
    annotate("text", x = Inf, y = Inf, vjust = 1, hjust = 1,
             label = paste0("# Genes < FDR\n",
                            sum(emremlOut[,fdrCol] < .01)," | 0.01\n",
                            sum(emremlOut[,fdrCol] < .10)," | 0.10\n",
                            sum(emremlOut[,fdrCol] < .20)," | 0.20")) +
    xlab(paste0("p-value - ",LMvar)) + 
    ggtitle(paste0(focalTreatment,"only - ",LM," | ",permVar," permutations"), 
            subtitle = paste0(totalPerms," permutation runs, model", modelName,"permuted p-value density plot in red"))
  ggsave(paste0(modelDir,"/",LM,"_pval_",LMvar,"_wPerms.png"), width = 6, height = 4)
  
  if (LMvar == "status") {
    ggplot() +
      geom_histogram(aes(x = permWghtStatR), fill = "tomato", bins = 20, alpha = .5) +
      geom_vline(xintercept = mean(permWghtStatR), col = "tomato", lwd = 2) +
      geom_vline(xintercept = trueWghtStatR, col = "dodgerblue3", lwd = 2) +
      geom_vline(xintercept = 0, col = "grey50", lwd = .5) +
      xlim(c(-.66,.66)) +
      xlab(paste0("Beta_Status ~ Beta_Weight Correlations"))
    ggsave(paste0(modelDir,"/",LM,"_status_weight_betaCorr_wPerms.png"), width = 6, height = 4)
    
    ggplot() +
      geom_histogram(aes(x = metaWghtStatR), fill = "tomato", bins = 10, alpha = .5) +
      geom_vline(xintercept = mean(metaWghtStatR), col = "tomato", lwd = 2) +
      geom_vline(xintercept = trueMetaWghtStatR, col = "dodgerblue3", lwd = 2) +
      geom_vline(xintercept = 0, col = "grey50", lwd = .5) +
      xlab(paste0("Status ~ Weight Correlations"))
    ggsave(paste0(modelDir,"/",LM,"_status_weight_rawCorr_wPerms.png"), width = 6, height = 4)
  }
}


```



investigating the effect of status on response
```{r}
trtThresh <- 1
qvalThresh <- .1
trt <- "LPS"
fxVar <- "status"

if (trt == "LPS") {
  model <- "F"
} else {
  model <- "G"
}

#get list of genes sig (qval < .1) in Ctrl only, LPS only
trtEmreml <- read.table(paste0("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_",trt,
                               "only_10000genes_5lCPM_model",model,"_basic+Preg+WeightwKmatrix_emmreml_wPermFDR_v1_status"))
ctrlEmreml <- read.table(paste0("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_Ctrl",
                               "only_10000genes_5lCPM_modelF_basic+Preg+WeightwKmatrix_emmreml_wPermFDR_v1_status"))
pvalCol <- which(colnames(trtEmreml) == paste0("pval_",fxVar))
betaCol <- which(colnames(trtEmreml) == paste0("beta_",fxVar))

#get LPS-Ctrl response betas
respEmreml <- read.table(paste0("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_",trt,
                                "_10000genes_5lCPM_model",model,"_basicResp+Preg+Weight_emmreml"))

respEmreml$stdbeta_status <- respEmreml$beta_status / sqrt(respEmreml$var_beta_status)

trtCtrlEmreml <- read.table(paste0("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_",trt,
                                   "_10000genes_5lCPM_model",model,"_basic+Preg+WeightwKmatrix_emmreml"))



trtCtrlEmremlPerm <- read.table(paste0("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_",trt,
                                   "_10000genes_5lCPM_model",model,"_basic+Preg+WeightwKmatrix_emmreml_wPermFDR_v1_treatment"))

trtCtrlEmremlPerm <- trtCtrlEmremlPerm[order(rownames(trtCtrlEmremlPerm)),]



### to respEmreml, need to add:
# treatment effect - is the gene up or down reg in LPS Treatment
trtUpGeneList <- rownames(trtCtrlEmreml)[trtCtrlEmreml$beta_treatment > 0]
trtDnGeneList <- rownames(trtCtrlEmreml)[trtCtrlEmreml$beta_treatment < 0]

# status effect at LPS - is the gene up or down reg w.r.t. status?
trtSigUpGeneList <- rownames(trtEmreml)[trtEmreml$fdr_status < qvalThresh & trtEmreml[,betaCol] > 0]
trtSigDnGeneList <- rownames(trtEmreml)[trtEmreml$fdr_status < qvalThresh & trtEmreml[,betaCol] < 0]

# status effect at Ctrl - is the gene up or down reg w.r.t. status?
ctrlSigUpGeneList <- rownames(ctrlEmreml)[ctrlEmreml$fdr_status < qvalThresh & ctrlEmreml[,betaCol] > 0]
ctrlSigDnGeneList <- rownames(ctrlEmreml)[ctrlEmreml$fdr_status < qvalThresh & ctrlEmreml[,betaCol] < 0]

# status effect overall, is the gene up or down reg w.r.t. status in both?

respEmreml$trtEffect <- ifelse(rownames(respEmreml) %in% trtUpGeneList, "Up",
                               ifelse(rownames(respEmreml) %in% trtDnGeneList, "Down", "none"))

respEmreml$statEffectCtrl <- ifelse(rownames(respEmreml) %in% ctrlSigUpGeneList, "Up",
                               ifelse(rownames(respEmreml) %in% ctrlSigDnGeneList, "Down", "none"))

respEmreml$statEffectTrt <- ifelse(rownames(respEmreml) %in% trtSigUpGeneList, "Up",
                               ifelse(rownames(respEmreml) %in% trtSigDnGeneList, "Down", "none"))

respEmreml$statEffectBoth <- ifelse(respEmreml$statEffectCtrl == "Up" | respEmreml$statEffectTrt == "Up", "Up",
                                    ifelse(respEmreml$statEffectCtrl == "Down" | respEmreml$statEffectTrt == "Down", "Down", "none"))


respEmreml$groupFig3 <- ifelse(respEmreml$trtEffect == "Up" & respEmreml$statEffectBoth == "Down", "GroupI",
                               ifelse(respEmreml$trtEffect == "Up" & respEmreml$statEffectBoth == "Up", "GroupII",
                                      ifelse(respEmreml$trtEffect == "Down" & respEmreml$statEffectBoth == "Down", "GroupIII",
                                             ifelse(respEmreml$trtEffect == "Down" & respEmreml$statEffectBoth == "Up", "GroupIV","none"))))


print(table(respEmreml$groupFig3))

respEmremlSigInStatus <- subset(respEmreml, groupFig3 != "none")

ggplot() +
  geom_histogram(aes(x = respEmremlSigInStatus$stdbeta_status,
                     fill = respEmremlSigInStatus$groupFig3), bins = 50) +
  labs(fill = "Sig\nTrt\nEffect?") +
  geom_vline(xintercept = 0, col = "gray30") +
  ggtitle(paste0(trt,"-Ctrl Response Status Betas"),
          subtitle = paste0("qval thresh q < ",qvalThresh))

ggsave(paste0("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/Response_Status_Betas_by_TrtResp_",
              trt,"vCtrl_q",qvalThresh,"_hist.png"), height = 4, width = 6)

#use subset to test interaction betas
#LPS
nestEmreml <- read.table(paste0("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_",trt,
                                       "_10000genes_5lCPM_model",model,"_nested+Preg+WeightwKmatrix_emmreml"))

#only keep significant genes
#only keep significant genes
statSigGeneList <- c(rownames(trtEmreml)[trtEmreml$fdr_status <= .1],
                     rownames(ctrlEmreml)[ctrlEmreml$fdr_status <= .1])
statSigGeneList <- unique(statSigGeneList)
trtSigGeneList <- rownames(trtCtrlEmreml)[qvalue(trtCtrlEmreml$pval_treatment)$qvalue <= 1]
sigGeneList <- intersect(statSigGeneList,trtSigGeneList)

nestEmreml$beta_interaction <- nestEmreml$beta_Trt.Status - nestEmreml$beta_Ctrl.Status
nestEmreml$var_beta_interaction <- sqrt(nestEmreml$var_beta_Trt.Status^2 +
                                                 nestEmreml$var_beta_Ctrl.Status^2)
nestEmreml$t_interaction <- nestEmreml$beta_interaction / sqrt(nestEmreml$var_beta_interaction)
nestEmreml$pval_interaction <- 2*pt(q=abs(nestEmreml$t_interaction), df=10000, lower.tail=FALSE)

nestEmremlSigStatus <- nestEmreml[rownames(nestEmreml) %in% sigGeneList,]

ggplot() +
  geom_histogram(aes(x = nestEmremlSigStatus$pval_interaction), bins = 100)

ggplot() +
  geom_histogram(aes(x = nestEmremlSigStatus$pval_treatment), bins = 100)
ggplot() +
  geom_histogram(aes(x = nestEmremlSigStatus$pval_Trt.Status), bins = 100)
ggplot() +
  geom_histogram(aes(x = nestEmremlSigStatus$pval_Ctrl.Status), bins = 100)

sum(qvalue(nestEmremlSigStatus$pval_interaction)$qvalue < .1)

sigInteractionLPSgene <- rownames(nestEmremlSigStatus)[qvalue(nestEmremlSigStatus$pval_interaction)$qvalue < .1]

grpIsigInt <- sigInteractionLPSgene[sigInteractionLPSgene %in% rownames(subset(respEmreml, groupFig3 == "GroupI"))]
grpIIsigInt <- sigInteractionLPSgene[sigInteractionLPSgene %in% rownames(subset(respEmreml, groupFig3 == "GroupII"))]
grpIIIsigInt <- sigInteractionLPSgene[sigInteractionLPSgene %in% rownames(subset(respEmreml, groupFig3 == "GroupIII"))]

unique(gene.set[gene.set$Associated.Gene.Name %in% grpIsigInt,])
unique(gene.set[gene.set$Associated.Gene.Name %in% grpIIsigInt,])
unique(gene.set[gene.set$Associated.Gene.Name %in% grpIsigInt,])
    

groupIIgeneList <- rownames(subset(respEmreml, groupFig3 == "GroupII"))
groupIgeneList <- rownames(subset(respEmreml, groupFig3 == "GroupI"))
groupIIIgeneList <- rownames(subset(respEmreml, groupFig3 == "GroupIII"))
noGroupGeneList <- rownames(subset(respEmreml, groupFig3 == "none"))
nongrpIIgeneList <- rownames(subset(respEmreml, groupFig3 != "GroupII"))

gene.set <- read.table(paste0("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/data/","hallmarkPathways_wCategories.csv"), sep = ",")

colnames(gene.set) <- c("Associated.Gene.Name","GO.Term.Accession","Category")

sort(table(unique(gene.set[gene.set$Associated.Gene.Name %in% groupIIgeneList,])$GO.Term.Accession))

### Enrichment analysis, FET for presence/absence in hallmark pathways
fetPVal <- vector()
adjFETpval <- vector()
inPath <- vector()
pathTotal <- vector()
geneGrp <- vector()
oddsRatio <- vector()
pathNames <- vector()

pway <- "HALLMARK_TNFA_SIGNALING_VIA_NFKB"
geneGroup <- "GroupII"

for (geneGroup in c("GroupI","GroupII","GroupIII","GroupIV")) {
  print(geneGroup)
  groupGeneList <- rownames(subset(respEmreml, groupFig3 == geneGroup))
  nonGroupGeneList <- rownames(subset(respEmreml, groupFig3 != geneGroup))
  
  for (pway in unique(subset(gene.set, Category %in% c("immune","signaling"))$GO.Term.Accession)) {
    #check for 2 genes
    pathwayGII <- length(intersect(groupGeneList,
                            unique(gene.set$Associated.Gene.Name[gene.set$GO.Term.Accession == pway])))
    if (pathwayGII >= 2) {
      nPathwayGII <- length(unique(gene.set$Associated.Gene.Name[gene.set$GO.Term.Accession == pway])) - pathwayGII
      nPathwayBck <- length(nonGroupGeneList) - nPathwayGII
      pathwayBck <- length(groupGeneList) - pathwayGII
      
      dat <- data.frame("Pathway" = c(pathwayGII, nPathwayGII),
                        "Non-Pathway" = c(pathwayBck, nPathwayBck),
                        row.names = c(geneGroup, "Background"),
                        stringsAsFactors = FALSE
                        )
      colnames(dat) <- c("Non-Pathway", "Pathway")
      
      #test <- fisher.test(dat, alternative = "greater")
      test <- fisher.test(dat)
      test
      print(paste0(pway," ",test$p.value))
      if (test$p.value < .1) {
        print(dat)
      }
      pathNames <- c(pathNames,pway)
      fetPVal <- c(fetPVal,test$p.value)
      if (geneGroup == "GroupI") {
        adjFETpval <- c(adjFETpval,test$p.value * 11)
      } else if (geneGroup == "GroupII") {
        adjFETpval <- c(adjFETpval,test$p.value * 5)
      } else if (geneGroup == "GroupIII") {
        adjFETpval <- c(adjFETpval,test$p.value * 15)
      } else if (geneGroup == "GroupIV") {
        adjFETpval <- c(adjFETpval,test$p.value * 3)
      } 
      oddsRatio <- c(oddsRatio,test$estimate)
      inPath <- c(inPath,pathwayGII)
      pathTotal <- c(pathTotal,nPathwayGII)
      geneGrp <- c(geneGrp,geneGroup)
    }
  }
}

fetResults <- cbind.data.frame(geneGrp,pathNames,fetPVal,adjFETpval,oddsRatio,inPath,pathTotal)
colnames(fetResults)[2] <- "gseaHallmark"
fetResults$gseaHallmark <- str_remove_all(fetResults$gseaHallmark, pattern = "HALLMARK_")
#fetResults <- fetResults[order(fetResults$fetPVal),]

fetResults <- fetResults[order(fetResults$geneGrp, fetResults$fetPVal),]





subset(fetResults, inPath > 1 & fetPVal < .05 & oddsRatio < 1)

subset(fetResults, adjFETpval < .05 & oddsRatio < 1)


### pick an example gene
lpsEmreml <- read.table("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_LPSonly_10000genes_5lCPM_modelF_basic+Preg+WeightwKmatrix_emmreml")
lpsTrtEmreml <- read.table("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_LPS_10000genes_5lCPM_modelF_basic+Preg+WeightwKmatrix_emmreml")
garEmreml <- read.table("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_Gardonly_10000genes_5lCPM_modelG_basic+Preg+WeightwKmatrix_emmreml")
dexEmreml <- read.table("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_Dexonly_10000genes_5lCPM_modelG_basic+Preg+WeightwKmatrix_emmreml")

possibleGenes <- intersect(intersect(rownames(dexEmreml)[dexEmreml$pval_status < .05],
                              rownames(garEmreml)[garEmreml$pval_status < .05]),
                           intersect(rownames(lpsEmreml)[lpsEmreml$pval_status < .05],
                              rownames(lpsTrtEmreml)[lpsTrtEmreml$pval_treatment < .05]))



for (trt in c("LPS","Dex","Gard")) {
    #trt <- "Dex"
    if (trt == "LPS") {
      model <- "F"
    } else {
      model <- "G"
    }
  
  trtCtrlResids <- readRDS(paste0("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_",trt,
                                  "_10000genes_5lCPM_model",model,"_resids.RDS"))
  trtCtrlMeta <- readRDS(paste0("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_",trt,
                                "_10000genes_5lCPM_metadata.RDS"))
  
  trtCtrlMeta$trtXstatus <- paste0(trtCtrlMeta$status_dom_sub,"_",trtCtrlMeta$Trt)
  trtCtrlMeta$StatusNum <- ifelse(trtCtrlMeta$status_dom_sub == "D", 1, 0)
  
  
  #plotGene <- groupIIgeneList[runif(1,min = 1, max = length(groupIIgeneList))]
  #plotGene <- "IL6"
  plotGene <- "CCR7"
  #plotGene <- "PTGS2"
  
  trtCtrlMeta$geneResid <- trtCtrlResids[which(rownames(trtCtrlResids) == plotGene),]
  
  ggplot(data = trtCtrlMeta,
         aes(x = StatusNum,
             y = geneResid,
             col = Trt)) +
    #geom_violin(draw_quantiles = c(.5)) +
    geom_point(alpha = .9) +
    geom_smooth(method = "lm", formula = "y~x", se = FALSE) +
    xlim(c(-.5,1.5)) +
    xlab("Status, S=0 & D=1") +
    ggtitle(paste0(plotGene," | ",trt))
  ggsave(paste0("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/GenePlot_",plotGene,"_",trt,"_status_lineplot.png"), height = 4, width = 6)
  
  #plot the log2fc response too
  respResids <- readRDS(paste0("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_",trt,
                               "_10000genes_5lCPM_model",model,"_responseResids.RDS"))
  respMeta <- readRDS(paste0("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_",trt,
                             "_10000genes_5lCPM_responseMetadata.RDS"))
  
  respMeta$status_dom_sub <- factor(respMeta$status_dom_sub, levels = c("S","D"))
  
  respMeta$geneResid <- respResids[which(rownames(respResids) == plotGene),]
  
  ggplot(data = respMeta,
         aes(x = status_dom_sub,
             y = geneResid)) +
    geom_boxplot(draw_quantiles = c(.5)) +
    geom_point(alpha = .33) +
    ylab(paste0("Response Residual (",trt," - Ctrl)")) +
    ggtitle(plotGene)
  ggsave(paste0("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/GenePlot_",plotGene,"_",trt,"_Log2FC_boxplot.png"), height = 4, width = 6)
}


respEmreml
```


```{r interaction model from nested}
trt <- "LPS"
model <- "F"

#get list of genes sig (qval < .1) in Ctrl only, LPS only
trtEmreml <- read.table(paste0("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_",trt,
                               "only_10000genes_5lCPM_model",model,"_basic+Preg+WeightwKmatrix_emmreml_wPermFDR_v1_status"))
ctrlEmreml <- read.table(paste0("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_Ctrl",
                               "only_10000genes_5lCPM_modelF_basic+Preg+WeightwKmatrix_emmreml_wPermFDR_v1_status"))

trtCtrlEmreml <- read.table(paste0("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_",trt,
                                       "_10000genes_5lCPM_model",model,"_basic+Preg+WeightwKmatrix_emmreml"))
    
#bring in nested
nestEmreml <- read.table(paste0("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_",trt,
                                       "_10000genes_5lCPM_model",model,"_nested+Preg+WeightwKmatrix_emmreml"))

#only keep significant genes
statSigGeneList <- c(rownames(trtEmreml)[trtEmreml$fdr_status <= .1],
                     rownames(ctrlEmreml)[ctrlEmreml$fdr_status <= .1])
statSigGeneList <- unique(statSigGeneList)
trtSigGeneList <- rownames(trtCtrlEmreml)[qvalue(trtCtrlEmreml$pval_treatment)$qvalue <= .1]
sigGeneList <- intersect(statSigGeneList,trtSigGeneList)

nestEmreml$beta_interaction <- nestEmreml$beta_Trt.Status - nestEmreml$beta_Ctrl.Status
nestEmreml$var_beta_interaction <- sqrt(nestEmreml$var_beta_Trt.Status^2 +
                                                 nestEmreml$var_beta_Ctrl.Status^2)
nestEmreml$t_interaction <- nestEmreml$beta_interaction / sqrt(nestEmreml$var_beta_interaction)
nestEmreml$pval_interaction <- 2*pt(q=abs(nestEmreml$t_interaction), df=10000, lower.tail=FALSE)

nestEmremlSigStatus <- nestEmreml[rownames(nestEmreml) %in% statSigGeneList,]

ggplot() +
  geom_histogram(aes(x = nestEmremlSigStatus$pval_interaction), bins = 100)

ggplot() +
  geom_histogram(aes(x = nestEmremlSigStatus$pval_treatment), bins = 100)
ggplot() +
  geom_histogram(aes(x = nestEmremlSigStatus$pval_Trt.Status), bins = 100)
ggplot() +
  geom_histogram(aes(x = nestEmremlSigStatus$pval_Ctrl.Status), bins = 100)

sum(qvalue(nestEmremlSigStatus$pval_interaction)$qvalue < .1)

sigInteractionLPSgene <- rownames(nestEmremlSigStatus)[qvalue(nestEmremlSigStatus$pval_interaction)$qvalue < .1]

grpIsigInt <- sigInteractionLPSgene[sigInteractionLPSgene %in% rownames(subset(respEmreml, groupFig3 == "GroupI"))]
grpIIsigInt <- sigInteractionLPSgene[sigInteractionLPSgene %in% rownames(subset(respEmreml, groupFig3 == "GroupII"))]
grpIIIsigInt <- sigInteractionLPSgene[sigInteractionLPSgene %in% rownames(subset(respEmreml, groupFig3 == "GroupIII"))]

unique(gene.set[gene.set$Associated.Gene.Name %in% grpIsigInt,])
unique(gene.set[gene.set$Associated.Gene.Name %in% grpIIsigInt,])
unique(gene.set[gene.set$Associated.Gene.Name %in% grpIIIsigInt,])
#add interaction p-values

### run a model/generate plots that are just socially sig genes in Ctrl or LPS


```


GSEA of treatment, status, pregnancy
```{r GSEA}
#emreml results

## loop through data sets

c <- "status"
trt <- "LPS"
thresh <- 5
m <- "emmreml"
sex <- "F"

for (sex in c("M","F")) {
  if (sex == "F") {
    effectSet <- c("status","preg","treatment")
  } else {
    effectSet <- c("status","treatment")
  }
  
  
  #for (thresh in c(5,7,"status")) {
  for (thresh in c(5)) {
    allSigTrtPathways <- data.frame()
    allTrtPathways <- data.frame()
    
    for (c in effectSet) {
      
      if (c == "treatment") {
        trtList <- c("LPS","Gard","Dex")
      } else {
        trtList <- c("LPS","Gard","Dex","Ctrl")
      }
      
      for (trt in trtList) {
        shortName <- paste0(sex,"_",c,"_",trt,"Trt_",thresh)
        
        if (trt %in% c("LPS","Ctrl")) {
          modelA <- "modelF"
          modelApreg <- "modelR"
          modelAstat <- "modelP"
          } else {
            modelA <- "modelG"
            modelApreg <- "modelS"
            modelAstat <- "modelQ"
          }
        
        if (sex == "F") {
          if (c == "status") {
            emmremlResults <- read.table(paste0(outDir,"allBatches_",sex,"_",trt,"only_10000genes_",thresh,"lCPM_",modelA,"_basic+Preg+WeightwKmatrix_emmreml"))
            stdBetas <- emmremlResults$beta_status / sqrt(emmremlResults$var_beta_status)
          } else if (c == "preg") {
            emmremlResults <- read.table(paste0(outDir,"allBatches_",sex,"_",trt,"only_10000genes_",thresh,"lCPM_",modelA,"_basic+Preg+WeightwKmatrix_emmreml"))
            stdBetas <- emmremlResults$beta_preg / sqrt(emmremlResults$var_beta_preg)
          } else if (c == "treatment") {
            emmremlResults <- read.table(paste0(outDir,"allBatches_",sex,"_",trt,"_10000genes_",thresh,"lCPM_",modelA,"_basic+Preg+WeightwKmatrix_emmreml"))
            stdBetas <- emmremlResults$beta_treatment / sqrt(emmremlResults$var_beta_treatment)
          }
        } else {
          if (c == "status") {
            emmremlResults <- read.table(paste0(outDir,"allBatches_",sex,"_",trt,"only_10000genes_",thresh,"lCPM_",modelA,"_basic+WeightwKmatrix_emmreml"))
            stdBetas <- emmremlResults$beta_status / sqrt(emmremlResults$var_beta_status)
          } else if (c == "preg") {
            emmremlResults <- read.table(paste0(outDir,"allBatches_",sex,"_",trt,"only_10000genes_",thresh,"lCPM_",modelA,"_basic+WeightwKmatrix_emmreml"))
            stdBetas <- emmremlResults$beta_preg / sqrt(emmremlResults$var_beta_preg)
          } else if (c == "treatment") {
            emmremlResults <- read.table(paste0(outDir,"allBatches_",sex,"_",trt,"_10000genes_",thresh,"lCPM_",modelA,"_basic+WeightwKmatrix_emmreml"))
            stdBetas <- emmremlResults$beta_treatment / sqrt(emmremlResults$var_beta_treatment)
          }
        }
        
        modelResults <- as.data.frame(stdBetas)
        rownames(modelResults) <- rownames(emmremlResults)
        colnames(modelResults) <- "stdBeta"
        
        ### using modelResults run a GSEA, and save the output
        
        ### Permutation-based GSEA tests
        ### Originally developed by Jordan Anderson and Noah Snyder-Mackler. Modified by Tauras Vilgalys. 
        ### Last updated 2020 September 16
        
        ###outside R, build a csv file with geneName,hallmarkPathway
        
        # Parameter options on lines 11 and 12
        # Data and files to load on lines 16 and 30
        ## parameters to consider: 
        pvals <- F ## are data in pvalues or betas? If betas, set pvals to FALSE
        weight <- 1 ## Weight to place on betas. Typically either a 1 or a 0, where 0 weights all effects equally.
                ## weight=0 acts like a KS test, and is typically used with pvalues. 
        
        sampleBetas <- as.data.frame(rownames(modelResults))
        sampleBetas$Betas <- modelResults$stdBeta
        
        colnames(sampleBetas) <- c("Gene","Betas")
    
        gene.list <- sampleBetas$Gene
      
        ## get nperm permutations of the vector of Betas
        nperm <- 1000
        for (i in 1:nperm) {
          tmpB <- sample(sampleBetas$Beta)
          cbind(sampleBetas,tmpB) -> sampleBetas; rm(tmpB)
        }; rm(i)#; head(sampleBetas)
    
        ## Get the list of gene sets to test. here the file starts with gene names and the associated GO terms. 
        ## i downloaded these hallmark gene sets from https://www.gsea-msigdb.org/gsea/msigdb/collections.jsp
    
        gene.set <- read.table(paste0(dataDir,"hallmarkPathways_wCategories.csv"), sep = ",")
        
        colnames(gene.set) <- c("Associated.Gene.Name","GO.Term.Accession","Category")
    
        gene.set$gene_GO <- paste0(gene.set$Associated.Gene.Name,"_",gene.set$GO.Term.Accession)
    
        # Trim for expressed genes where B was estimated
        gene.set2 <- subset(gene.set, gene.set$Associated.Gene.Name %in% sampleBetas$Gene)
    
        # Remove duplicated gene/GO pairings
        gene.set2 <- gene.set2[!duplicated(gene.set2$gene_GO),]
    
        ## Get list of gene ontologies to test, and the number of genes in each
        GO_res <- as.data.frame(matrix(ncol=2, nrow=length(unique(gene.set2$GO.Term.Accession)), unique(gene.set2$GO.Term.Accession)))
        GO_res$V2 <- 0
        GO_res <- GO_res[!(GO_res$V1 == ""),]; colnames(GO_res) <- c("GO", "n_genes")
        
        for (i in 1:nrow(GO_res)) {
          GO_res$n_genes[i] <- sum(gene.set2$GO.Term.Accession == GO_res$GO[i])
        }
        
        rm(i)
        GO_res$n_genes <- as.numeric(GO_res$n_genes)
        
        ## Only consider gene sets with at least 5 measured genes
        GO_res <- subset(GO_res, n_genes >= 5)
        
        ## Create a matrix of just the effect sizes, moving the gene names to the row.names
        effect_size_matrix <- sampleBetas[,-1]
        row.names(effect_size_matrix) <- sampleBetas[,1]
    
        ## Run GSEA function for each GO term (defined earlier in GO_res$GO)
        output<-GSEA(effect_size_matrix = effect_size_matrix,
                     gene_set_list = unique(GO_res$GO),
                     gene_GO_associations = gene.set2,
                     weight = 1, pvals=F)
        
        names(output) <- c("NES_matrix_upreg","NES_matrix_downreg","p_values")
      
        # Normalize + ES' by the mean of + ES values, and - ES' by the mean of - ES null values for that gene set. 
        ##This should effectively control for differences in gene set size
        pos.ES.norm<-matrix(0,nrow=nrow(GO_res),ncol=nperm+1)
        neg.ES.norm<-matrix(0,nrow=nrow(GO_res),ncol=nperm+1)
        
        for (i in 1:length(GO_res$GO)) {
          pos.ES <- unlist(output$NES_matrix_upreg[i,])
          names(pos.ES)<-NULL
          neg.ES <- unlist(output$NES_matrix_downreg[i,])
          names(neg.ES)<-NULL
    
          #for a given gene set, N, calculate the mean of the positive max ES for estimated+ permuted ES.
          #calculate the mean of the max negative ES for estimate+ permuted ES also
           pos.m <- mean(pos.ES)
           neg.m <- mean(abs(neg.ES))
           
           #Then for each permutation in that geneset, scale ES by this mean value, depending on if it was max + or max - ES
           for (j in 1:(nperm+1)){
             pos.ES.norm[i,j]<-pos.ES[j]/pos.m
             neg.ES.norm[i,j]<-neg.ES[j]/neg.m
           }
        }
        
        # FDR? 
        ## Calculate p-values for one of the null distributions as well
        perm_upreg <- apply(output$NES_matrix_upreg[,-1],1,function(x) {ee <- ecdf(x[2:length(x)]); 1 - ee(x[1])})
        perm_downreg <- apply(output$NES_matrix_downreg[,-1],1,function(x) {ee <- ecdf(x[2:length(x)]); 1 - ee(x[1])})
        ## Use perm.fdr function (written by Joaquin Sanz, available from Snyder-Mackler et al. 2016: 
        ## https://github.com/nsmackler/status_genome_2016/blob/master/perm_FDR.R)
        
        fdr_up <- perm.fdr(output$p_values, perm_upreg, "pval_upreg", "fdr_vsPerm1")
        fdr_down <- perm.fdr(output$p_values, perm_upreg, "pval_downreg", "fdr_vsPerm1")
        
        upregNES <- output$NES_matrix_upreg[,1]
        dnregNES <- output$NES_matrix_downreg[,1]
        
        cbind(as.character(GO_res$GO),output$p_values)
    
        #plot significant effects
        betasOnly <- as.data.frame(effect_size_matrix$Betas)
        rownames(betasOnly) <- rownames(effect_size_matrix)
        colnames(betasOnly) <- c("betas")
        betasOnly$rank <- rank(-betasOnly$betas)
      
        dir.create(paste0(figDir,shortName))
        
        for (r in 1:2) {
          if(sum(output$p_values[,r] < .05) > 0) {
            for (g in which(output$p_values[,r] < .05)) {
              pathwayName <- as.character(GO_res$GO)[g]
              pathwayGenes <- subset(gene.set2, GO.Term.Accession == pathwayName)$Associated.Gene.Name
              betasOnly$pathway <- FALSE
              betasOnly$pathway <- rownames(effect_size_matrix) %in% pathwayGenes
              
              betasPathway <- subset(betasOnly, pathway)
              ggplot() +
                geom_linerange(aes(x = betasOnly$rank, ymax = betasOnly$betas, ymin = 0), alpha = 66/dim(effect_size_matrix)[1]) +
                geom_linerange(aes(x = betasPathway$rank, ymax = betasPathway$betas, ymin = 0)) +
                xlab("Rank") + ggtitle(pathwayName, 
                                       subtitle = paste0(subset(GO_res, GO == pathwayName)$n_genes," genes, ",
                                                         "p = ",as.numeric(output$p_values[g,r]),"    ",colnames(output$p_values)[r],
                                                         "      ",m,"-",shortName))
              ggsave(paste0(figDir,"/",shortName,"/gsea_stickPlot_",m,"_Enriched_",pathwayName,".png"), width = 6, height = 4)
            }
          }
        }
        
        gseaTreatment <- output
        
        sigList <- gseaTreatment$p_values[,1] < .05 | gseaTreatment$p_values[,2] < .05
        
        sigPathways <- as.data.frame(cbind(as.character(GO_res$GO),
                                           gseaTreatment$p_values,
                                           gseaTreatment$NES_matrix_upreg[,1],
                                           gseaTreatment$NES_matrix_downreg[,1],
                                           shortName,
                                           m,
                                           c,
                                           trt)[sigList,])
        
        allPathways <- as.data.frame(cbind(as.character(GO_res$GO),
                                           gseaTreatment$p_values,gseaTreatment$NES_matrix_upreg[,1],
                                           gseaTreatment$NES_matrix_downreg[,1],
                                           shortName,
                                           m,
                                           c,
                                           trt))
      
        if( dim(allSigTrtPathways)[1] == 0) {
          allSigTrtPathways <- sigPathways
          allTrtPathways <- allPathways
        } else {
            allSigTrtPathways <- rbind(allSigTrtPathways,sigPathways)
            allTrtPathways <- rbind(allTrtPathways,allPathways)
        }
      }
    }
    
    write.table(x = allSigTrtPathways, file = paste0(outDir,"gsea_",sex,"_trtstatuspreg_GSEA_sigResults.txt"), quote = F, row.names = F, sep = ",")
    write.table(x = allTrtPathways, file = paste0(outDir,"gsea_",sex,"_trtstatuspreg_GSEA_allResults.txt"), quote = F, row.names = F, sep = ",")
    
    
    geneSetSort <- unique(gene.set[,2:3])
    geneSetSort <- geneSetSort[order(geneSetSort$Category),]
    geneSetSort$pathwayShort <- str_remove(geneSetSort$GO.Term.Accession, "HALLMARK_")
    
    plotAllSigTrtPathways <- allSigTrtPathways
    
    colnames(plotAllSigTrtPathways)[c(1,4,5)] <- c("pathway","ESup","ESdown") 
    colnames(plotAllSigTrtPathways)[c(6,7)] <- c("species","model") 
    
    plotAllSigTrtPathways$sigPVal <- ifelse(plotAllSigTrtPathways$pval_upreg < plotAllSigTrtPathways$pval_downreg,
                                            plotAllSigTrtPathways$pval_upreg, plotAllSigTrtPathways$pval_downreg)
    
    plotAllSigTrtPathways$sigES <- ifelse(plotAllSigTrtPathways$pval_upreg < plotAllSigTrtPathways$pval_downreg,
                                            plotAllSigTrtPathways$ESup, plotAllSigTrtPathways$ESdown)
    
    plotAllSigTrtPathways$sigPValAdj <- ifelse(plotAllSigTrtPathways$sigPVal == 0, .0001, plotAllSigTrtPathways$sigPVal)
    
    plotAllSigTrtPathways$sigES <- as.numeric(plotAllSigTrtPathways$sigES)
    plotAllSigTrtPathways$sigPValAdj <- as.numeric(plotAllSigTrtPathways$sigPValAdj)
    
    
    ggplot(data = plotAllSigTrtPathways, aes(x = species, y = pathway, col = sigES, size = -log10(sigPValAdj))) +
      geom_point()
    
    
    #todo - remove "hallmark_"
    plotAllSigTrtPathways$pathway <- str_remove(plotAllSigTrtPathways$pathway, "HALLMARK_")
    
    
    #change angle of names (bottom)
    
    ggplot(data = plotAllSigTrtPathways, aes(x = species, y = pathway, col = sigES, size = -log10(sigPValAdj))) +
      geom_point() +
      theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
    
    
    #viridis color scheme
    
    #sort by number of conditions & up/down reg elo
    pathwayList <- unique(plotAllSigTrtPathways$pathway)
    meanES <- vector()
    
    for (p in pathwayList) {
      meanES <- c(meanES,mean(subset(plotAllSigTrtPathways, pathway == p)$sigES))
    }
    
    pathwayList <- pathwayList[order(-meanES)]
    
    plotAllSigTrtPathways$pathway <- factor(plotAllSigTrtPathways$pathway, levels = pathwayList)
    
    
    
    pathwayList <- unique(plotAllSigTrtPathways$pathway)
    meanES <- vector()
    
    for (p in pathwayList) {
      meanES <- c(meanES,mean(subset(plotAllSigTrtPathways, pathway == p)$sigES))
    }
    
    pathwayList <- pathwayList[order(-meanES)]
    
    plotAllSigTrtPathways$pathway <- factor(plotAllSigTrtPathways$pathway, levels = pathwayList)
    
    
    ggplot(data = subset(plotAllSigTrtPathways, c == "status"), aes(x = species, y = pathway, col = sigES, size = -log10(sigPValAdj))) +
      geom_point() + scale_color_viridis(limits = c(-1,1)) + ggtitle("status - emreml - sig GSEA pathways") +
      theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
    ggsave(paste0(figDir,"GSEA_condition_status_plot_viridis_emreml_",thresh,"_",sex,".png"), width = 8, height = 6)
    
    ggplot(data = subset(plotAllSigTrtPathways, c == "preg"), aes(x = species, y = pathway, col = sigES, size = -log10(sigPValAdj))) +
      geom_point() + scale_color_viridis(limits = c(-1,1)) + ggtitle("preg - emreml - sig GSEA pathways") +
      theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
    ggsave(paste0(figDir,"GSEA_condition_preg_plot_viridis_emreml_",thresh,"_",sex,".png"), width = 8, height = 6)
    
    
    immunePathways <- c("TNFA_SIGNALING_VIA_NFKB",
                        "INFLAMMATORY_RESPONSE",
                        "INTERFERON_ALPHA_RESPONSE",
                        "INTERFERON_GAMMA_RESPONSE",
                        "COMPLEMENT",
                        "CTRA_UP",
                        "CTRA_DOWN")
    
    ggplot(data = subset(plotAllSigTrtPathways, pathway %in% immunePathways & c == "status"), 
           aes(x = species, y = pathway, col = sigES, size = -log10(sigPValAdj))) +
      geom_point() + scale_color_viridis(limits = c(-1,1)) + ggtitle("status - emreml - sig GSEA pathways") +
      theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
    ggsave(paste0(figDir,"GSEA_condition_status_plot_viridis_emreml_ImmunePaths_",thresh,"_",sex,".png"), width = 8, height = 6)
    
    ggplot(data = subset(plotAllSigTrtPathways, pathway %in% immunePathways & c == "preg"), 
           aes(x = species, y = pathway, col = sigES, size = -log10(sigPValAdj))) +
      geom_point() + scale_color_viridis(limits = c(-1,1)) + ggtitle("preg - emreml - sig GSEA pathways") +
      theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
    ggsave(paste0(figDir,"GSEA_condition_preg_plot_viridis_emreml_ImmunePaths_",thresh,"_",sex,".png"), width = 8, height = 6)
    
    ggplot(data = subset(plotAllSigTrtPathways, pathway %in% immunePathways), 
           aes(x = species, y = pathway, col = sigES, size = -log10(sigPValAdj))) +
      geom_point() + scale_color_viridis(limits = c(-1,1)) + ggtitle("preg - emreml - sig GSEA pathways") +
      theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
    ggsave(paste0(figDir,"GSEA_condition_plot_viridis_emreml_ImmunePaths_",thresh,"_",sex,".png"), width = 8, height = 4)
    
    
    plotAllSigTrtPathways$pathway <- factor(plotAllSigTrtPathways$pathway, levels = geneSetSort$pathwayShort)
    
    ## plot the pathways with multiple sig hits
    sigPaths <- names(table(plotAllSigTrtPathways$pathway)[table(plotAllSigTrtPathways$pathway) > 2])
    
    ggplot(data = subset(plotAllSigTrtPathways, pathway %in% sigPaths), aes(x = species, y = pathway, col = sigES, size = -log10(sigPValAdj))) +
      geom_point() + scale_color_viridis(limits = c(-1,1)) + ggtitle("emreml - sig GSEA in > 2 pathways") +
      theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
    
    ggsave(paste0(figDir,"GSEA_condition_plot_viridis_emreml_SigPaths_gt2_",thresh,"_",sex,".png"), width = 8, height = 6)
    
    p <- "MYC_TARGETS_V1"
    t <- "LPS"
    
    for (t in unique(plotAllSigTrtPathways$trt)) {
      print(t)
      pathSig <- vector()
      
      for (p in geneSetSort$pathwayShort) {
        cSig <- subset(plotAllSigTrtPathways, trt == t & pathway == p)$c
        
        if (length(cSig) == 2) {
          pathSig <- c(pathSig,"both")
        } else if (length(cSig) == 1) {
          if (cSig == "preg") {
            pathSig <- c(pathSig,"preg")
          } else {
            pathSig <- c(pathSig,"status")
          }
        } else {
          pathSig <- c(pathSig,"none")
        }
      }
      gseaFETMat <- matrix(c(sum(pathSig == "both"),
                             sum(pathSig == "preg"),
                             sum(pathSig == "status"),
                             sum(pathSig == "none")), nrow = 2)
    
      #fishP <- signif(fisher.test(fishDat)$p.value, digits = 3)
      #sigGeneDiff <- signif(fishDat - chisq.test(fishDat)$expected, digits = 3)
      #excessBoth <- sigGeneDiff[1,1]
      fishP <- signif(fisher.test(gseaFETMat)$p.value, digits = 3)
      sigGeneDiff <- signif(gseaFETMat - chisq.test(gseaFETMat)$expected, digits = 3)
      excessBoth <- sigGeneDiff[1,1]
      print(paste0(fishP," ",excessBoth))
    }
  }
}
```


Treatment GSEA M+F (Fig 1/paper)
```{r GSEA}
#emreml results

## loop through data sets

trt <- "LPS"
m <- "emmreml"

sex <- "MF"
allSigTrtPathways <- data.frame()
allTrtPathways <- data.frame()
effectSet <- c("treatment")
c <- "treatment"
trtList <- c("LPS","Gard","Dex")
thresh <- 5
  
for (trt in trtList) {
    shortName <- paste0(sex,"_",c,"_",trt,"Trt_",thresh)
    
    if (trt %in% c("LPS","Ctrl")) {
      model <- "modelF"
      } else {
        model <- "modelG"
      }
    
    emmremlResults <- read.table(paste0(outDir,"allBatches_",sex,"_",trt,"_10000genes_",thresh,"lCPM_",model,"_basicMF+WeightwKmatrix_emmreml"))
    stdBetas <- emmremlResults$beta_treatment / sqrt(emmremlResults$var_beta_treatment)
    
    modelResults <- as.data.frame(stdBetas)
    rownames(modelResults) <- rownames(emmremlResults)
    colnames(modelResults) <- "stdBeta"
    
    ### using modelResults run a GSEA, and save the output
    
    ### Permutation-based GSEA tests
    ### Originally developed by Jordan Anderson and Noah Snyder-Mackler. Modified by Tauras Vilgalys. 
    ### Last updated 2020 September 16
    
    ###outside R, build a csv file with geneName,hallmarkPathway
    
    # Parameter options on lines 11 and 12
    # Data and files to load on lines 16 and 30
    ## parameters to consider: 
    pvals <- F ## are data in pvalues or betas? If betas, set pvals to FALSE
    weight <- 1 ## Weight to place on betas. Typically either a 1 or a 0, where 0 weights all effects equally.
            ## weight=0 acts like a KS test, and is typically used with pvalues. 
    
    sampleBetas <- as.data.frame(rownames(modelResults))
    sampleBetas$Betas <- modelResults$stdBeta
    
    colnames(sampleBetas) <- c("Gene","Betas")

    gene.list <- sampleBetas$Gene
  
    ## get nperm permutations of the vector of Betas
    nperm <- 60000
    for (i in 1:nperm) {
      tmpB <- sample(sampleBetas$Beta)
      cbind(sampleBetas,tmpB) -> sampleBetas; rm(tmpB)
    }; rm(i)#; head(sampleBetas)

    ## Get the list of gene sets to test. here the file starts with gene names and the associated GO terms. 
    ## i downloaded these hallmark gene sets from https://www.gsea-msigdb.org/gsea/msigdb/collections.jsp

    gene.set <- read.table(paste0(dataDir,"hallmarkPathways_wCategories.csv"), sep = ",")
    
    colnames(gene.set) <- c("Associated.Gene.Name","GO.Term.Accession","Category")

    gene.set$gene_GO <- paste0(gene.set$Associated.Gene.Name,"_",gene.set$GO.Term.Accession)

    # Trim for expressed genes where B was estimated
    gene.set2 <- subset(gene.set, gene.set$Associated.Gene.Name %in% sampleBetas$Gene)

    # Remove duplicated gene/GO pairings
    gene.set2 <- gene.set2[!duplicated(gene.set2$gene_GO),]

    ## Get list of gene ontologies to test, and the number of genes in each
    GO_res <- as.data.frame(matrix(ncol=2, nrow=length(unique(gene.set2$GO.Term.Accession)), unique(gene.set2$GO.Term.Accession)))
    GO_res$V2 <- 0
    GO_res <- GO_res[!(GO_res$V1 == ""),]; colnames(GO_res) <- c("GO", "n_genes")
    
    for (i in 1:nrow(GO_res)) {
      GO_res$n_genes[i] <- sum(gene.set2$GO.Term.Accession == GO_res$GO[i])
    }
    
    rm(i)
    GO_res$n_genes <- as.numeric(GO_res$n_genes)
    
    ## Only consider gene sets with at least 5 measured genes
    GO_res <- subset(GO_res, n_genes >= 5)
    
    ## Create a matrix of just the effect sizes, moving the gene names to the row.names
    effect_size_matrix <- sampleBetas[,-1]
    row.names(effect_size_matrix) <- sampleBetas[,1]

    ## Run GSEA function for each GO term (defined earlier in GO_res$GO)
    output<-GSEA(effect_size_matrix = effect_size_matrix,
                 gene_set_list = unique(GO_res$GO),
                 gene_GO_associations = gene.set2,
                 weight = 1, pvals=F)
    
    names(output) <- c("NES_matrix_upreg","NES_matrix_downreg","p_values")
  
    # Normalize + ES' by the mean of + ES values, and - ES' by the mean of - ES null values for that gene set. 
    ##This should effectively control for differences in gene set size
    pos.ES.norm<-matrix(0,nrow=nrow(GO_res),ncol=nperm+1)
    neg.ES.norm<-matrix(0,nrow=nrow(GO_res),ncol=nperm+1)
    
    for (i in 1:length(GO_res$GO)) {
      pos.ES <- unlist(output$NES_matrix_upreg[i,])
      names(pos.ES)<-NULL
      neg.ES <- unlist(output$NES_matrix_downreg[i,])
      names(neg.ES)<-NULL

      #for a given gene set, N, calculate the mean of the positive max ES for estimated+ permuted ES.
      #calculate the mean of the max negative ES for estimate+ permuted ES also
       pos.m <- mean(pos.ES)
       neg.m <- mean(abs(neg.ES))
       
       #Then for each permutation in that geneset, scale ES by this mean value, depending on if it was max + or max - ES
       for (j in 1:(nperm+1)){
         pos.ES.norm[i,j]<-pos.ES[j]/pos.m
         neg.ES.norm[i,j]<-neg.ES[j]/neg.m
       }
    }
    
    # FDR? 
    ## Calculate p-values for one of the null distributions as well
    perm_upreg <- apply(output$NES_matrix_upreg[,-1],1,function(x) {ee <- ecdf(x[2:length(x)]); 1 - ee(x[1])})
    perm_downreg <- apply(output$NES_matrix_downreg[,-1],1,function(x) {ee <- ecdf(x[2:length(x)]); 1 - ee(x[1])})
    ## Use perm.fdr function (written by Joaquin Sanz, available from Snyder-Mackler et al. 2016: 
    ## https://github.com/nsmackler/status_genome_2016/blob/master/perm_FDR.R)
    
    fdr_up <- perm.fdr(output$p_values, perm_upreg, "pval_upreg", "fdr_vsPerm1")
    fdr_down <- perm.fdr(output$p_values, perm_upreg, "pval_downreg", "fdr_vsPerm1")
    
    upregNES <- output$NES_matrix_upreg[,1]
    dnregNES <- output$NES_matrix_downreg[,1]
    
    cbind(as.character(GO_res$GO),output$p_values)

    #plot significant effects
    betasOnly <- as.data.frame(effect_size_matrix$Betas)
    rownames(betasOnly) <- rownames(effect_size_matrix)
    colnames(betasOnly) <- c("betas")
    betasOnly$rank <- rank(-betasOnly$betas)
  
    dir.create(paste0(figDir,shortName))
    
    for (r in 1:2) {
      if(sum(output$p_values[,r] < .05) > 0) {
        for (g in which(output$p_values[,r] < .05)) {
          pathwayName <- as.character(GO_res$GO)[g]
          pathwayGenes <- subset(gene.set2, GO.Term.Accession == pathwayName)$Associated.Gene.Name
          betasOnly$pathway <- FALSE
          betasOnly$pathway <- rownames(effect_size_matrix) %in% pathwayGenes
          
          betasPathway <- subset(betasOnly, pathway)
          ggplot() +
            geom_linerange(aes(x = betasOnly$rank, ymax = betasOnly$betas, ymin = 0), alpha = 66/dim(effect_size_matrix)[1]) +
            geom_linerange(aes(x = betasPathway$rank, ymax = betasPathway$betas, ymin = 0)) +
            xlab("Rank") + ggtitle(pathwayName, 
                                   subtitle = paste0(subset(GO_res, GO == pathwayName)$n_genes," genes, ",
                                                     "p = ",as.numeric(output$p_values[g,r]),"    ",colnames(output$p_values)[r],
                                                     "      ",m,"-",shortName))
          ggsave(paste0(figDir,"/",shortName,"/gsea_stickPlot_",m,"_Enriched_",pathwayName,".png"), width = 6, height = 4)
        }
      }
    }
    
    gseaTreatment <- output
    
    nPaths <- dim(gseaTreatment$p_values)[1]
    
    sigList <- gseaTreatment$p_values[,1] < .05 | gseaTreatment$p_values[,2] < .05
    
    sigPathways <- as.data.frame(cbind(as.character(GO_res$GO),
                                       gseaTreatment$p_values,
                                       gseaTreatment$NES_matrix_upreg[,1],
                                       gseaTreatment$NES_matrix_downreg[,1],
                                       gseaTreatment$p_values * nPaths,
                                       shortName,
                                       m,
                                       c,
                                       trt)[sigList,])
    
    allPathways <- as.data.frame(cbind(as.character(GO_res$GO),
                                       gseaTreatment$p_values,
                                       gseaTreatment$NES_matrix_upreg[,1],
                                       gseaTreatment$NES_matrix_downreg[,1],
                                       gseaTreatment$p_values * nPaths,
                                       shortName,
                                       m,
                                       c,
                                       trt))
  
    if( dim(allSigTrtPathways)[1] == 0) {
      allSigTrtPathways <- sigPathways
      allTrtPathways <- allPathways
    } else {
        allSigTrtPathways <- rbind(allSigTrtPathways,sigPathways)
        allTrtPathways <- rbind(allTrtPathways,allPathways)
    }
}

colnames(allSigTrtPathways)[6:7] <- c("adjPval_upreg","adjPval_downreg")
colnames(allTrtPathways)[6:7] <- c("adjPval_upreg","adjPval_downreg")

for (c in 2:7) {
  allTrtPathways[,c] <- as.numeric(allTrtPathways[,c])
  allSigTrtPathways[,c] <- as.numeric(allSigTrtPathways[,c])
}

write.table(x = allSigTrtPathways, file = paste0(outDir,"gsea_",sex,"_trtstatuspreg_GSEA_sigResults.txt"), quote = F, row.names = F, sep = ",")
write.table(x = allTrtPathways, file = paste0(outDir,"gsea_",sex,"_trtstatuspreg_GSEA_allResults.txt"), quote = F, row.names = F, sep = ",")


geneSetSort <- unique(gene.set[,2:3])
geneSetSort <- geneSetSort[order(geneSetSort$Category),]
geneSetSort$pathwayShort <- str_remove(geneSetSort$GO.Term.Accession, "HALLMARK_")

plotAllSigTrtPathways <- allSigTrtPathways

colnames(plotAllSigTrtPathways)[c(1,4,5)] <- c("pathway","ESup","ESdown")
colnames(plotAllSigTrtPathways)[c(8,9)] <- c("species","model") 

plotAllSigTrtPathways$sigPVal <- ifelse(plotAllSigTrtPathways$pval_upreg < plotAllSigTrtPathways$pval_downreg,
                                        plotAllSigTrtPathways$pval_upreg, plotAllSigTrtPathways$pval_downreg)

plotAllSigTrtPathways$sigES <- ifelse(plotAllSigTrtPathways$pval_upreg < plotAllSigTrtPathways$pval_downreg,
                                        plotAllSigTrtPathways$ESup, plotAllSigTrtPathways$ESdown)

plotAllSigTrtPathways$sigPValAdj <- ifelse(plotAllSigTrtPathways$sigPVal == 0, .0001, plotAllSigTrtPathways$sigPVal)

plotAllSigTrtPathways$sigES <- as.numeric(plotAllSigTrtPathways$sigES)
plotAllSigTrtPathways$sigPValAdj <- as.numeric(plotAllSigTrtPathways$sigPValAdj)


ggplot(data = plotAllSigTrtPathways, aes(x = species, y = pathway, col = sigES, size = -log10(sigPValAdj))) +
  geom_point()


#todo - remove "hallmark_"
plotAllSigTrtPathways$pathway <- str_remove(plotAllSigTrtPathways$pathway, "HALLMARK_")


#change angle of names (bottom)

ggplot(data = plotAllSigTrtPathways, aes(x = species, y = pathway, col = sigES, size = -log10(sigPValAdj))) +
  geom_point() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))


#viridis color scheme

#sort by number of conditions & up/down reg elo
pathwayList <- unique(plotAllSigTrtPathways$pathway)
meanES <- vector()

for (p in pathwayList) {
  meanES <- c(meanES,mean(subset(plotAllSigTrtPathways, pathway == p)$sigES))
}

pathwayList <- pathwayList[order(-meanES)]

plotAllSigTrtPathways$pathway <- factor(plotAllSigTrtPathways$pathway, levels = pathwayList)



pathwayList <- unique(plotAllSigTrtPathways$pathway)
meanES <- vector()

for (p in pathwayList) {
  meanES <- c(meanES,mean(subset(plotAllSigTrtPathways, pathway == p)$sigES))
}

pathwayList <- pathwayList[order(-meanES)]

plotAllSigTrtPathways$pathway <- factor(plotAllSigTrtPathways$pathway, levels = pathwayList)


ggplot(data = subset(plotAllSigTrtPathways, c == "status"), aes(x = species, y = pathway, col = sigES, size = -log10(sigPValAdj))) +
  geom_point() + scale_color_viridis(limits = c(-1,1)) + ggtitle("status - emreml - sig GSEA pathways") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
ggsave(paste0(figDir,"GSEA_condition_status_plot_viridis_emreml_",thresh,"_",sex,".png"), width = 8, height = 6)

ggplot(data = subset(plotAllSigTrtPathways, c == "preg"), aes(x = species, y = pathway, col = sigES, size = -log10(sigPValAdj))) +
  geom_point() + scale_color_viridis(limits = c(-1,1)) + ggtitle("preg - emreml - sig GSEA pathways") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
ggsave(paste0(figDir,"GSEA_condition_preg_plot_viridis_emreml_",thresh,"_",sex,".png"), width = 8, height = 6)


immunePathways <- c("TNFA_SIGNALING_VIA_NFKB",
                    "INFLAMMATORY_RESPONSE",
                    "INTERFERON_ALPHA_RESPONSE",
                    "INTERFERON_GAMMA_RESPONSE",
                    "COMPLEMENT",
                    "CTRA_UP",
                    "CTRA_DOWN")

ggplot(data = subset(plotAllSigTrtPathways, pathway %in% immunePathways & c == "status"), 
       aes(x = species, y = pathway, col = sigES, size = -log10(sigPValAdj))) +
  geom_point() + scale_color_viridis(limits = c(-1,1)) + ggtitle("status - emreml - sig GSEA pathways") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
ggsave(paste0(figDir,"GSEA_condition_status_plot_viridis_emreml_ImmunePaths_",thresh,"_",sex,".png"), width = 8, height = 6)

ggplot(data = subset(plotAllSigTrtPathways, pathway %in% immunePathways & c == "preg"), 
       aes(x = species, y = pathway, col = sigES, size = -log10(sigPValAdj))) +
  geom_point() + scale_color_viridis(limits = c(-1,1)) + ggtitle("preg - emreml - sig GSEA pathways") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
ggsave(paste0(figDir,"GSEA_condition_preg_plot_viridis_emreml_ImmunePaths_",thresh,"_",sex,".png"), width = 8, height = 6)

ggplot(data = subset(plotAllSigTrtPathways, pathway %in% immunePathways), 
       aes(x = species, y = pathway, col = sigES, size = -log10(sigPValAdj))) +
  geom_point() + scale_color_viridis(limits = c(-1,1)) + ggtitle("preg - emreml - sig GSEA pathways") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
ggsave(paste0(figDir,"GSEA_condition_plot_viridis_emreml_ImmunePaths_",thresh,"_",sex,".png"), width = 8, height = 4)


plotAllSigTrtPathways$pathway <- factor(plotAllSigTrtPathways$pathway, levels = geneSetSort$pathwayShort)

## plot the pathways with multiple sig hits
sigPaths <- names(table(plotAllSigTrtPathways$pathway)[table(plotAllSigTrtPathways$pathway) > 2])

ggplot(data = subset(plotAllSigTrtPathways, pathway %in% sigPaths), aes(x = species, y = pathway, col = sigES, size = -log10(sigPValAdj))) +
  geom_point() + scale_color_viridis(limits = c(-1,1)) + ggtitle("emreml - sig GSEA in > 2 pathways") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))

ggsave(paste0(figDir,"GSEA_condition_plot_viridis_emreml_SigPaths_gt2_",thresh,"_",sex,".png"), width = 8, height = 6)



```

```{r TLR4 analysis}

# 2018 Lea
# 234/542 of the MyD88-dependent genes and 
# 165/400 of the TRIF-dependent genes identified by [2008 Ramsey] had expressed orthologs in our dataset

# 2016 NSM
# Ramsey et al reported 334 genes that were up-regulated upon stimulation via the MyD88-dependent pathway and 
# 274 by that were up-regulated via the TRIF-dependent pathway, 
# among which 238 and 168 rhesus macaque orthologues were included in our LPS challenge data set, respectively.

## 2008 Ramsey
# Groups include those primarily induced (C11, C12, C15, C17, C21, C26) and downregulated (C7, C29) by the MyD88-dependent pathway, 
# and those primarily induced (C6, C8, C22, C24) and downregulated (C4, C5, C10, C20) by the TRIF-dependent pathway.

# Table S4, "Cluster" column identified above

tlr4genes <- read.csv2(paste0(dataDir,"2008Ramsey_TLR4only_genelist.csv"), sep = ",")

table(tlr4genes$Path)
table(tlr4genes$Path.Dir)

# check genes in LPS Female condition
varA <- "LPS"
modelA <- "F"
emremlResults <- read.table(paste0(outDir,"allBatches_F_",varA,"only_10000genes_5lCPM_model",
                                   modelA,"_basic+Preg+WeightwKmatrix_emmreml"))

head(rownames(emremlResults))
tlr4genes$geneName <- toupper(tlr4genes$GeneSymbol)

tlr4genes$inMrktData <- ifelse(tlr4genes$geneName %in% rownames(emremlResults), TRUE, FALSE) 

table(subset(tlr4genes, inMrktData)$Path.Dir)

# percent kept
100 * (table(subset(tlr4genes, inMrktData)$Path.Dir) / table(tlr4genes$Path.Dir))

paste0(sum(tlr4genes$Path == "MyD88" & tlr4genes$inMrktData),"/",sum(tlr4genes$Path == "MyD88"),
       " of the MyD88-dependent genes and ",
       sum(tlr4genes$Path == "TRIF" & tlr4genes$inMrktData),"/",sum(tlr4genes$Path == "TRIF"),
       " of the TRIF-dependent genes identified by [2008 Ramsey] had expressed orthologs in our dataset")

tlr4mrkt <- subset(tlr4genes, inMrktData)

### get group gene lists
trtThresh <- 1
qvalThresh <- .1
trt <- "LPS"

fxVar <- "status"
#qvalThresh <- 0.2
if (trt == "LPS") {
  model <- "F"
} else {
model <- "G"
}

#get list of genes sig (qval < .1) in Ctrl only, LPS only
trtEmreml <- read.table(paste0("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_",trt,
                               "only_10000genes_5lCPM_model",model,"_basic+Preg+WeightwKmatrix_emmreml_wPermFDR_v1_status"))
ctrlEmreml <- read.table(paste0("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_Ctrl",
                               "only_10000genes_5lCPM_modelF_basic+Preg+WeightwKmatrix_emmreml_wPermFDR_v1_status"))
pvalCol <- which(colnames(trtEmreml) == paste0("pval_",fxVar))
fdrCol <- which(colnames(trtEmreml) == paste0("fdr_",fxVar))
betaCol <- which(colnames(trtEmreml) == paste0("beta_",fxVar))

#get LPS-Ctrl response betas
respEmreml <- read.table(paste0("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_",trt,
                                "_10000genes_5lCPM_model",model,"_basicResp+Preg+Weight_emmreml"))

respEmreml$stdbeta_status <- respEmreml$beta_status / sqrt(respEmreml$var_beta_status)

trtCtrlEmreml <- read.table(paste0("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_",trt,
                                   "_10000genes_5lCPM_model",model,"_basic+Preg+WeightwKmatrix_emmreml"))

### to respEmreml, need to add:
# treatment effect - is the gene up or down reg in LPS Treatment
trtUpGeneList <- rownames(trtCtrlEmreml)[trtCtrlEmreml$beta_treatment > 0]
trtDnGeneList <- rownames(trtCtrlEmreml)[trtCtrlEmreml$beta_treatment < 0]

# status effect at LPS - is the gene up or down reg w.r.t. status?
trtSigUpGeneList <- rownames(trtEmreml)[trtEmreml[,fdrCol] < qvalThresh & trtEmreml[,betaCol] > 0]
trtSigDnGeneList <- rownames(trtEmreml)[trtEmreml[,fdrCol] < qvalThresh & trtEmreml[,betaCol] < 0]

# status effect at Ctrl - is the gene up or down reg w.r.t. status?
ctrlSigUpGeneList <- rownames(ctrlEmreml)[ctrlEmreml[,fdrCol] < qvalThresh & ctrlEmreml[,betaCol] > 0]
ctrlSigDnGeneList <- rownames(ctrlEmreml)[ctrlEmreml[,fdrCol] < qvalThresh & ctrlEmreml[,betaCol] < 0]

# status effect overall, is the gene up or down reg w.r.t. status in both?

respEmreml$trtEffect <- ifelse(rownames(respEmreml) %in% trtUpGeneList, "Up",
                               ifelse(rownames(respEmreml) %in% trtDnGeneList, "Down", "none"))

respEmreml$statEffectCtrl <- ifelse(rownames(respEmreml) %in% ctrlSigUpGeneList, "Up",
                               ifelse(rownames(respEmreml) %in% ctrlSigDnGeneList, "Down", "none"))

respEmreml$statEffectTrt <- ifelse(rownames(respEmreml) %in% trtSigUpGeneList, "Up",
                               ifelse(rownames(respEmreml) %in% trtSigDnGeneList, "Down", "none"))

respEmreml$statEffectBoth <- ifelse(respEmreml$statEffectCtrl == "Up" | respEmreml$statEffectTrt == "Up", "Up",
                                    ifelse(respEmreml$statEffectCtrl == "Down" | respEmreml$statEffectTrt == "Down", "Down", "none"))


respEmreml$groupFig3 <- ifelse(respEmreml$trtEffect == "Up" & respEmreml$statEffectBoth == "Down", "GroupI",
                               ifelse(respEmreml$trtEffect == "Up" & respEmreml$statEffectBoth == "Up", "GroupII",
                                      ifelse(respEmreml$trtEffect == "Down" & respEmreml$statEffectBoth == "Down", "GroupIII",
                                             ifelse(respEmreml$trtEffect == "Down" & respEmreml$statEffectBoth == "Down", "GroupIV","none"))))
    
    
print(table(respEmreml$groupFig3))

groupIgeneList <- rownames(subset(respEmreml, groupFig3 == "GroupI"))
groupIIgeneList <- rownames(subset(respEmreml, groupFig3 == "GroupII"))
groupIIIgeneList <- rownames(subset(respEmreml, groupFig3 == "GroupIII"))

#fisher's exact test, enrichment among Group I/II/III/IV
fetPVal <- vector()
inPath <- vector()
pathTotal <- vector()
geneGrp <- vector()
oddsRatio <- vector()
testedPway <- vector()
pway <- "MyD88-up"
geneGroup <- "GroupIII"

for (geneGroup in names(table(respEmreml$groupFig3))[names(table(respEmreml$groupFig3)) != "none"]) {
  
  groupGeneList <- rownames(subset(respEmreml, groupFig3 == geneGroup))
  nonGroupGeneList <- rownames(subset(respEmreml, groupFig3 != geneGroup))
    
  for (pway in unique(tlr4mrkt$Path.Dir)) {
    
    pathwayGII <- length(intersect(groupGeneList,
                                   tlr4mrkt$geneName[tlr4mrkt$Path.Dir == pway]))
    if (pathwayGII >= 1) {
      nPathwayGII <- length(tlr4mrkt$geneName[tlr4mrkt$Path.Dir == pway]) - pathwayGII
      nPathwayBck <- length(nonGroupGeneList) - nPathwayGII
      pathwayBck <- length(groupGeneList) - pathwayGII
      
      dat <- data.frame(
        "Non-Pathway" = c(pathwayBck, nPathwayBck),
        "Pathway" = c(pathwayGII, nPathwayGII),
        row.names = c(geneGroup, "Background"),
        stringsAsFactors = FALSE
      )
      colnames(dat) <- c("Non-Pathway", "Pathway")
      
      ### only conduct test if there are 2 genes in the pathway, only conduct a one-sided test
      test <- fisher.test(dat, alternative = "greater")
      test
      print(paste0(pway," ",test$p.value))
      if (test$p.value < .1) {
        print(dat)
      }
      testedPway <- c(testedPway,pway)
      fetPVal <- c(fetPVal,test$p.value)
      oddsRatio <- c(oddsRatio,test$estimate)
      inPath <- c(inPath,pathwayGII)
      pathTotal <- c(pathTotal,nPathwayGII)
      geneGrp <- c(geneGrp,geneGroup)
    }
  }
}

fetResults <- cbind.data.frame(geneGrp,testedPway,fetPVal,oddsRatio,inPath,pathTotal)
colnames(fetResults)[2] <- "TLR4_gene_group"
#fetResults <- fetResults[order(fetResults$fetPVal),]

fetResults <- fetResults[order(fetResults$geneGrp, fetResults$fetPVal),]

fetResults

intersect(groupIIgeneList,tlr4mrkt$geneName[tlr4mrkt$Path.Dir == "MyD88-up"])
intersect(groupIgeneList,tlr4mrkt$geneName[tlr4mrkt$Path.Dir == "MyD88-up"])

##plot GE in Trif UP and MyD88 up in D and S
trtMeta <- readRDS(paste0("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_LPSonly_10000genes_5lCPM_metadata.RDS"))
trtResids <- readRDS(paste0("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_LPSonly_10000genes_5lCPM_modelF_resids.RDS"))
ctrlMeta <- readRDS(paste0("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_Ctrlonly_10000genes_5lCPM_metadata.RDS"))
ctrlResids <- readRDS(paste0("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_Ctrlonly_10000genes_5lCPM_modelF_resids.RDS"))

myD88upGeneList <- tlr4mrkt$geneName[tlr4mrkt$Path.Dir == "MyD88-up"]
myD88dnGeneList <- tlr4mrkt$geneName[tlr4mrkt$Path.Dir == "MyD88-down"]
trifupGeneList <- tlr4mrkt$geneName[tlr4mrkt$Path.Dir == "TRIF-up"]
trifdnGeneList <- tlr4mrkt$geneName[tlr4mrkt$Path.Dir == "TRIF-down"]

myD88GIIlist <- c("IL6","SERPINB2","EDN1")

trtMeta$myD88upMedian <- apply(trtResids[rownames(trtResids) %in% myD88upGeneList,],2,median)
#trtMeta$myD88upMedian <- colMeans(trtResids[rownames(trtResids) %in% myD88upGeneList,])
trtMeta$trifupMedian <- apply(trtResids[rownames(trtResids) %in% trifupGeneList,],2,median)
trtMeta$myD88upGIIMedian <- apply(trtResids[rownames(trtResids) %in% myD88GIIlist,],2,median)


ggplot(data = trtMeta,
       aes(x = status_dom_sub,
           y = myD88upMedian)) +
  geom_violin(draw_quantiles = c(.25,.5,.75)) +
  geom_point()

ggplot(data = trtMeta,
       aes(x = status_dom_sub,
           y = trifupMedian)) +
  geom_violin(draw_quantiles = c(.25,.5,.75)) +
  geom_point()

ggplot(data = trtMeta,
       aes(x = status_dom_sub,
           y = myD88upGIIMedian)) +
  geom_violin(draw_quantiles = c(.25,.5,.75)) +
  geom_point()


ctrlMeta$myD88upMedian <- apply(ctrlResids[rownames(ctrlResids) %in% myD88upGeneList,],2,median)
#trtMeta$myD88upMedian <- colMeans(trtResids[rownames(trtResids) %in% myD88upGeneList,])
ctrlMeta$trifupMedian <- apply(ctrlResids[rownames(ctrlResids) %in% trifupGeneList,],2,median)


ggplot(data = ctrlMeta,
       aes(x = status_dom_sub,
           y = myD88upMedian)) +
  geom_violin(draw_quantiles = c(.25,.5,.75)) +
  geom_point()

ggplot(data = ctrlMeta,
       aes(x = status_dom_sub,
           y = trifupMedian)) +
  geom_violin(draw_quantiles = c(.25,.5,.75)) +
  geom_point()

tlr4mrkt$lpsStatusBeta <- 0

for (r in 1:dim(tlr4mrkt)[1]) {
  tlr4mrkt$lpsStatusBeta[r] <- trtEmreml$beta_status[which(rownames(trtEmreml) %in% tlr4mrkt$geneName[r])]
}

ggplot(data = subset(tlr4mrkt, `Path.Dir` %in% c("MyD88-up","TRIF-up")),
       aes(x = `Path.Dir`,
           y = lpsStatusBeta)) +
  geom_violin(draw_quantiles = c(.25,.5,.75)) +
  geom_jitter(width = .03)



```


```{r cell type heterogeneity}
marrow <- read.csv2(paste0(dataDir,"/tabulaMuris/Marrow-counts.csv"), sep = ",")
dim(marrow)

rownames(marrow) <- marrow$X
marrow <- marrow[,-1]

marrowMeans <- rowMeans(marrow)

sum(marrowMeans == 0)

marrow <- marrow[marrowMeans > 0,]
dim(marrow)

marrowMeans <- rowMeans(marrow)

ggplot() + geom_histogram(aes(x = marrowMeans), bins = 100) + scale_x_log10()

sortMarrow <- marrow[order(marrowMeans, decreasing = T),]

murisGeneList <- toupper(rownames(sortMarrow)[1:5000])

write.table(murisGeneList, 
            file = "~/Dropbox (Personal)/meerkats/RNA/meerkatGE/data/topGenes.txt",
            row.names = FALSE, col.names = FALSE)

#how many are in all treatment
allTrtEmmreml <- read.table("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_allTrt_10000genes_5lCPM_modelG_allTrtF+Preg+WeightwKmatrix_emmreml")
allCounts <- readRDS("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_allTrt_10000genes_5lCPM_counts.RDS")
usedCounts <- readRDS("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_allTrt_10000genes_5lCPM_filteredCounts.RDS")

#how many of the top N-thousand were in the meerkat dataset?
for ( topNthou in c(100, 500, (1:10)*1000) ) {
  #topNthou <- 1000
  
  print(topNthou)
  
  murisGeneList <- toupper(rownames(sortMarrow)[1:topNthou])
  
  inMrkt <- sum(murisGeneList %in% rownames(allCounts))
  
  mrktInMuris <- sum(murisGeneList %in% rownames(allTrtEmmreml))
  
  percentIn <- sum(murisGeneList %in% rownames(allTrtEmmreml)) / sum(murisGeneList %in% rownames(allCounts))
  
  print(paste0(inMrkt," ",mrktInMuris," ",percentIn))
}

### FACS sorting
facsClasses <- read.csv2(paste0(dataDir,"tabulaMuris/facs_Marrow_cell_ontology_class_classes.csv"), sep = ",")
colnames(facsClasses) <- c("classNum","cellType")

facsMarkers <- read.csv2(paste0(dataDir,"tabulaMuris/facs_Marrow_cell_ontology_class_markers.csv"), sep = ",", dec = ".")
facsMarkers$geneUpper <- toupper(facsMarkers$gene)
dim(facsMarkers)

facsMarkers <- subset(facsMarkers, geneUpper %in% rownames(allCounts))
dim(facsMarkers)

facsMarkers <- subset(facsMarkers, geneUpper %in% rownames(usedCounts))
dim(facsMarkers)


### Droplet sorting
dropClasses <- read.csv2(paste0(dataDir,"tabulaMuris/droplet_Marrow_cell_ontology_class_classes.csv"), sep = ",")
colnames(dropClasses) <- c("classNum","cellType")

dropMarkers <- read.csv2(paste0(dataDir,"tabulaMuris/droplet_Marrow_cell_ontology_class_markers.csv"), sep = ",", dec = ".")


### how many FACS markers are in the meerkat data
dropMarkers$geneUpper <- toupper(dropMarkers$gene)
dim(dropMarkers)

dropMarkers <- subset(dropMarkers, geneUpper %in% rownames(allCounts))
dim(dropMarkers)

dropMarkers <- subset(dropMarkers, geneUpper %in% rownames(usedCounts))
dim(dropMarkers)
```


### Celltype Heterogeneity -- MOUSE

Check for enrichment of genes that are markers for specific cell types within the status-associated genes
If the effect of status is actually being caused by increases in T cells (for instance), then more of the status-associated genes than the expected 60% might also be T cell marker genes. If the distribution of marker genes is as expected, it could indicate that differences in cell composition arent responsible for the effects of status

```{r }
# is presence in muris enriched for status?
emmreml <- read.table("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_LPSonly_10000genes_5lCPM_modelF_basic+Preg+WeightwKmatrix_emmreml_wPermFDR_v1_status")
  
sigMuris <- sum(emmreml$fdr_status < .1 & rownames(emmreml) %in% toupper(rownames(marrow)))
nsMuris <- sum(emmreml$fdr_status > .1 & rownames(emmreml) %in% toupper(rownames(marrow)))

FET1 <- sigMuris
FET2 <- nsMuris
FET3 <- sum(emmreml$fdr_status < .1) - sigMuris
FET4 <- sum(emmreml$fdr_status > .1) - nsMuris
  
fetMat <- matrix(data = c(FET1,FET2,FET3,FET4), ncol = 2, nrow = 2, byrow = T)
  
fetResult <- fisher.test(fetMat)



#droplet sorted, mouse marrow
clusterList <- sort(unique(dropMarkers$cluster))

c <- clusterList[2]

for (c in clusterList) {
  emmreml <- read.table("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_LPSonly_10000genes_5lCPM_modelF_basic+Preg+WeightwKmatrix_emmreml_wPermFDR_v1_status")
  
  emmremlMuris <- emmreml[rownames(emmreml) %in% toupper(rownames(marrow)),]
  
  cMarkerGenes <- subset(dropMarkers, cluster == c)$geneUpper
  
  percentInSsur <- round(mean(cMarkerGenes %in% rownames(emmreml)), digits = 2)
  
  totSig <- sum(emmremlMuris$fdr_status < .1)
  totNS <- sum(emmremlMuris$fdr_status > .1)
  
  #only genes in the emmreml analysis
  cMarkerGenes <- cMarkerGenes[cMarkerGenes %in% rownames(emmreml)]
  
  #sigMarker <- sum(emmremlMuris$fdr_status < .1 & rownames(emmremlMuris) %in% cMarkerGenes)
  #nsMarker <- sum(emmremlMuris$fdr_status > .1 & rownames(emmremlMuris) %in% cMarkerGenes)
  sigMarker <- sum(emmremlMuris$fdr_status < .1 & rownames(emmremlMuris) %in% cMarkerGenes)
  nsMarker <- sum(emmremlMuris$fdr_status > .1 & rownames(emmremlMuris) %in% cMarkerGenes)
  
  FET1 <- sigMarker
  FET2 <- nsMarker
  FET3 <- totSig - sigMarker
  FET4 <- totNS - nsMarker
  
  fetMat <- matrix(data = c(FET1,FET2,FET3,FET4), ncol = 2, nrow = 2, byrow = T)
  
  fetResult <- fisher.test(fetMat)
  
  print(paste0(dropClasses[(c+1),2],";",
               c,";",
               round(fetResult$p.value, digits = 2),";",
               round(fetResult$p.value * length(clusterList), digits = 4),";",
               round(fetResult$estimate, digits = 2),";",
               percentInSsur))
}

dropClasses

#2,9,12
mrkrTCell <- subset(dropMarkers, cluster == 1)$geneUpper
mrkrBearly <- subset(dropMarkers, cluster == 8)$geneUpper
mrkrBlate <- subset(dropMarkers, cluster == 11)$geneUpper

mrkrTCell <- mrkrTCell[mrkrTCell %in% rownames(emmreml)]
mrkrBearly <- mrkrBearly[mrkrBearly %in% rownames(emmreml)]
mrkrBlate <- mrkrBlate[mrkrBlate %in% rownames(emmreml)]

length(mrkrTCell)
length(mrkrBearly)
length(mrkrBlate)

length(intersect(mrkrTCell,mrkrBearly))
length(intersect(mrkrBlate,mrkrBearly))
length(intersect(mrkrTCell,mrkrBlate))

length(intersect(mrkrTCell,intersect(mrkrBearly,mrkrBlate)))


cMarkerGenes
emmremlMuris$markerGene <- ifelse(rownames(emmremlMuris) %in% cMarkerGenes, TRUE, FALSE)

ggplot(subset(emmremlMuris, fdr_status < .1),
       aes(x = beta_status,
           fill = markerGene)) +
  geom_histogram(bins = 25)


#sig genes:
length(subset(emmremlMuris, fdr_status < .1 & beta_status > 0)$markerGene)
mean(subset(emmremlMuris, fdr_status < .1 & beta_status > 0)$markerGene)

length(subset(emmremlMuris, fdr_status < .1 & beta_status < 0)$markerGene)
mean(subset(emmremlMuris, fdr_status < .1 & beta_status < 0)$markerGene)

posBetaMarkers <- rownames(subset(emmremlMuris, fdr_status < .1 & beta_status > 0 & markerGene))
negBetaMarkers <- rownames(subset(emmremlMuris, fdr_status < .1 & beta_status < 0 & markerGene))

subset(dropMarkers, cluster == 1 & geneUpper %in% posBetaMarkers)
mean(subset(dropMarkers, cluster == 1 & geneUpper %in% negBetaMarkers)$avg_logFC > 1)
mean(subset(dropMarkers, cluster == 1 & geneUpper %in% negBetaMarkers)$avg_logFC < 1)

mean(subset(dropMarkers, cluster == 1)$avg_logFC > 1)
```

So that would involve markers that indicate a directionally consistent fraction (e.g., more T cells) being enriched in directionally consistent changes in gene expression linked to status, over and above their representation in the pool of genes tested as a whole in the gene expression analysis.

Is there enrichment of matching directionality between marker gene & sig gene compared to all significant marker genes
```{r}
c <- 4
# is presence in muris enriched for status?
emmreml <- read.table("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_LPSonly_10000genes_5lCPM_modelF_basic+Preg+WeightwKmatrix_emmreml_wPermFDR_v1_status")

sigEmmreml <- subset(emmreml, fdr_status < .1)
sigMurisEmmreml <- sigEmmreml[rownames(sigEmmreml) %in% toupper(rownames(marrow)),]
sigMurisEmmremlCluster <- sigMurisEmmreml[rownames(sigMurisEmmreml) %in% subset(dropMarkers, cluster == c)$geneUpper,]
sigMurisEmmremlCluster$geneUpper <- rownames(sigMurisEmmremlCluster)

sigMurisEmmremlClusterWMrkrData <- merge(sigMurisEmmremlCluster, subset(dropMarkers, cluster == c), by = "geneUpper")

#1 = + / Up
FET1 <- dim(subset(sigMurisEmmremlClusterWMrkrData, beta_status > 0 & avg_logFC > 1))[1]
#2 = - / Up
FET2 <- dim(subset(sigMurisEmmremlClusterWMrkrData, beta_status < 0 & avg_logFC > 1))[1]
#3 = + / Down
FET3 <- dim(subset(sigMurisEmmremlClusterWMrkrData, beta_status > 0 & avg_logFC < 1))[1]
#4 = - / Down
FET4 <- dim(subset(sigMurisEmmremlClusterWMrkrData, beta_status < 0 & avg_logFC < 1))[1]
  
fetMat <- matrix(data = c(FET1,FET2,FET3,FET4), ncol = 2, nrow = 2, byrow = T)
  
fetResult <- fisher.test(fetMat)



#droplet sorted, mouse marrow
clusterList <- sort(unique(dropMarkers$cluster))

c <- clusterList[2]

for (c in clusterList) {
  emmreml <- read.table("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_LPSonly_10000genes_5lCPM_modelF_basic+Preg+WeightwKmatrix_emmreml_wPermFDR_v1_status")
  
  sigEmmreml <- subset(emmreml, fdr_status < .1)
  sigMurisEmmreml <- sigEmmreml[rownames(sigEmmreml) %in% toupper(rownames(marrow)),]
  sigMurisEmmremlCluster <- sigMurisEmmreml[rownames(sigMurisEmmreml) %in% subset(dropMarkers, cluster == c)$geneUpper,]
  sigMurisEmmremlCluster$geneUpper <- rownames(sigMurisEmmremlCluster)
  
  sigMurisEmmremlClusterWMrkrData <- merge(sigMurisEmmremlCluster, subset(dropMarkers, cluster == c), by = "geneUpper")
  
  #1 = + / Up
  FET1 <- dim(subset(sigMurisEmmremlClusterWMrkrData, beta_status > 0 & avg_logFC > 1))[1]
  #2 = - / Up
  FET2 <- dim(subset(sigMurisEmmremlClusterWMrkrData, beta_status < 0 & avg_logFC > 1))[1]
  #3 = + / Down
  FET3 <- dim(subset(sigMurisEmmremlClusterWMrkrData, beta_status > 0 & avg_logFC < 1))[1]
  #4 = - / Down
  FET4 <- dim(subset(sigMurisEmmremlClusterWMrkrData, beta_status < 0 & avg_logFC < 1))[1]
    
  fetMat <- matrix(data = c(FET1,FET2,FET3,FET4), ncol = 2, nrow = 2, byrow = T)
    
  fetResult <- fisher.test(fetMat)
  
  print(paste0(dropClasses[(c+1),2],";",
               c,";",
               dim(sigMurisEmmremlClusterWMrkrData)[1],";",
               round(fetResult$p.value, digits = 2),";",
               round(fetResult$p.value * length(clusterList), digits = 4),";",
               round(fetResult$estimate, digits = 2),";",
               percentInSsur))
  
}
```


Use the marker list to estimate the cell type proportions in each sample
If there arent sample-to-sample differences in cell composition, then cell composition likely isnt causing the effects were measuring in status. Main issue here is that we have no ground truth (a flow-sorted meerkat sample with RNAseq), so it will be hard to infer what a given difference in marker genes means for differences in cell composition, or what a good threshold is for a sample-to-sample difference

CIBERSORT

```{r cibersort fxn}
## load CIBERSORT functions:


#Core algorithm
CoreAlg <- function(X, y){
  
  #try different values of nu
  svn_itor <- 3
  
  res <- function(i){
    if(i==1){nus <- 0.25}
    if(i==2){nus <- 0.5}
    if(i==3){nus <- 0.75}
    model<-svm(X,y,type="nu-regression",kernel="linear",nu=nus,scale=F)
    model
  }
  
  if(Sys.info()['sysname'] == 'Windows') out <- mclapply(1:svn_itor, res, mc.cores=1) else
    out <- mclapply(1:svn_itor, res, mc.cores=svn_itor)
  
  nusvm <- rep(0,svn_itor)
  corrv <- rep(0,svn_itor)
  
  #do cibersort
  t <- 1
  while(t <= svn_itor) {
    weights = t(out[[t]]$coefs) %*% out[[t]]$SV
    weights[which(weights<0)]<-0
    w<-weights/sum(weights)
    u <- sweep(X,MARGIN=2,w,'*')
    k <- apply(u, 1, sum)
    nusvm[t] <- sqrt((mean((k - y)^2)))
    corrv[t] <- cor(k, y)
    t <- t + 1
  }
  
  #pick best model
  rmses <- nusvm
  mn <- which.min(rmses)
  model <- out[[mn]]
  
  #get and normalize coefficients
  q <- t(model$coefs) %*% model$SV
  q[which(q<0)]<-0
  w <- (q/sum(q))
  
  mix_rmse <- rmses[mn]
  mix_r <- corrv[mn]
  
  newList <- list("w" = w, "mix_rmse" = mix_rmse, "mix_r" = mix_r)
  
}

#do permutations
doPerm <- function(perm, X, Y){
  itor <- 1
  Ylist <- as.list(data.matrix(Y))
  dist <- matrix()
  
  while(itor <= perm){
    #print(itor)
    
    #random mixture
    yr <- as.numeric(Ylist[sample(length(Ylist),dim(X)[1])])
    
    #standardize mixture
    yr <- (yr - mean(yr)) / sd(yr)
    
    #run CIBERSORT core algorithm
    result <- CoreAlg(X, yr)
    
    mix_r <- result$mix_r
    
    #store correlation
    if(itor == 1) {dist <- mix_r}
    else {dist <- rbind(dist, mix_r)}
    
    itor <- itor + 1
  }
  newList <- list("dist" = dist)
}

#main function
CIBERSORT <- function(sig_matrix, mixture_file, perm=0, QN=TRUE){
  
  #read in data
  X <- data.matrix(sig_matrix)
  Y <- data.matrix(mixture_file)
  
  #order
  X <- X[order(rownames(X)),]
  Y <- Y[order(rownames(Y)),]
  
  P <- perm #number of permutations
  
  #anti-log if max < 50 in mixture file
  if(max(Y) < 50) {Y <- 2^Y}
  
  #quantile normalization of mixture file
  if(QN == TRUE){
    tmpc <- colnames(Y)
    tmpr <- rownames(Y)
    Y <- normalize.quantiles(Y)
    colnames(Y) <- tmpc
    rownames(Y) <- tmpr
  }
  
  #intersect genes
  Xgns <- row.names(X)
  Ygns <- row.names(Y)
  YintX <- Ygns %in% Xgns
  Y <- Y[YintX,]
  XintY <- Xgns %in% row.names(Y)
  X <- X[XintY,]
  
  #standardize sig matrix
  X <- (X - mean(X)) / sd(as.vector(X))
  
  #empirical null distribution of correlation coefficients
  if(P > 0) {nulldist <- sort(doPerm(P, X, Y)$dist)}
  
  #print(nulldist)
  
  header <- c('Mixture',colnames(X),"P-value","Correlation","RMSE")
  #print(header)
  
  output <- matrix()
  itor <- 1
  mixtures <- dim(Y)[2]
  pval <- 9999
  
  #iterate through mixtures
  while(itor <= mixtures){
    
    y <- Y[,itor]
    
    #standardize mixture
    y <- (y - mean(y)) / sd(y)
    
    #run SVR core algorithm
    result <- CoreAlg(X, y)
    
    #get results
    w <- result$w
    mix_r <- result$mix_r
    mix_rmse <- result$mix_rmse
    
    #calculate p-value
    if(P > 0) {pval <- 1 - (which.min(abs(nulldist - mix_r)) / length(nulldist))}
    
    #print output
    out <- c(colnames(Y)[itor],w,pval,mix_r,mix_rmse)
    if(itor == 1) {output <- out}
    else {output <- rbind(output, out)}
    
    itor <- itor + 1
    
  }
  
  #save results
  write.table(rbind(header,output), file="CIBERSORT-Results.txt", sep="\t", row.names=F, col.names=F, quote=F)
  
  #return matrix object containing all results
  obj <- rbind(header,output)
  obj <- obj[,-1]
  obj <- obj[-1,]
  obj <- matrix(as.numeric(unlist(obj)),nrow=nrow(obj))
  rownames(obj) <- colnames(Y)
  colnames(obj) <- c(colnames(X),"P-value","Correlation","RMSE")
  obj
}

```


```{r CIBERSORT}

#ctrlCounts <- readRDS("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_LPSonly_10000genes_5lCPM_filteredCounts.RDS")
ctrlCounts <- readRDS("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_Ctrlonly_10000genes_5lCPM_filteredCounts.RDS")
#ctrlMeta <- readRDS("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_LPSonly_10000genes_5lCPM_metadata.RDS")
ctrlMeta <- readRDS("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_Ctrlonly_10000genes_5lCPM_metadata.RDS")

#limit counts and markers to common genes

mouseMarkers <- intersect(unique(dropMarkers$geneUpper),rownames(ctrlCounts))

ctrlCountsFilt <- ctrlCounts[rownames(ctrlCounts) %in% mouseMarkers,]
dropMarkersFilt <- subset(dropMarkers, geneUpper %in% mouseMarkers)

csortWeights <- matrix(1, nrow = length(mouseMarkers), ncol = length(unique(dropMarkers$cluster)))

rownames(csortWeights) <- rownames(ctrlCountsFilt)
colnames(csortWeights) <- unique(dropMarkers$cluster)

r <- rownames(csortWeights)[432]
c <- colnames(csortWeights)[1]

for (r in rownames(csortWeights)) {
  for (c in colnames(csortWeights)) {
    gWeight <- subset(dropMarkersFilt, cluster == c & geneUpper == r)
    if (dim(gWeight)[1] == 1) {
      csortWeights[which(rownames(csortWeights) == r),
                   which(colnames(csortWeights) == c)] <- gWeight$avg_logFC
    }
  }
}


## estimate composition
ssurCtrl_deconvolute=CIBERSORT(csortWeights,ctrlCountsFilt,perm = 0,QN = TRUE)

csortWeights10fold <- 10^csortWeights
ssurCtrl_deconvolute10fold <- CIBERSORT(csortWeights10fold,ctrlCountsFilt,perm = 0,QN = TRUE)

csortWeights20fold <- 20^csortWeights
ssurCtrl_deconvolute20fold <- CIBERSORT(csortWeights20fold,ctrlCountsFilt,perm = 0,QN = TRUE)

colnames(ssurCtrl_deconvolute20fold)[1:12] <- c("monocyte","T cell","macrophage","early pro-B cell","granulocytopoietic cell","<NA>",
                                                "granulocyte","erythroblast","hematopoietic precursor cell","proerythroblast","late pro-B cell","basophil")

cellPropEst <- melt(ssurCtrl_deconvolute20fold[,1:12])

ggplot(cellPropEst,
       aes(y = value,
           group = Var2,
           fill = Var2)) +
  geom_boxplot()



dropMarkersPBMC <- subset(dropMarkers, cluster %in% c(0,1,8,10,11))

mouseMarkers <- intersect(unique(dropMarkersPBMC$geneUpper),rownames(ctrlCounts))

ctrlCountsFilt <- ctrlCounts[rownames(ctrlCounts) %in% mouseMarkers,]
dropMarkersFilt <- subset(dropMarkersPBMC, geneUpper %in% mouseMarkers)

csortWeights <- matrix(1, nrow = length(mouseMarkers), ncol = length(unique(dropMarkersPBMC$cluster)))

rownames(csortWeights) <- rownames(ctrlCountsFilt)
colnames(csortWeights) <- unique(dropMarkersPBMC$cluster)

r <- rownames(csortWeights)[432]
c <- colnames(csortWeights)[1]

for (r in rownames(csortWeights)) {
  for (c in colnames(csortWeights)) {
    gWeight <- subset(dropMarkersFilt, cluster == c & geneUpper == r)
    if (dim(gWeight)[1] == 1) {
      csortWeights[which(rownames(csortWeights) == r),
                   which(colnames(csortWeights) == c)] <- gWeight$avg_logFC
    }
  }
}


## estimate composition
#ssurCtrl_deconvolute=CIBERSORT(csortWeights,ctrlCountsFilt,perm = 0,QN = TRUE)

csortWeights10fold <- 10^csortWeights
ssurCtrl_deconvolute10foldDropPBMC <- CIBERSORT(csortWeights10fold,ctrlCountsFilt,perm = 0,QN = TRUE)

csortWeights20fold <- 20^csortWeights
ssurCtrl_deconvolute20foldDropPBMC <- CIBERSORT(csortWeights20fold,ctrlCountsFilt,perm = 0,QN = TRUE)


colnames(ssurCtrl_deconvolute20foldDropPBMC)[1:5] <- c("monocyte","T cell","macrophage","early pro-B cell","late pro-B cell")

cellPropEst <- melt(ssurCtrl_deconvolute20foldDropPBMC[,1:5])

ggplot(cellPropEst,
       aes(y = value,
           group = Var2,
           fill = Var2)) +
  geom_boxplot()


### is there any connection to status?
ctrlMeta$tCellProp <- ssurCtrl_deconvolute20foldDropPBMC[,2]
ctrlMeta$bCellProp <- ssurCtrl_deconvolute20foldDropPBMC[,5]
ctrlMeta$monocyteProp <- ssurCtrl_deconvolute20foldDropPBMC[,1]

ctrlMeta$earlyBCellProp <- ssurCtrl_deconvolute20foldDropPBMC[,4]
ctrlMeta$macrophageProp <- ssurCtrl_deconvolute20foldDropPBMC[,3]


ttestStat <- t.test(subset(ctrlMeta, status_dom_sub == "S")$tCellProp,
                    subset(ctrlMeta, status_dom_sub == "D")$tCellProp)
ggplot(ctrlMeta,
       aes(y = tCellProp,
           x = status_dom_sub,
           fill = status_dom_sub)) +
  ggtitle("", subtitle = paste0("p = ",round(ttestStat$p.value, digits = 3),
                                " | T = ",round(ttestStat$statistic, digits = 3))) +
  geom_violin(draw_quantiles = c(.25,.5,.75))

ttestStat <- t.test(subset(ctrlMeta, status_dom_sub == "S")$bCellProp,
       subset(ctrlMeta, status_dom_sub == "D")$bCellProp)

ggplot(ctrlMeta,
       aes(y = bCellProp,
           x = status_dom_sub,
           fill = status_dom_sub)) +
  ggtitle("", subtitle = paste0("p = ",round(ttestStat$p.value, digits = 3),
                                " | T = ",round(ttestStat$statistic, digits = 3))) +
  geom_violin(draw_quantiles = c(.25,.5,.75))

ttestStat <- t.test(subset(ctrlMeta, status_dom_sub == "S")$monocyteProp,
       subset(ctrlMeta, status_dom_sub == "D")$monocyteProp)

ggplot(ctrlMeta,
       aes(y = monocyteProp,
           x = status_dom_sub,
           fill = status_dom_sub)) +
  ggtitle("", subtitle = paste0("p = ",round(ttestStat$p.value, digits = 3),
                                " | T = ",round(ttestStat$statistic, digits = 3))) +
  geom_violin(draw_quantiles = c(.25,.5,.75))

saveRDS(object = ctrlMeta, file = paste0(dataDir,"ctrlMetawCIBERSORT.RDS"))




#lpsMeta <- ctrlMeta

### look across single capture
# lpsCiber <- lpsMeta[,colnames(lpsMeta) %in% c("sampleID","captureref","status_dom_sub","tCellProp","bCellProp","monocyteProp")]
# ctrlCiber <- ctrlMeta[,colnames(ctrlMeta) %in% c("sampleID","captureref","status_dom_sub","tCellProp","bCellProp","monocyteProp")]
# 
# ciberMerge <- merge(lpsCiber, ctrlCiber, by = "captureref", suffixes = c(".lps",".ctrl"))
# 
# ggplot(ciberMerge,
#        aes(x = tCellProp.ctrl,
#            y = tCellProp.lps)) +
#   geom_point(aes(col = status_dom_sub.ctrl)) +
#   geom_smooth(method = "lm", formula = "y~x")
# 
# ggplot(ciberMerge,
#        aes(x = bCellProp.ctrl,
#            y = bCellProp.lps)) +
#   geom_point(aes(col = status_dom_sub.ctrl)) +
#   geom_smooth(method = "lm", formula = "y~x")
# 
# ggplot(ciberMerge,
#        aes(x = monocyteProp.ctrl,
#            y = monocyteProp.lps)) +
#   geom_point(aes(col = status_dom_sub.ctrl)) +
#   geom_smooth(method = "lm", formula = "y~x")
# 
# cor.test(ciberMerge$tCellProp.lps,ciberMerge$tCellProp.ctrl)
# cor.test(ciberMerge$bCellProp.lps,ciberMerge$bCellProp.ctrl)
# cor.test(ciberMerge$monocyteProp.lps,ciberMerge$monocyteProp.ctrl)
# 
# 
# ciberStack <- rbind.data.frame(lpsCiber,ctrlCiber)
# 
# ttestStat <- t.test(subset(ciberStack, status_dom_sub == "S")$tCellProp,
#                     subset(ciberStack, status_dom_sub == "D")$tCellProp)
# 
# ggplot(ciberStack,
#        aes(y = tCellProp,
#            x = status_dom_sub,
#            fill = status_dom_sub)) +
#   ggtitle("", subtitle = paste0("p = ",round(ttestStat$p.value, digits = 3))) +
#   geom_violin(draw_quantiles = c(.25,.5,.75))
# 
# ttestStat <- t.test(subset(ciberStack, status_dom_sub == "S")$tCellProp,
#                     subset(ciberStack, status_dom_sub == "D")$tCellProp)
# 
# ggplot(ciberStack,
#        aes(y = bCellProp,
#            x = status_dom_sub,
#            fill = status_dom_sub)) +
#   ggtitle("", subtitle = paste0("p = ",round(ttestStat$p.value, digits = 3))) +
#   geom_violin(draw_quantiles = c(.25,.5,.75))
# 
# ttestStat <- t.test(subset(ciberStack, status_dom_sub == "S")$monocyteProp,
#                     subset(ciberStack, status_dom_sub == "D")$monocyteProp)
# 
# ggplot(ciberStack,
#        aes(y = monocyteProp,
#            x = status_dom_sub,
#            fill = status_dom_sub)) +
#   ggtitle("", subtitle = paste0("p = ",round(ttestStat$p.value, digits = 3))) +
#   geom_violin(draw_quantiles = c(.25,.5,.75))



```
how does a PCA of cell composition look?
how much variance do PC1 & 2 explains
do either PC1 or PC2 correlate with status
does including PC1 or PC2 into the model change effect of status


```{r cell comp and status}
finalCellComp <- as.matrix(ssurCtrl_deconvolute20foldDropPBMC[,1:5])

ccPCA <- prcomp(t(finalCellComp))

ctrlMeta$ccPC1 <- ccPCA$rotation[,1]
ctrlMeta$ccPC2 <- ccPCA$rotation[,2]
ctrlMeta$ccPC3 <- ccPCA$rotation[,3]

ggplot(ctrlMeta,
       aes(x = ccPC1,
           y = ccPC2,
           col = status_dom_sub)) +
  geom_point()

ggplot(ctrlMeta,
       aes(x = ccPC1,
           y = ccPC3,
           col = status_dom_sub)) +
  geom_point()

ggplot(ctrlMeta,
       aes(x = ccPC2,
           y = ccPC3,
           col = status_dom_sub)) +
  geom_point()

ggplot(ctrlMeta,
       aes(x = status_dom_sub,
           y = ccPC1,
           fill = status_dom_sub)) +
  geom_violin() +
  geom_point()

cor.test(ctrlMeta$ccPC1, as.numeric(ctrlMeta$status_dom_sub == "S"))

cor.test(ctrlMeta$ccPC2, as.numeric(ctrlMeta$status_dom_sub == "S"))

cor.test(ctrlMeta$ccPC3, as.numeric(ctrlMeta$status_dom_sub == "S"))

ggplot(ctrlMeta,
       aes(x = as.numeric(ctrlMeta$status_dom_sub == "S"),
           y = ccPC1)) +
  geom_point() +
  geom_smooth(method = "lm", formula = "y~x")

ggplot(ctrlMeta,
       aes(x = as.numeric(ctrlMeta$status_dom_sub == "S"),
           y = ccPC1)) +
  geom_jitter(width = .1) +
  geom_smooth(method = "lm", formula = "y~x")

ccPCA$sdev / sum(ccPCA$sdev)


cellCompPCs <- ccPCA$rotation
cellCompPCMeta <- ctrlMeta[,which(colnames(ctrlMeta) %in% c("sampleID","correctsampleID","captureref"))]

saveRDS(object = cellCompPCs, file = paste0(dataDir,"cellCompPCs.RDS"))
saveRDS(object = cellCompPCMeta, file = paste0(dataDir,"cellCompPCmeta.RDS"))
```


Drop the cell composition aspect and just use B Cells and T Cells
```{r run emmreml with cell composition PCs}

focalTreatment <- "LPS"

for (focalTreatment in c("LPS","Dex","Gard","Ctrl")) {
  
  print(focalTreatment)

  # test a change
  
  batchList <- "allBatches"
  
  if (focalTreatment %in% c("LPS","Ctrl")) {
    modelName <- "F"; model <- "F"
  } else {
    modelName <- "G"; model <- "G"
  }
  minGenes <- 10000
  minLogCPM <- 5
  # 
  Kmat <- TRUE
  Kfile <- "~/Dropbox (Personal)/meerkats/KMPsamples/KMPrelatedness_matrix.RDS"
  # 
  singleTrt <- TRUE
  combineTrt <- FALSE
  # 
  if (singleTrt) {
   gatherSubset <- paste0(batchList,"_",focalSex,"_",focalTreatment,"only_",minGenes,"genes_",minLogCPM,"lCPM")
  } else if (combineTrt)  {
   gatherSubset <- paste0(batchList,"_",focalSex,"_allTrt_",minGenes,"genes_",minLogCPM,"lCPM")
  } else {
   gatherSubset <- paste0(batchList,"_",focalSex,"_",focalTreatment,"_",minGenes,"genes_",minLogCPM,"lCPM")
  }
  # 
  perms <- FALSE
  
  # focalTreatment <- "Gard"; modelName <- "G"
  # focalSex <- "F"; LM <- "basic+Preg+WeightwKmatrix"
  #focalSex <- "M"; LM <- "basic+WeightwKmatrix"
  
  
  LM <- "basic+Preg+Weight+BTcell"
  
  #runEmreml <- function(workDir, gatherSubset, model, LM, perms = FALSE, permVar = "treatment", permNum = 1, 
  #                      Kmat = TRUE, Kfile = "~/Dropbox (Personal)/meerkats/KMPsamples/KMPrelatedness_matrix.RDS",
  #                      permVersion = 1)
  
  dataDir <- paste0(workDir,"data/")
  outDir <- paste0(workDir,"output/")
  figDir <- paste0(workDir,"figures/")
  
  dir.create(paste0(figDir,gatherSubset,"/model",model))
  modelDir <- paste0(figDir,gatherSubset,"/model",model)
  if (perms == TRUE) {
    modelDir <- paste0(figDir,gatherSubset,"/model",model,"/perms")
    dir.create(modelDir)
  }
  
  #read in data
  subsetMeta <- readRDS(file = paste0(outDir,"/",gatherSubset,"_metadata.RDS"))
  subsetCountsDF <- readRDS(file = paste0(outDir,"/",gatherSubset,"_counts.RDS"))
  filterCountsDF <- readRDS(file = paste0(outDir,"/",gatherSubset,"_filteredCounts.RDS"))
  
  resids <- readRDS(file = paste0(outDir,"/",gatherSubset,"_model",model,"_resids.RDS"))
  
  if (Kmat) {
    K <- readRDS(Kfile)
  }
  
  ### load counts, residuals, and metadata
  #subsetMeta <- read.table(file = paste0(outDir,modelName,"/",modelName,"_metadata"))
  #filterCountsDF <- read.table(file = paste0(outDir,modelName,"/",modelName,"_filtCounts"))
  colnames(filterCountsDF) <- subsetMeta$sampleID
  
  ### CHECK THAT THEY MATCH?!?
  
  #resids <- read.table(file = paste0(outDir,modelName,"/",modelName,"_residuals"))
  resids <- as.matrix(resids)
  colnames(resids) <- subsetMeta$sampleID
  
  nSamples <- dim(subsetMeta)[1]
  nIndividuals <- length(unique(subsetMeta$correctsampleID))
  nGenes <- dim(resids)[1]
  
  
  #Z - individual to sample mapping
  sampleIndMap <- diag(0,nrow=nSamples,ncol=nIndividuals)
  rownames(sampleIndMap) <- subsetMeta$sampleID
  colnames(sampleIndMap) <- unique(subsetMeta$correctsampleID)
  
  for (i in 1:nSamples) {
    for (j in 1:nIndividuals) {
      sampleIndMap[i,j] <- 1 * (subsetMeta$correctsampleID[i] == colnames(sampleIndMap)[j])
    }
  }
  
  if (Kmat) {
    #start with identity
    identityK <- diag(1,nrow=nIndividuals,ncol=nIndividuals)
    rownames(identityK) <- unique(subsetMeta$correctsampleID)
    colnames(identityK) <- unique(subsetMeta$correctsampleID)
    
    Ksubset <- K[rownames(K) %in% subsetMeta$kmpID,colnames(K) %in% subsetMeta$kmpID]
    
    #look through identityK and replace 0's with correct value from Ksubset
    for (i in 1:nIndividuals) {
      for (j in 1:nIndividuals) {
        if (i != j) {
          iKMPid <- subset(subsetMeta, correctsampleID == rownames(identityK)[i])$kmpID[1]
          jKMPid <- subset(subsetMeta, correctsampleID == rownames(identityK)[j])$kmpID[1]
          
          if (iKMPid %in% rownames(Ksubset) & jKMPid %in% colnames(Ksubset)) {
            identityK[i,j] <- Ksubset[which(rownames(Ksubset) == iKMPid),
                                      which(colnames(Ksubset) == jKMPid)]
          }
        }
      }
    }
    
  } else {
    #K relatedness, none as of yet
    identityK <- diag(1,nrow=nIndividuals,ncol=nIndividuals)
  }
  
  
  subsetMeta$permID <- paste0(subsetMeta$animal_id,"_",subsetMeta$captureref)
  
  #add CC data to metadata
  ctrlMetawCIBER <- readRDS(paste0(dataDir,"ctrlMetawCIBERSORT.RDS"))
  #cellCompPCmeta <- readRDS(paste0(dataDir,"cellCompPCmeta.RDS"))
  
  subsetMeta$ccPair <- paste0(subsetMeta$correctsampleID,"_",subsetMeta$captureref)
  ctrlMetawCIBER$ccPair <- paste0(ctrlMetawCIBER$correctsampleID,"_",ctrlMetawCIBER$captureref)
  
  subsetMeta$bCellProp <- NA
  subsetMeta$tCellProp <- NA
  
  avgBcell <- mean(ctrlMetawCIBER$bCellProp)
  avgTcell <- mean(ctrlMetawCIBER$tCellProp)
  
  for (s in 1:dim(subsetMeta)[1]) {
    if (subsetMeta$ccPair[s] %in% ctrlMetawCIBER$ccPair) {
      subsetMeta$bCellProp[s] <- ctrlMetawCIBER$bCellProp[which(ctrlMetawCIBER$ccPair == subsetMeta$ccPair[s])]
      subsetMeta$tCellProp[s] <- ctrlMetawCIBER$tCellProp[which(ctrlMetawCIBER$ccPair == subsetMeta$ccPair[s])]
    } else {
      subsetMeta$bCellProp[s] <- avgBcell
      subsetMeta$tCellProp[s] <- avgTcell
    }
  }

  
  

  
  fixed_cov <- subsetMeta[which(colnames(subsetMeta) %in% c("Trt","sex","status_dom_sub","ageAtSample","permID","animal_id","pregLact","climSeason","avgWeight6mo","captureref","bCellProp","tCellProp"))]
  
  colnames(fixed_cov)[which(colnames(fixed_cov) == "Trt")] <- "treatment"
  colnames(fixed_cov)[which(colnames(fixed_cov) == "status_dom_sub")] <- "status"
  colnames(fixed_cov)[which(colnames(fixed_cov) == "ageAtSample")] <- "age"
  colnames(fixed_cov)[which(colnames(fixed_cov) == "pregLact")] <- "reprod"
  colnames(fixed_cov)[which(colnames(fixed_cov) == "climSeason")] <- "season"
  colnames(fixed_cov)[which(colnames(fixed_cov) == "avgWeight6mo")] <- "weight"
  
  #singleTrt <- FALSE
  #combineTrt <- FALSE
  ## learn if this is single treatment from the data
  if (length(unique(fixed_cov$treatment)) == 1) {
    singleTrt <- TRUE
    trtName <- unique(fixed_cov$treatment)
  } else if (length(unique(fixed_cov$treatment)) == 2) {
    trtName <- unique(fixed_cov$treatment)[which(unique(fixed_cov$treatment) != "Ctrl")]
  } else {
    combineTrt <- TRUE
    trtName <- "all"
  }

  fixed_cov$age <- scale(fixed_cov$age, scale = F, center = T)
  fixed_cov$weightUnScaled <- fixed_cov$weight
  animalWeights <- unique(fixed_cov$weight)
  scaledWeights <- scale(animalWeights, scale = F, center = T)
  ###WORKTODO###
  #need to fix the weights to be truly scaled (each individual counts once)
  for (n in 1:dim(fixed_cov)[1]) {
    matchRow <- which(animalWeights == fixed_cov$weightUnScaled[n])
    fixed_cov$weight[n] <- scaledWeights[matchRow]
  }
  
  #same thing for status only
  dWeights <- unique(subset(fixed_cov, status == "D")$weightUnScaled)
  dWeightsScaled <- scale(dWeights, scale = F, center = T)
  sWeights <- unique(subset(fixed_cov, status == "S")$weightUnScaled)
  sWeightsScaled <- scale(sWeights, scale = F, center = T)
  fixed_cov$weightStatusScaled <- 0
  
  for (n in 1:dim(fixed_cov)[1]) {
    if (fixed_cov$status[n] == "D") {
      matchRow <- which(dWeights == fixed_cov$weightUnScaled[n])
      fixed_cov$weightStatusScaled[n] <- dWeightsScaled[matchRow]
    } else {
      matchRow <- which(sWeights == fixed_cov$weightUnScaled[n])
      fixed_cov$weightStatusScaled[n] <- sWeightsScaled[matchRow]
    }
  }
  
  if (singleTrt == FALSE & combineTrt == FALSE) {
    fixed_cov$treatment <- factor(fixed_cov$treatment,
                                  levels = c("Ctrl",
                                           trtName))
  } else if (singleTrt == FALSE & combineTrt == TRUE) {
    fixed_cov$treatment <- factor(fixed_cov$treatment,
                                  levels = c("Ctrl",
                                             "Dex",
                                             "Gard",
                                             "LPS"))
  }
  
  fixed_cov$status <- factor(fixed_cov$status, levels = c("S","D"))
  fixed_cov$sex <- factor(fixed_cov$sex, levels = c("F","M"))
  fixed_cov$reprod <- factor(fixed_cov$reprod, levels = c("none","preg","lact"))
  fixed_cov$preg <- ifelse(fixed_cov$reprod == "preg", TRUE, FALSE)
  
  fixed_cov$season <- factor(fixed_cov$season, levels = c("warm-wet","cold-dry"))
  
  if ( perms ) {
    if (permVersion == 1) {
      trueFixed_Cov <- fixed_cov
      if ( permVar == "treatment" ) {
        #permute treatment variable here
        possibleTrts <- unique(fixed_cov$treatment)
        
        chall <- unique(fixed_cov$permID)[22]
        for (chall in unique(fixed_cov$permID)) {
          covRows <- which(fixed_cov$permID == chall)
          
          actualTrts <- fixed_cov$treatment[covRows]
          #permTrts <- actualTrts
          
          #if there are 2 entries, randomly flip or not
          if (length(actualTrts) == 2) {
            if (runif(1) < .5) {
              permTrts <- actualTrts[c(2,1)]
            } else {
              permTrts <- actualTrts
            }
          } else {
            #otherwise randomly pick trt or control
            if (runif(1) < .5) {
              permTrts <- possibleTrts[1]
            } else {
              permTrts <- possibleTrts[2]
            }
          }
          fixed_cov$treatment[covRows] <- permTrts
        }
        
      } else if ( permVar %in% c("status","preg","weight","age") ) {
        #fixed_cov$animalStatus <- paste0(fixed_cov$animal_id,"_",fixed_cov$status)
        #get all captures and their status/weight/age/preg
        captStat <- unique(fixed_cov[, which(colnames(fixed_cov) %in% c("captureref",permVar))])
        varColNum <- which(colnames(fixed_cov) == permVar)
        trueCaptStat <- captStat
        
        #permute statuses
        captStat[,2] <- sample(captStat[,2], size = length(captStat[,2]), replace = FALSE)
        
        #fill new status back into fixed_cov DF
        for (sampleEntry in 1:dim(fixed_cov)[1]) {
          captID <- fixed_cov$captureref[sampleEntry]
          permStatus <- subset(captStat, captureref == captID)[,2]
          fixed_cov[sampleEntry,varColNum] <- permStatus
        }
      } 
    } else if (permVersion == 2) {
      #shuffle order of columns in residual object
      trueResids <- resids
      resids <- resids[,sample(1:dim(resids)[2], size = dim(resids)[2], replace = FALSE)]
      
      ### shuffle WITH BLOCKS?!?
      
    } else if (permVersion == 3) {
      trueFixed_Cov <- fixed_cov
      #get all captures and their status+weight+age+preg
      #### TO-DO sort out male/female here (drop preg in male models)
      captStat <- unique(fixed_cov[, which(colnames(fixed_cov) %in% c("captureref","status","weight","age","preg"))])
      #varColNum <- which(colnames(fixed_cov) == permVar)
      trueCaptStat <- captStat
      
      #permute CAPTURES
      captStat[,1] <- sample(captStat[,1], size = length(captStat[,1]), replace = FALSE)
      
      #fill new status/weight/age/preg back into fixed_cov DF
      for (sampleEntry in 1:dim(fixed_cov)[1]) {
        captID <- fixed_cov$captureref[sampleEntry]
        
        fixed_cov$status[sampleEntry] <- subset(captStat, captureref == captID)$status
        fixed_cov$age[sampleEntry] <- (subset(captStat, captureref == captID)$age)
        fixed_cov$weight[sampleEntry] <- (subset(captStat, captureref == captID)$weight)
        fixed_cov$preg[sampleEntry] <- subset(captStat, captureref == captID)$preg
      }
    }
  }
  
  #?#?# differentiate trt-ctrl v condition analysis here
  if (singleTrt) {
    if (LM == "nested") {
      LM <- "basic"
    }
  }
  
  
  if ( LM == "basic" ) {
    
    if (singleTrt) {
      
      #basic model
      fixedCovMM <- model.matrix(~ fixed_cov$status + 
                                   fixed_cov$age)
      
      emmremlOutput <- as.data.frame(diag(0,ncol = dim(fixedCovMM)[2] * 3, nrow = dim(resids)[1]))
      rownames(emmremlOutput) <- rownames(resids)
      colnames(emmremlOutput) <- c('beta_intercept','beta_status','beta_age',
                                   'var_beta_intercept','var_beta_status','var_beta_age',
                                   'pval_intercept','pval_status','pval_age')
    } else {

      #basic model
      fixedCovMM <- model.matrix(~ fixed_cov$treatment + 
                                   fixed_cov$status + 
                                   fixed_cov$age)
      
      emmremlOutput <- as.data.frame(diag(0,ncol = dim(fixedCovMM)[2] * 3, nrow = dim(resids)[1]))
      rownames(emmremlOutput) <- rownames(resids)
      colnames(emmremlOutput) <- c('beta_intercept','beta_treatment','beta_status','beta_age',
                                   'var_beta_intercept','var_beta_treatment','var_beta_status','var_beta_age',
                                   'pval_intercept','pval_treatment','pval_status','pval_age')
    }
      } else if ( LM == "basic+Weight" ) {
        
        if (singleTrt) {
          
          #basic model
          fixedCovMM <- model.matrix(~ fixed_cov$status + 
                                       fixed_cov$age +
                                       fixed_cov$weight)
          
          emmremlOutput <- as.data.frame(diag(0,ncol = dim(fixedCovMM)[2] * 3, nrow = dim(resids)[1]))
          rownames(emmremlOutput) <- rownames(resids)
          colnames(emmremlOutput) <- c('beta_intercept','beta_status','beta_age','beta_weight',
                                       'var_beta_intercept','var_beta_status','var_beta_age','var_beta_weight',
                                       'pval_intercept','pval_status','pval_age','pval_weight')
        } else {
          
          #basic model
          fixedCovMM <- model.matrix(~ fixed_cov$treatment + 
                                       fixed_cov$status + 
                                       fixed_cov$age +
                                       fixed_cov$weight)
          
          emmremlOutput <- as.data.frame(diag(0,ncol = dim(fixedCovMM)[2] * 3, nrow = dim(resids)[1]))
          rownames(emmremlOutput) <- rownames(resids)
          colnames(emmremlOutput) <- c('beta_intercept','beta_treatment','beta_status','beta_age','beta_weight',
                                       'var_beta_intercept','var_beta_treatment','var_beta_status','var_beta_age','var_beta_weight',
                                       'pval_intercept','pval_treatment','pval_status','pval_age','pval_weight')
        }
        
        } else if ( LM == "basicMF+Weight" ) {
          
          if (singleTrt) {
            
            #basic model
            fixedCovMM <- model.matrix(~ fixed_cov$status + 
                                         fixed_cov$age +
                                         fixed_cov$weight +
                                         fixed_cov$sex)
            
            emmremlOutput <- as.data.frame(diag(0,ncol = dim(fixedCovMM)[2] * 3, nrow = dim(resids)[1]))
            rownames(emmremlOutput) <- rownames(resids)
            colnames(emmremlOutput) <- c('beta_intercept','beta_status','beta_age','beta_weight','beta_sex',
                                         'var_beta_intercept','var_beta_status','var_beta_age','var_beta_weight','var_beta_sex',
                                         'pval_intercept','pval_status','pval_age','pval_weight','pval_sex')
          } else {
            
            #basic model
            fixedCovMM <- model.matrix(~ fixed_cov$treatment + 
                                         fixed_cov$status + 
                                         fixed_cov$age +
                                         fixed_cov$weight +
                                         fixed_cov$sex)
            
            emmremlOutput <- as.data.frame(diag(0,ncol = dim(fixedCovMM)[2] * 3, nrow = dim(resids)[1]))
            rownames(emmremlOutput) <- rownames(resids)
            colnames(emmremlOutput) <- c('beta_intercept','beta_treatment','beta_status','beta_age','beta_weight','beta_sex',
                                         'var_beta_intercept','var_beta_treatment','var_beta_status','var_beta_age','var_beta_weight','var_beta_sex',
                                         'pval_intercept','pval_treatment','pval_status','pval_age','pval_weight','pval_sex')
          }
          
        }else if (LM == "nested") {
        
        fixedCovMM <- model.matrix(~ fixed_cov$treatment + 
                                     fixed_cov$status:fixed_cov$treatment + 
                                     fixed_cov$age)
        
        emmremlOutput <- as.data.frame(diag(0,ncol = dim(fixedCovMM)[2] * 3, nrow = dim(resids)[1]))
        rownames(emmremlOutput) <- rownames(resids)
        colnames(emmremlOutput) <- c('beta_intercept','beta_treatment','beta_age',
                                     'beta_Ctrl:Status','beta_Trt:Status',
                                     'var_beta_intercept','var_beta_treatment','var_beta_age',
                                     'var_beta_Ctrl:Status','var_beta_Trt:Status',
                                     'pval_intercept','pval_treatment','pval_age',
                                     'pval_Ctrl:Status','pval_Trt:Status')
        
        #want to compare S - D in each of LPS and Ctrl
        fixedCovMM[,4] <- ifelse(fixed_cov$treatment == trtName, 0,
                                 ifelse(fixed_cov$status == "D", 1, -1))
        
        fixedCovMM[,5] <- ifelse(fixed_cov$treatment == "Ctrl", 0,
                                 ifelse(fixed_cov$status == "D", 1, -1))
      } else if (LM == "nested+Weight") {
        
        fixedCovMM <- model.matrix(~ fixed_cov$treatment + 
                                     fixed_cov$status:fixed_cov$treatment + 
                                     fixed_cov$age +
                                     fixed_cov$weight)
        
        emmremlOutput <- as.data.frame(diag(0,ncol = dim(fixedCovMM)[2] * 3, nrow = dim(resids)[1]))
        rownames(emmremlOutput) <- rownames(resids)
        colnames(emmremlOutput) <- c('beta_intercept','beta_treatment','beta_age','beta_weight',
                                     'beta_Ctrl:Status','beta_Trt:Status',
                                     'var_beta_intercept','var_beta_treatment','var_beta_age','var_beta_weight',
                                     'var_beta_Ctrl:Status','var_beta_Trt:Status',
                                     'pval_intercept','pval_treatment','pval_age','pval_weight',
                                     'pval_Ctrl:Status','pval_Trt:Status')
        
        #want to compare S - D in each of LPS and Ctrl
        fixedCovMM[,5] <- ifelse(fixed_cov$treatment == trtName, 0,
                                 ifelse(fixed_cov$status == "D", 1, -1))
        
        fixedCovMM[,6] <- ifelse(fixed_cov$treatment == "Ctrl", 0,
                                 ifelse(fixed_cov$status == "D", 1, -1))
      } else if (LM == "nestedWPreg") {
        
        fixedCovMM <- model.matrix(~ fixed_cov$treatment + 
                                     fixed_cov$status:fixed_cov$treatment + 
                                     fixed_cov$age + 
                                     fixed_cov$preg)
        
        emmremlOutput <- as.data.frame(diag(0,ncol = dim(fixedCovMM)[2] * 3, nrow = dim(resids)[1]))
        rownames(emmremlOutput) <- rownames(resids)
        colnames(emmremlOutput) <- c('beta_intercept','beta_treatment','beta_age','beta_preg',
                                     'beta_Ctrl:Status','beta_Trt:Status',
                                     'var_beta_intercept','var_beta_treatment','var_beta_age','var_beta_preg',
                                     'var_beta_Ctrl:Status','var_beta_Trt:Status',
                                     'pval_intercept','pval_treatment','pval_age','pval_preg',
                                     'pval_Ctrl:Status','pval_Trt:Status')
        
        #want to compare S - D in each of LPS and Ctrl
        fixedCovMM[,5] <- ifelse(fixed_cov$treatment == trtName, 0,
                                 ifelse(fixed_cov$status == "D", 1, -1))
        
        fixedCovMM[,6] <- ifelse(fixed_cov$treatment == "Ctrl", 0,
                                 ifelse(fixed_cov$status == "D", 1, -1))
      } else if (LM == "nested+Preg+Weight") {
        fixedCovMM <- model.matrix(~ fixed_cov$treatment + 
                                     fixed_cov$status:fixed_cov$treatment + 
                                     fixed_cov$age + 
                                     fixed_cov$preg + 
                                     fixed_cov$weight)
        
        emmremlOutput <- as.data.frame(diag(0,ncol = dim(fixedCovMM)[2] * 3, nrow = dim(resids)[1]))
        rownames(emmremlOutput) <- rownames(resids)
        colnames(emmremlOutput) <- c('beta_intercept','beta_treatment','beta_age','beta_preg','beta_weight',
                                     'beta_Ctrl:Status','beta_Trt:Status',
                                     'var_beta_intercept','var_beta_treatment','var_beta_age','var_beta_preg','var_beta_weight',
                                     'var_beta_Ctrl:Status','var_beta_Trt:Status',
                                     'pval_intercept','pval_treatment','pval_age','pval_preg','pval_weight',
                                     'pval_Ctrl:Status','pval_Trt:Status')
        
        #want to compare S - D in each of LPS and Ctrl
        fixedCovMM[,6] <- ifelse(fixed_cov$treatment == trtName, 0,
                                 ifelse(fixed_cov$status == "D", 1, -1))
        
        fixedCovMM[,7] <- ifelse(fixed_cov$treatment == "Ctrl", 0,
                                 ifelse(fixed_cov$status == "D", 1, -1))
        
      } else if (LM == "nestedPreg") {
        
        ###only test preg-v-non here
        fixedCovMM <- model.matrix(~ fixed_cov$status + 
                                     fixed_cov$preg:fixed_cov$status + 
                                     fixed_cov$age)
        
        emmremlOutput <- as.data.frame(diag(0,ncol = dim(fixedCovMM)[2] * 3, nrow = dim(resids)[1]))
        rownames(emmremlOutput) <- rownames(resids)
        colnames(emmremlOutput) <- c('beta_intercept','beta_status','beta_age',
                                     'beta_Sub:Preg','beta_Dom:Preg',
                                     'var_beta_intercept','var_beta_status','var_beta_age',
                                     'var_beta_Sub:Preg','var_beta_Dom:Preg',
                                     'pval_intercept','pval_status','pval_age',
                                     'pval_Sub:Preg','pval_Dom:Preg')
        
        #need to center the nested variables, -1,0,1
        #dom + preg
        fixedCovMM[,4] <- ifelse(fixed_cov$status == "S" & fixed_cov$preg == TRUE, 1, 
                                 ifelse(fixed_cov$status == "S" & fixed_cov$preg == FALSE, -1, 0))
        #sub + preg
        fixedCovMM[,5] <- ifelse(fixed_cov$status == "D" & fixed_cov$preg == TRUE, 1, 
                                 ifelse(fixed_cov$status == "D" & fixed_cov$preg == FALSE, -1, 0))
        
      } else if (LM == "nestedStatus") {
        
        ###only test preg-v-non here
        fixedCovMM <- model.matrix(~ fixed_cov$preg + 
                                     fixed_cov$status:fixed_cov$preg + 
                                     fixed_cov$age)
        
        emmremlOutput <- as.data.frame(diag(0,ncol = dim(fixedCovMM)[2] * 3, nrow = dim(resids)[1]))
        rownames(emmremlOutput) <- rownames(resids)
        colnames(emmremlOutput) <- c('beta_intercept','beta_preg','beta_age',
                                     'beta_NonPreg:Status','beta_Preg:Status',
                                     'var_beta_intercept','var_beta_preg','var_beta_age',
                                     'var_beta_NonPreg:Status','var_beta_Preg:Status',
                                     'pval_intercept','pval_preg','pval_age',
                                     'pval_NonPreg:Status','pval_Preg:Status')
        
        #need to center the nested variables, -1,0,1
        #non-preg + dom -> -1
        fixedCovMM[,4] <- ifelse(fixed_cov$preg == TRUE, 0, 
                                 ifelse(fixed_cov$status == "D", 1, -1))
        #preg + dom -> -1
        fixedCovMM[,5] <- ifelse(fixed_cov$preg == FALSE, 0, 
                                 ifelse(fixed_cov$status == "D", 1, -1))
        
      } else if (LM == "basic+Preg") {
        if (singleTrt) {
          
          fixedCovMM <- model.matrix(~ fixed_cov$status + 
                                       fixed_cov$age +
                                       fixed_cov$preg)
          
          emmremlOutput <- as.data.frame(diag(0,ncol = dim(fixedCovMM)[2] * 3, nrow = dim(resids)[1]))
          rownames(emmremlOutput) <- rownames(resids)
          colnames(emmremlOutput) <- c('beta_intercept','beta_status','beta_age','beta_preg',
                                       'var_beta_intercept','var_beta_status','var_beta_age','var_beta_preg',
                                       'pval_intercept','pval_status','pval_age','pval_preg')
        } else {
          fixedCovMM <- model.matrix(~ fixed_cov$treatment + 
                                       fixed_cov$status + 
                                       fixed_cov$age +
                                       fixed_cov$preg)
          
          emmremlOutput <- as.data.frame(diag(0,ncol = dim(fixedCovMM)[2] * 3, nrow = dim(resids)[1]))
          rownames(emmremlOutput) <- rownames(resids)
          colnames(emmremlOutput) <- c('beta_intercept','beta_treatment','beta_status','beta_age','beta_preg',
                                       'var_beta_intercept','var_beta_treatment','var_beta_status','var_beta_age','var_beta_preg',
                                       'pval_intercept','pval_treatment','pval_status','pval_age','pval_preg')
          
        }
      } else if (LM == "basic+Preg+Weight") {
        if (singleTrt) {
          
          fixedCovMM <- model.matrix(~ fixed_cov$status + 
                                       fixed_cov$age +
                                       fixed_cov$preg +
                                       fixed_cov$weight)
          
          emmremlOutput <- as.data.frame(diag(0,ncol = dim(fixedCovMM)[2] * 3, nrow = dim(resids)[1]))
          rownames(emmremlOutput) <- rownames(resids)
          colnames(emmremlOutput) <- c('beta_intercept','beta_status','beta_age','beta_preg','beta_weight',
                                       'var_beta_intercept','var_beta_status','var_beta_age','var_beta_preg','var_beta_weight',
                                       'pval_intercept','pval_status','pval_age','pval_preg','pval_weight')
        } else {
          fixedCovMM <- model.matrix(~ fixed_cov$treatment + 
                                       fixed_cov$status + 
                                       fixed_cov$age +
                                       fixed_cov$preg +
                                       fixed_cov$weight)
          
          emmremlOutput <- as.data.frame(diag(0,ncol = dim(fixedCovMM)[2] * 3, nrow = dim(resids)[1]))
          rownames(emmremlOutput) <- rownames(resids)
          colnames(emmremlOutput) <- c('beta_intercept','beta_treatment','beta_status','beta_age','beta_preg','beta_weight',
                                       'var_beta_intercept','var_beta_treatment','var_beta_status','var_beta_age','var_beta_preg','var_beta_weight',
                                       'pval_intercept','pval_treatment','pval_status','pval_age','pval_preg','pval_weight')
          
        }
      } else if (LM == "basicPreg+Weight") {
        
        if (singleTrt) {
          
          fixedCovMM <- model.matrix(~ fixed_cov$status + 
                                       fixed_cov$age +
                                       fixed_cov$preg +
                                       fixed_cov$weight)
          
          emmremlOutput <- as.data.frame(diag(0,ncol = dim(fixedCovMM)[2] * 3, nrow = dim(resids)[1]))
          rownames(emmremlOutput) <- rownames(resids)
          colnames(emmremlOutput) <- c('beta_intercept','beta_status','beta_age','beta_preg','beta_weight',
                                       'var_beta_intercept','var_beta_status','var_beta_age','var_beta_preg','var_beta_weight',
                                       'pval_intercept','pval_status','pval_age','pval_preg','pval_weight')
        } else {
          fixedCovMM <- model.matrix(~ fixed_cov$treatment + 
                                       fixed_cov$status + 
                                       fixed_cov$age +
                                       fixed_cov$preg +
                                       fixed_cov$weight)
          
          emmremlOutput <- as.data.frame(diag(0,ncol = dim(fixedCovMM)[2] * 3, nrow = dim(resids)[1]))
          rownames(emmremlOutput) <- rownames(resids)
          colnames(emmremlOutput) <- c('beta_intercept','beta_treatment','beta_status','beta_age','beta_preg','beta_weight',
                                       'var_beta_intercept','var_beta_treatment','var_beta_status','var_beta_age','var_beta_preg','var_beta_weight',
                                       'pval_intercept','pval_treatment','pval_status','pval_age','pval_preg','pval_weight')
          
        }
      } else if (LM == "basic+Preg+Weight+BTcell") {
        
        if (singleTrt) {
          
          fixedCovMM <- model.matrix(~ fixed_cov$status + 
                                       fixed_cov$age +
                                       fixed_cov$preg +
                                       fixed_cov$weight + 
                                       fixed_cov$bCellProp +
                                       fixed_cov$tCellProp)
          
          emmremlOutput <- as.data.frame(diag(0,ncol = dim(fixedCovMM)[2] * 3, nrow = dim(resids)[1]))
          rownames(emmremlOutput) <- rownames(resids)
          colnames(emmremlOutput) <- c('beta_intercept','beta_status','beta_age','beta_preg','beta_weight','beta_bCell','beta_tCell',
                                       'var_beta_intercept','var_beta_status','var_beta_age','var_beta_preg','var_beta_weight','var_beta_bCell','var_beta_tCell',
                                       'pval_intercept','pval_status','pval_age','pval_preg','pval_weight','pval_bCell','pval_tCell')
        } else {
          fixedCovMM <- model.matrix(~ fixed_cov$treatment + 
                                       fixed_cov$status + 
                                       fixed_cov$age +
                                       fixed_cov$preg +
                                       fixed_cov$weight + 
                                       fixed_cov$ccPC1 +
                                       fixed_cov$ccPC2)
          
          emmremlOutput <- as.data.frame(diag(0,ncol = dim(fixedCovMM)[2] * 3, nrow = dim(resids)[1]))
          rownames(emmremlOutput) <- rownames(resids)
          colnames(emmremlOutput) <- c('beta_intercept','beta_treatment','beta_status','beta_age','beta_preg','beta_weight','beta_ccPC1','beta_ccPC2',
                                       'var_beta_intercept','var_beta_treatment','var_beta_status','var_beta_age','var_beta_preg','var_beta_weight','var_beta_ccPC1','var_beta_ccPC2',
                                       'pval_intercept','pval_treatment','pval_status','pval_age','pval_preg','pval_weight','pval_ccPC1','pval_ccPC2')
        }
          
        } else if (LM == "nestedWeight") {
        
        #center the weights by status first
        ###WORKTODO###
        
        
        if (unique(fixed_cov$sex) == "F" ) {
          if (singleTrt) {
            
            fixedCovMM <- model.matrix(~ fixed_cov$status + 
                                         fixed_cov$age +
                                         fixed_cov$preg +
                                         fixed_cov$weightStatusScaled:fixed_cov$status)
            
            emmremlOutput <- as.data.frame(diag(0,ncol = dim(fixedCovMM)[2] * 3, nrow = dim(resids)[1]))
            rownames(emmremlOutput) <- rownames(resids)
            colnames(emmremlOutput) <- c('beta_intercept','beta_status','beta_age','beta_preg','beta_weight:sub','beta_weight:dom',
                                         'var_beta_intercept','var_beta_status','var_beta_age','var_beta_preg','var_beta_weight:sub','var_beta_weight:dom',
                                         'pval_intercept','pval_status','pval_age','pval_preg','pval_weight:sub','pval_weight:dom')
    

          } else {
            fixedCovMM <- model.matrix(~ fixed_cov$treatment + 
                                         fixed_cov$status + 
                                         fixed_cov$age +
                                         fixed_cov$preg +
                                         fixed_cov$weightStatusScaled:fixed_cov$status)
            
            emmremlOutput <- as.data.frame(diag(0,ncol = dim(fixedCovMM)[2] * 3, nrow = dim(resids)[1]))
            rownames(emmremlOutput) <- rownames(resids)
            colnames(emmremlOutput) <- c('beta_intercept','beta_treatment','beta_status','beta_age','beta_preg','beta_weight:sub','beta_weight:dom',
                                         'var_beta_intercept','var_beta_treatment','var_beta_status','var_beta_age','var_beta_preg','var_beta_weight:sub','var_beta_weight:dom',
                                         'pval_intercept','pval_treatment','pval_status','pval_age','pval_preg','pval_weight:sub','pval_weight:dom')
            
          }
        } else {
          #males only, drop preg
          if (singleTrt) {
            
            fixedCovMM <- model.matrix(~ fixed_cov$status + 
                                         fixed_cov$age +
                                         fixed_cov$weightStatusScaled:fixed_cov$status)
            
            emmremlOutput <- as.data.frame(diag(0,ncol = dim(fixedCovMM)[2] * 3, nrow = dim(resids)[1]))
            rownames(emmremlOutput) <- rownames(resids)
            colnames(emmremlOutput) <- c('beta_intercept','beta_status','beta_age','beta_weight:sub','beta_weight:dom',
                                         'var_beta_intercept','var_beta_status','var_beta_age','var_beta_weight:sub','var_beta_weight:dom',
                                         'pval_intercept','pval_status','pval_age','pval_weight:sub','pval_weight:dom')
          } else {
            fixedCovMM <- model.matrix(~ fixed_cov$treatment + 
                                         fixed_cov$status + 
                                         fixed_cov$age +
                                         fixed_cov$weightStatusScaled:fixed_cov$status)
            
            emmremlOutput <- as.data.frame(diag(0,ncol = dim(fixedCovMM)[2] * 3, nrow = dim(resids)[1]))
            rownames(emmremlOutput) <- rownames(resids)
            colnames(emmremlOutput) <- c('beta_intercept','beta_treatment','beta_status','beta_age','beta_weight:sub','beta_weight:dom',
                                         'var_beta_intercept','var_beta_treatment','var_beta_status','var_beta_age','var_beta_weight:sub','var_beta_weight:dom',
                                         'pval_intercept','pval_treatment','pval_status','pval_age','pval_weight:sub','pval_weight:dom')
            
          }
        }
      } else if (LM == "basicSeason") {
        
        fixedCovMM <- model.matrix(~ fixed_cov$status + 
                                     fixed_cov$age +
                                     fixed_cov$season)
        
        emmremlOutput <- as.data.frame(diag(0,ncol = dim(fixedCovMM)[2] * 3, nrow = dim(resids)[1]))
        rownames(emmremlOutput) <- rownames(resids)
        colnames(emmremlOutput) <- c('beta_intercept','beta_status','beta_age','beta_season',
                                     'var_beta_intercept','var_beta_status','var_beta_age','var_beta_season',
                                     'pval_intercept','pval_status','pval_age','pval_season')
      } else if (LM == "allTrtF") {
        fixedCovMM <- model.matrix(~ fixed_cov$status +
                                     fixed_cov$age +
                                     fixed_cov$treatment)
        
        emmremlOutput <- as.data.frame(diag(0,ncol = dim(fixedCovMM)[2] * 3, nrow = dim(resids)[1]))
        rownames(emmremlOutput) <- rownames(resids)
        colnames(emmremlOutput) <- c('beta_intercept','beta_status','beta_age','beta_trtDex','beta_trtGard','beta_trtLPS',
                                     'var_beta_intercept','var_beta_status','var_beta_age','var_beta_trtDex','var_beta_trtGard','var_beta_trtLPS',
                                     'pval_intercept','pval_status','pval_age','pval_trtDex','pval_trtGard','pval_trtLPS')
      } else if (LM == "allTrtM") {
        fixedCovMM <- model.matrix(~ fixed_cov$status +
                                     fixed_cov$age +
                                     fixed_cov$treatment)
        
        emmremlOutput <- as.data.frame(diag(0,ncol = dim(fixedCovMM)[2] * 3, nrow = dim(resids)[1]))
        rownames(emmremlOutput) <- rownames(resids)
        colnames(emmremlOutput) <- c('beta_intercept','beta_status','beta_age','beta_trtDex','beta_trtGard','beta_trtLPS',
                                     'var_beta_intercept','var_beta_status','var_beta_age','var_beta_trtDex','var_beta_trtGard','var_beta_trtLPS',
                                     'pval_intercept','pval_status','pval_age','pval_trtDex','pval_trtGard','pval_trtLPS')
      } else if (LM == "allTrtF+Weight") {
        fixedCovMM <- model.matrix(~ fixed_cov$status +
                                     fixed_cov$age +
                                     fixed_cov$weight +
                                     fixed_cov$treatment)
        
        emmremlOutput <- as.data.frame(diag(0,ncol = dim(fixedCovMM)[2] * 3, nrow = dim(resids)[1]))
        rownames(emmremlOutput) <- rownames(resids)
        colnames(emmremlOutput) <- c('beta_intercept','beta_status','beta_age','beta_weight','beta_trtDex','beta_trtGard','beta_trtLPS',
                                     'var_beta_intercept','var_beta_status','var_beta_age','var_beta_weight','var_beta_trtDex','var_beta_trtGard','var_beta_trtLPS',
                                     'pval_intercept','pval_status','pval_age','pval_weight','pval_trtDex','pval_trtGard','pval_trtLPS')
      } else if (LM == "allTrtF+Preg") {
        fixedCovMM <- model.matrix(~ fixed_cov$status +
                                     fixed_cov$age +
                                     fixed_cov$preg +
                                     fixed_cov$treatment)
        
        emmremlOutput <- as.data.frame(diag(0,ncol = dim(fixedCovMM)[2] * 3, nrow = dim(resids)[1]))
        rownames(emmremlOutput) <- rownames(resids)
        colnames(emmremlOutput) <- c('beta_intercept','beta_status','beta_age','beta_preg','beta_trtDex','beta_trtGard','beta_trtLPS',
                                     'var_beta_intercept','var_beta_status','var_beta_age','var_beta_preg','var_beta_trtDex','var_beta_trtGard','var_beta_trtLPS',
                                     'pval_intercept','pval_status','pval_age','pval_preg','pval_trtDex','pval_trtGard','pval_trtLPS')
      } else if (LM == "allTrtF+Preg+Weight") {
        fixedCovMM <- model.matrix(~ fixed_cov$status +
                                     fixed_cov$age +
                                     fixed_cov$preg +
                                     fixed_cov$treatment +
                                     fixed_cov$weight)
        
        emmremlOutput <- as.data.frame(diag(0,ncol = dim(fixedCovMM)[2] * 3, nrow = dim(resids)[1]))
        rownames(emmremlOutput) <- rownames(resids)
        colnames(emmremlOutput) <- c('beta_intercept','beta_status','beta_age','beta_preg','beta_trtDex','beta_trtGard','beta_trtLPS','beta_weight',
                                     'var_beta_intercept','var_beta_status','var_beta_age','var_beta_preg','var_beta_trtDex','var_beta_trtGard','var_beta_trtLPS','var_beta_weight',
                                     'pval_intercept','pval_status','pval_age','pval_preg','pval_trtDex','pval_trtGard','pval_trtLPS','pval_weight')
      } else if (LM == "allTrtM+Weight") {
        fixedCovMM <- model.matrix(~ fixed_cov$status +
                                     fixed_cov$age +
                                     fixed_cov$treatment +
                                     fixed_cov$weight)
        
        emmremlOutput <- as.data.frame(diag(0,ncol = dim(fixedCovMM)[2] * 3, nrow = dim(resids)[1]))
        rownames(emmremlOutput) <- rownames(resids)
        colnames(emmremlOutput) <- c('beta_intercept','beta_status','beta_age','beta_trtDex','beta_trtGard','beta_trtLPS','beta_weight',
                                     'var_beta_intercept','var_beta_status','var_beta_age','var_beta_trtDex','var_beta_trtGard','var_beta_trtLPS','var_beta_weight',
                                     'pval_intercept','pval_status','pval_age','pval_trtDex','pval_trtGard','pval_trtLPS','pval_weight')
      } else if (LM == "allTrtMF+Weight") {
        fixedCovMM <- model.matrix(~ fixed_cov$status +
                                     fixed_cov$age +
                                     fixed_cov$treatment +
                                     fixed_cov$weight +
                                     fixed_cov$sex)
        
        emmremlOutput <- as.data.frame(diag(0,ncol = dim(fixedCovMM)[2] * 3, nrow = dim(resids)[1]))
        rownames(emmremlOutput) <- rownames(resids)
        colnames(emmremlOutput) <- c('beta_intercept','beta_status','beta_age','beta_trtDex','beta_trtGard','beta_trtLPS','beta_weight','beta_sex',
                                     'var_beta_intercept','var_beta_status','var_beta_age','var_beta_trtDex','var_beta_trtGard','var_beta_trtLPS','var_beta_weight','var_beta_sex',
                                     'pval_intercept','pval_status','pval_age','pval_trtDex','pval_trtGard','pval_trtLPS','pval_weight','pval_sex')
      }
  
  #vUcheck <- vector()
  #vEcheck <- vector()
  
  for (g in 1:nGenes) {
      temp <- emmreml(y = resids[g,],
                      X = fixedCovMM,
                      K = identityK,
                      Z = sampleIndMap,
                      varbetahat = T,varuhat=T,PEVuhat=T,test=T)
      
      p <- temp$pvalbeta[,"none"]
      varb <- temp$varbetahat
      b <- temp$betahat
      
      #vUcheck <- c(vUcheck,temp$Vu)
      #vEcheck <- c(vEcheck,temp$Ve)
      
      emmremlOutput[g,] <- as.vector(c(b,varb,p))
  }
  #system("say dingh")
  
  #ggplot() + geom_histogram(aes(x = emmremlOutput$pval_status), bins = 100)
  #ggplot() + geom_histogram(aes(x = emmremlOutput$pval_weight), bins = 100)
  #ggplot() + geom_point(aes(x = vUcheck, y = vEcheck))
  #sum(vUcheck > .5)
  #ggplot() + geom_point(aes(x = emmremlOutput$beta_weight, y = emmremlOutput$beta_status), alpha = .2)
  #cor(emmremlOutput$beta_status, emmremlOutput$beta_weight)
  #vUcheckPerm2rand <- vUcheck
  #vEcheckPerm2rand <- vEcheck
  
  #ggplot() + geom_point(aes(x = vUcheckPerm0, y = vUcheckPerm2rand), alpha = .2) + geom_abline(slope = 1, intercept = 0)
  #ggplot() + geom_point(aes(x = vEcheckPerm0, y = vEcheckPerm2rand), alpha = .2) + geom_abline(slope = 1, intercept = 0)
  
  #add some model info to the titles
  if (trtName %in% c("LPS","Gard","Dex")) {
    if (singleTrt) {
      trtTitle <- paste0(trtName,"only")
    } else {
      trtTitle <- paste0(trtName,"-Ctrl")
    }
  } else {
    trtTitle <- "Ctrlonly"
  }
  
  #if using a Kmat note it in the model
  if (Kmat) {
    LMbasic <- LM
    LM <- paste0(LM,"wKmatrix")
  } else {
    LMbasic <- LM
  }
  
    
  #figOut <- paste0(figDir,modelName)
  if ( perms == FALSE ) {
    ggplot() +
      geom_histogram(aes(x = emmremlOutput$pval_intercept), bins = 100) +
      xlab("p-value - intercept") + ggtitle(paste0("resid - ",model," | LM - ",LM," | ",trtName))
    ggsave(paste0(modelDir,"/",LM,"_pval_intercept.png"), width = 6, height = 4)
    
    if (singleTrt == FALSE & combineTrt == FALSE) {
      ggplot() +
        geom_histogram(aes(x = emmremlOutput$pval_treatment), bins = 100) +
        xlab("p-value - treatment") + ggtitle(paste0("resid - ",model," | LM - ",LM," | ",trtName))
      ggsave(paste0(modelDir,"/",LM,"_pval_treatment.png"), width = 6, height = 4)
    } else if (combineTrt == TRUE) {
      ggplot() +
        geom_histogram(aes(x = emmremlOutput$pval_trtDex), bins = 100) +
        xlab("p-value - Dex trt") + ggtitle(paste0("resid - ",model," | LM - ",LM," | ",trtName))
      ggsave(paste0(modelDir,"/",LM,"_pval_trtDex.png"), width = 6, height = 4)
      
      ggplot() +
        geom_histogram(aes(x = emmremlOutput$pval_trtGard), bins = 100) +
        xlab("p-value - Gard trt") + ggtitle(paste0("resid - ",model," | LM - ",LM," | ",trtName))
      ggsave(paste0(modelDir,"/",LM,"_pval_trtGard.png"), width = 6, height = 4)
      
      ggplot() +
        geom_histogram(aes(x = emmremlOutput$pval_trtLPS), bins = 100) +
        xlab("p-value - LPS trt") + ggtitle(paste0("resid - ",model," | LM - ",LM," | ",trtName))
      ggsave(paste0(modelDir,"/",LM,"_pval_trtLPS.png"), width = 6, height = 4)
    }
    
    ggplot() +
      geom_histogram(aes(x = emmremlOutput$pval_age), bins = 100) +
      xlab("p-value - age") + ggtitle(paste0("resid - ",model," | LM - ",LM," | ",trtName))
    ggsave(paste0(modelDir,"/",LM,"_pval_age.png"), width = 6, height = 4)
    
    if (LMbasic %in% c("basic","basic+Preg","nestedPreg","allTrtF","allTrtF+Weight","allTrtM",
                       "allTrtM+Weight","basic+Preg+Weight","basic+Preg+Weight+CC","basic+Weight","nestedWeight","allTrtF+Preg+Weight")) {
      ggplot() +
        geom_histogram(aes(x = emmremlOutput$pval_status), bins = 100) +
        xlab("p-value - status") + ggtitle(paste0("resid - ",model," | LM - ",LM," | ",trtName))
      ggsave(paste0(modelDir,"/",LM,"_pval_status.png"), width = 6, height = 4)
    }
    
    if (LMbasic %in% c("basic+Preg","nestedStatus","nestedWPreg","nested+Preg+Weight","allTrtF+Preg",
                       "nestedWeight","basic+Preg+Weight","basic+Preg+Weight+CC","basicPregWeightOnly","allTrtF+Preg+Weight")) {
      ggplot() +
        geom_histogram(aes(x = emmremlOutput$pval_preg), bins = 100) +
        xlab("p-value - preg") + ggtitle(paste0("resid - ",model," | LM - ",LM," | ",trtName))
      ggsave(paste0(modelDir,"/",LM,"_pval_preg.png"), width = 6, height = 4)
      
      #ggplot() +
      #  geom_histogram(aes(x = emmremlOutput$pval_lact), bins = 100) +
      #  xlab("p-value - lact") + ggtitle(paste0("resid - ",model," | LM - ",LM," | ",trtName))
      #ggsave(paste0(modelDir,"/",LM,"_pval_lact.png"), width = 6, height = 4)
    }
    
    if (LMbasic %in% c("nestedPreg")) {
      ggplot() +
        geom_histogram(aes(x = emmremlOutput$pval_Dom.Preg), bins = 100) +
        xlab("p-value - Dom:Preg") + ggtitle(paste0("resid - ",model," | LM - ",LM," | ",trtName))
      ggsave(paste0(modelDir,"/",LM,"_pval_DomxPreg.png"), width = 6, height = 4)
      
      ggplot() +
        geom_histogram(aes(x = emmremlOutput$pval_Sub.Preg), bins = 100) +
        xlab("p-value - Sub:Preg") + ggtitle(paste0("resid - ",model," | LM - ",LM," | ",trtName))
      ggsave(paste0(modelDir,"/",LM,"_pval_SubxPreg.png"), width = 6, height = 4)
    }
    
    if (LMbasic %in% c("nestedStatus")) {
      ggplot() +
        geom_histogram(aes(x = emmremlOutput$pval_Preg.Status), bins = 100) +
        xlab("p-value - Preg:Status") + ggtitle(paste0("resid - ",model," | LM - ",LM," | ",trtName))
      ggsave(paste0(modelDir,"/",LM,"_pval_DomxPreg.png"), width = 6, height = 4)
      
      ggplot() +
        geom_histogram(aes(x = emmremlOutput$pval_NonPreg.Status), bins = 100) +
        xlab("p-value - NotPreg:Status") + ggtitle(paste0("resid - ",model," | LM - ",LM," | ",trtName))
      ggsave(paste0(modelDir,"/",LM,"_pval_SubxPreg.png"), width = 6, height = 4)
    }
    
    if (LMbasic %in% c("basicSeason")) {
      ggplot() +
        geom_histogram(aes(x = emmremlOutput$pval_season), bins = 100) +
        xlab("p-value - season") + ggtitle(paste0("resid - ",model," | LM - ",LM," | ",trtName))
      ggsave(paste0(modelDir,"/",LM,"_pval_season.png"), width = 6, height = 4)
      
    }
      
    if (LMbasic %in% c("nested","nested+Weight","nestedWPreg","nested+Preg+Weight")) {
      ggplot() +
        geom_histogram(aes(x = emmremlOutput$`pval_Ctrl:Status`), bins = 100) +
        xlab("p-value - Ctrl:Status") + 
        ggtitle(paste0("resid - ",model," | LM - ",LM," | ",trtName))
      ggsave(paste0(modelDir,"/",LM,"_pval_CtrlxStatus.png"), width = 6, height = 4)
      
      ggplot() +
        geom_histogram(aes(x = emmremlOutput$`pval_Trt:Status`), bins = 100) +
        xlab("p-value - Trt:Status") + 
        ggtitle(paste0("resid - ",model," | LM - ",LM," | ",trtName))
      ggsave(paste0(modelDir,"/",LM,"_pval_TrtxStatus.png"), width = 6, height = 4)
    }
    
    #weight pvals
    if (LMbasic %in% c("nested+Weight","basic+Preg+Weight","basic+Preg+Weight+CC","nested+Preg+Weight","basic+Weight",
                       "allTrtF+Weight","allTrtM+Weight","basicPregWeightOnly","allTrtF+Preg+Weight")) {
      ggplot() +
        geom_histogram(aes(x = emmremlOutput$pval_weight), bins = 100) +
        xlab("p-value - weight") + ggtitle(paste0("resid - ",model," | LM - ",LM," | ",trtName))
      ggsave(paste0(modelDir,"/",LM,"_pval_weight.png"), width = 6, height = 4)
      
    }
    
    if (LMbasic %in% c("nestedWeight")) {
      ggplot() +
        geom_histogram(aes(x = emmremlOutput$`pval_weight:dom`), bins = 100) +
        xlab("p-value - Dom:Weight") + 
        ggtitle(paste0("resid - ",model," | LM - ",LM," | ",trtName))
      ggsave(paste0(modelDir,"/",LM,"_pval_DomxWeight.png"), width = 6, height = 4)
      
      ggplot() +
        geom_histogram(aes(x = emmremlOutput$`pval_weight:sub`), bins = 100) +
        xlab("p-value - Sub:Weight") + 
        ggtitle(paste0("resid - ",model," | LM - ",LM," | ",trtName))
      ggsave(paste0(modelDir,"/",LM,"_pval_SubxWeight.png"), width = 6, height = 4)
    }
    
    if (LMbasic %in% c("basic+Preg+Weight+CC")) {
      ggplot() +
        geom_histogram(aes(x = emmremlOutput$pval_ccPC1), bins = 100) +
        xlab("p-value - CellComp PC1") + 
        ggtitle(paste0("resid - ",model," | LM - ",LM," | ",trtName))
      ggsave(paste0(modelDir,"/",LM,"_pval_ccPC1.png"), width = 6, height = 4)
      
      ggplot() +
        geom_histogram(aes(x = emmremlOutput$pval_ccPC2), bins = 100) +
        xlab("p-value - CellComp PC2") + 
        ggtitle(paste0("resid - ",model," | LM - ",LM," | ",trtName))
      ggsave(paste0(modelDir,"/",LM,"_pval_ccPC2.png"), width = 6, height = 4)
    }
    
    
    #write out the emmreml
    write.table(x = emmremlOutput, file = paste0(outDir,"/",gatherSubset,"_model",model,"_",LM,"_emmreml"))
    
  } else {
    if ( permVar == "treatment" ) {
      ggplot() +
        geom_histogram(aes(x = emmremlOutput$pval_treatment), bins = 100) +
        xlab("p-value - treatment") + 
        ggtitle(paste0("resid - ",model," | LM - ",LM," | ",trtName," | perm",formatC(permNum, width=3, flag="0")))
      
      ggsave(paste0(modelDir,"/",LM,"_pval_treatment_perm",
                    formatC(permNum, width=3, flag="0"),".png"), width = 6, height = 4)
    } else if ( permVar == "status" ) {
      if (LMbasic == "basic") {
        ggplot() +
          geom_histogram(aes(x = emmremlOutput$pval_status), bins = 100) +
          xlab("p-value - status") + 
          ggtitle(paste0("resid - ",model," | LM - ",LM," | ",trtName," | perm",formatC(permNum, width=3, flag="0")))
        
        ggsave(paste0(modelDir,"/",LM,"_pval_status_perm",
                      formatC(permNum, width=3, flag="0"),".png"), width = 6, height = 4)
      }
      if (LMbasic %in% c("nested","nested+Weight")) {
        ggplot() +
          geom_histogram(aes(x = emmremlOutput$pval_Trt.Status), bins = 100) +
          xlab("p-value - Trt:Status") + 
          ggtitle(paste0("resid - ",model," | LM - ",LM," | ",trtName," | perm",formatC(permNum, width=3, flag="0")))
        
        ggsave(paste0(modelDir,"/",LM,"_pval_TrtxStatus_perm",
                      formatC(permNum, width=3, flag="0"),".png"), width = 6, height = 4)
        ggplot() +
          geom_histogram(aes(x = emmremlOutput$pval_Ctrl.Status), bins = 100) +
          xlab("p-value - Ctrl:Status") + 
          ggtitle(paste0("resid - ",model," | LM - ",LM," | ",trtName," | perm",formatC(permNum, width=3, flag="0")))
        
        ggsave(paste0(modelDir,"/",LM,"_pval_CtrlxStatus_perm",
                      formatC(permNum, width=3, flag="0"),".png"), width = 6, height = 4)
        
      }
      
    }
    
    write.table(x = emmremlOutput, file = paste0(outDir,"/",gatherSubset,"_model",model,"_",LM,"_emmreml_",
                                                 permVar,"_v",permVersion,"_perm",formatC(permNum, width=3, flag="0")))
    ### saved permuted fixed_cov
    saveRDS(object = fixed_cov, file = paste0(outDir,"/",gatherSubset,"_metadata_",permVar,"_v",permVersion,"_perm",
                                              formatC(permNum, width=3, flag="0"),".RDS"))
  }

}

```


```{r compare status beta before and after adding cell comp}

for (focalTreatment in c("LPS","Dex","Gard","Ctrl")) {
  
  if (focalTreatment %in% c("LPS","Ctrl")) {
    modelName <- "F"; model <- "F"
  } else {
    modelName <- "G"; model <- "G"
  }
  gatherSubset <- paste0(batchList,"_",focalSex,"_",focalTreatment,"only_",minGenes,"genes_",minLogCPM,"lCPM")
  
  emremlwCC <- read.table(paste0(outDir,"allBatches_F_",focalTreatment,"only_10000genes_5lCPM_model",model,"_basic+Preg+Weight+CCwKmatrix_emmreml"))
  emremlwCC$gene <- rownames(emremlwCC)
  emremlnoCC <- read.table(paste0(outDir,"allBatches_F_",focalTreatment,"only_10000genes_5lCPM_model",model,"_basic+Preg+WeightwKmatrix_emmreml"))
  emremlnoCC$gene <- rownames(emremlnoCC)
  
  nGenes <- dim(emremlwCC)[1]
  
  emremlCCComp <- merge(emremlwCC,emremlnoCC, by = "gene", suffixes = c(".wCC",".noCC"))
  
  ggplot(emremlCCComp,
         aes(x = beta_status.noCC,
                   y = beta_status.wCC)) +
    geom_point() +
    geom_smooth(method = "lm", formula = "y~x") +
    ggtitle(gatherSubset) +
    stat_cor(p.accuracy = 0.001, r.accuracy = 0.01)
  ggsave(paste0(figDir,"addCellProp_",focalTreatment,"_beta-betaPlot.png"), width = 6, height = 4)
  
  ggplot() +
    geom_point(aes(x = emremlnoCC$pval_status,
                   y = emremlwCC$pval_status)) +
    geom_abline(slope = 1, intercept = 0, col = "red") +
    scale_x_log10() + scale_y_log10()
  
  
  
  ggplot() +
    geom_point(aes(y = -log10(sort(emremlwCC$pval_status)),
                   x = -log10(sort(runif(nGenes)))), col = "green") +
    geom_point(aes(y = -log10(sort(emremlnoCC$pval_status)),
                   x = -log10(sort(runif(nGenes)))), col = "blue") +
    geom_abline(slope = 1, intercept = 0, col = "red") +
    ylab("Actual P-values (Green = w/CellProp)") +
    xlab("Flat Null P-values") +
    ggtitle(gatherSubset)
  ggsave(paste0(figDir,"addCCpc_",focalTreatment,"_qqPlot.png"), width = 6, height = 4)
  
}

```



Drop one sample from each animal that is used in the preg/non-preg OR dom/sub transition check
to start drop from emmreml analysis (seems excessive to drop from residual)

Preg-NonPreg : Drop the NONPREG version
Sub-Dom : Drop the SUB version

```{r longitudinal resamples}
#gather data
batchList <- "allBatches"
minGenes <- 10000
minLogCPM <- 5
singTrt <- FALSE
focalSex <- "F"
singTrt <- TRUE
LM <- "basic+Preg+Weight"

for (focalTreatment in c("LPS","Dex","Gard","Ctrl")) {
  for (rsVar in c("status","preg")) {
    gatherData(workDir = workDir, 
               batchList = batchList,
               focalSex = focalSex,
               focalTreatment = focalTreatment,
               minGenes = minGenes,
               minLogCPM = minLogCPM,
               singleTrt = singTrt, 
               longReSample = TRUE, 
               ReSampVar = rsVar)
  }
}

#generate residuals

for (focalTreatment in c("LPS","Dex","Gard","Ctrl")) {
  for (rsVar in c("status","preg")) {
  
  gatherSubset <- paste0(batchList,"_",focalSex,"_",focalTreatment,"only_",minGenes,"genes_",minLogCPM,"lCPM_",rsVar,"ReSamp")
  #"allBatches_F_LPSonly_10000genes_5lCPM_statusReSamp"
  
  if (focalTreatment %in% c("LPS","Ctrl")) {
    modelName <- "F"
  } else {
    modelName <- "G"
  }
  
  genResids(workDir = workDir, gatherSubset = gatherSubset, model = modelName)
  }
}



#emmreml
for (focalTreatment in c("LPS","Dex","Gard","Ctrl")) {
  for (rsVar in c("status","preg")) {
  
  print(paste0(focalSex," - ",focalTreatment," - ",rsVar))
  gatherSubset <- paste0(batchList,"_",focalSex,"_",focalTreatment,"only_",minGenes,"genes_",minLogCPM,"lCPM_",rsVar,"ReSamp")
  
  if (focalTreatment %in% c("LPS","Ctrl")) {
        modelName <- "F"
      } else {
        modelName <- "G"
      }
    runEmreml(workDir = workDir, gatherSubset = gatherSubset, model = modelName, LM = "basic+Preg+Weight",
              Kmat = TRUE, Kfile = "~/Dropbox (Personal)/meerkats/KMPsamples/KMPrelatedness_matrix.RDS")
  }
}

#permutations

permVar <- "status"
permVersion <- 1
rsVar <- "status"
for (permVar in c("status","preg")) {
  rsVar <- permVar
  for (focalTreatment in c("LPS","Ctrl","Dex","Gard")) {
    singleTrt <- TRUE
    
    if (focalTreatment %in% c("LPS","Ctrl")) {
      modelName <- "F"
    } else {
      modelName <- "G"
    }
    
    gatherSubset <- paste0(batchList,"_",focalSex,"_",focalTreatment,"only_",minGenes,"genes_",minLogCPM,"lCPM_",rsVar,"ReSamp")
    
    modelFemreml <- read.table("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_LPSonly_10000genes_5lCPM_statusReSamp_modelF_basic+Preg+WeightwKmatrix_emmreml")
    
    outputFiles <- list.files(outDir)
    
    for (permNum in 1:100) {
      print(paste0(focalTreatment," ",permVar," ",permNum))
      print(date())
      if (!paste0("allBatches_F_",focalTreatment,"only_10000genes_5lCPM_",permVar,"ReSamp_model",modelName,"_",LM,
                "wKmatrix_emmreml_",permVar,"_v",permVersion,"_perm",formatC(permNum, width=3, flag="0")) %in% outputFiles) {
        runEmreml(workDir = workDir, gatherSubset = gatherSubset, model = modelName, LM = "basic+Preg+Weight",
                  perms = TRUE, permVar = permVar, permNum = permNum, permVersion = 1, Kmat = TRUE,
                  Kfile <- "~/Dropbox (Personal)/meerkats/KMPsamples/KMPrelatedness_matrix.RDS")
      }
    }
  }
}

### used perm.fdr to calculate qvalue with the permutations

permVersion <- 1
permVar <- "status"
focalSex <- "F"
minGenes <- 10000
minLogCPM <- 5
focalTreatment <- "LPS"


for (focalTreatment in c("LPS","Dex","Gard","Ctrl")) {
  for (permVar in c("status","preg")) {
    LMvar <- permVar
    rsVar <- permVar
  
    if (focalTreatment %in% c("Dex","Gard")) {
      modelName <- "G"
    } else {
      modelName <- "F"
    }
    
    gatherSubset <- paste0(batchList,"_",focalSex,"_",focalTreatment,"only_",minGenes,"genes_",minLogCPM,"lCPM_",rsVar,"ReSamp")
    
    modelDir <- paste0(figDir,gatherSubset,"/model",modelName)
    
    LM <- "basic+Preg+Weight"
  
    #permVar is always "status" - labeled the original as such
    #permVar <- "status"
    #LMvar <- "treatment"
    
    totalPerms <- 100
    #get Emreml
    #"~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_MF_LPS_10000genes_5lCPM_modelF_basicMF+WeightwKmatrix_emmreml"
    emremlOut <- read.table(paste0(outDir,gatherSubset,"_model",modelName,"_",LM,"wKmatrix_emmreml"))
    
    #plot standard p-value
    ggplot(data = emremlOut, aes_string(x = paste0("pval_",LMvar))) +
      geom_histogram(bins = 100)
    
    #allBatches_F_LPS_10000genes_4lCPM_modelF_nested_emmreml_treatment_perm001
    permRun <- 8
    
    permPvals <- vector()
    
    for (permRun in 1:totalPerms) {
      #get permutations
      #"~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_LPSonly_10000genes_5lCPM_modelF_basic+Preg+WeightwKmatrix_emmreml_status_perm008"
      #"~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_LPSonly_10000genes_5lCPM_modelF_basic+Preg+WeightwKmatrix_emmreml_status_v1_perm008"
      permEmreml <- read.table(paste0(outDir,gatherSubset,"_model",modelName,"_",LM,"wKmatrix_emmreml_",
                                      permVar,"_v",permVersion,"_perm",formatC(permRun,width = 3, flag = "0")))
      
      permCol <- which(colnames(permEmreml) == paste0("pval_",LMvar))
      
      permPvals <- c(permPvals,permEmreml[,permCol])
    }
    
    #permutation pvals divided by number of runs should match emreml
    
    ( length(permPvals) / totalPerms ) == dim(emremlOut)[1]
    
    emremlOut <- perm.fdr(input_df = emremlOut,
                          perm_df = permPvals,
                          Pvals_col_name = paste0("pval_",LMvar),
                          name = LMvar)
    
    
    
    ggplot() +
      geom_histogram(aes_string(x = permPvals), bins = 100, fill = "tomato", alpha = .5) +
      xlab(paste0("permuted p-value - ",LMvar)) + 
      ggtitle(paste0(focalTreatment,"only - ",LM," | ",permVar," permutations"), 
              subtitle = paste0(totalPerms," permutation runs, model", modelName))
    ggsave(paste0(modelDir,"/perms/",LM,"_pval_",LMvar,"_v",permVersion,"_wPerms.png"), width = 6, height = 4)
    
    fdrCol <- which(colnames(emremlOut) == paste0("fdr_",LMvar))
    
    ggplot() +
      geom_histogram(aes(x = emremlOut[,permCol], y = ..density..), fill = "dodgerblue2", bins = 100, alpha = .5) +
      geom_density(aes(x = permPvals, y = ..density..), adjust = .25, fill = "tomato", color = "tomato", alpha = .25) +
      annotate("text", x = Inf, y = Inf, vjust = 1, hjust = 1,
               label = paste0("# Genes < FDR\n",
                              sum(emremlOut[,fdrCol] < .01)," | 0.01\n",
                              sum(emremlOut[,fdrCol] < .10)," | 0.10\n",
                              sum(emremlOut[,fdrCol] < .20)," | 0.20")) +
      xlab(paste0("p-value - ",LMvar)) + 
      ggtitle(paste0(focalTreatment,"only - ",LM," | ",permVar," permutations"), 
              subtitle = paste0(totalPerms," permutation runs, model", modelName,"permuted p-value density plot in red"))
    ggsave(paste0(modelDir,"/",LM,"_pval_",LMvar,"_v",permVersion,"_wPerms.png"), width = 6, height = 4)
    
    write.table(x = emremlOut, file = paste0(outDir,gatherSubset,"_model",modelName,"_",LM,
                                             "wKmatrix_emmreml_wPermFDR_v",permVersion,"_",permVar))
    #read.table(paste0(outDir,gatherSubset,"_model",modelName,"_",LM,"wKmatrix_emmreml"))
  }
}



```



```{r check status and preg change in resampled animals}
geneByGenePlot <- TRUE
varA <- "LPS"

wTestPValPreg <- vector()
tTestStatPreg <- vector()
wTestPValStatus <- vector()
tTestStatStatus <- vector()
emmremlStdBetaStatus <- vector()
emmremlPValStatus <- vector()
emmremlStdBetaPreg <- vector()
emmremlPValPreg <- vector()
emmremlQValPreg <- vector()
emmremlQValStatus <- vector()

#vectors for gene by gene line plot
geneNamePlot <- vector()
pregPval <- vector()
statPval <- vector()
meanGEpreg <- vector()
meanGEnonpreg <- vector()
meanGEdom <- vector()
meanGEsub <- vector()
treatment <- vector()

for (varA in c("LPS","Gard","Dex","Ctrl")) {
  #print(varA)
  if (varA %in% c("LPS","Ctrl")) {
    modelA <- "modelF"
  } else {
    modelA <- "modelG"
  }
  
  ### use the model without the repeated animals for the gene info
  emmremlResultsPreg <- read.table(paste0(outDir,"allBatches_F_",varA,"only_10000genes_5lCPM_pregReSamp_",
                                          modelA,"_basic+Preg+WeightwKmatrix_emmreml_wPermFDR_v1_preg"))
  emmremlResultsPreg$qval_preg <- qvalue(emmremlResultsPreg$pval_preg)$qvalue
  
  emmremlResultsStat <- read.table(paste0(outDir,"allBatches_F_",varA,"only_10000genes_5lCPM_statusReSamp_",
                                          modelA,"_basic+Preg+WeightwKmatrix_emmreml_wPermFDR_v1_status"))
  emmremlResultsStat$qval_status <- qvalue(emmremlResultsStat$pval_status)$qvalue
  
  residsStatus <- readRDS(paste0(outDir,"allBatches_F_",varA,"only_10000genes_5lCPM_",modelA,"_resids.RDS"))
  metaStatus <- readRDS(paste0(outDir,"allBatches_F_",varA,"only_10000genes_5lCPM_",modelA,"_dfMeta.RDS"))
  
  residsPreg <- readRDS(paste0(outDir,"allBatches_F_",varA,"only_10000genes_5lCPM_",modelA,"_resids.RDS"))
  metaPreg <- readRDS(paste0(outDir,"allBatches_F_",varA,"only_10000genes_5lCPM_",modelA,"_dfMeta.RDS"))
  
  metaPreg$preg <- ifelse(metaPreg$pregLact == "preg", TRUE, FALSE)
  metaStatus$preg <- ifelse(metaStatus$pregLact == "preg", TRUE, FALSE)
  
  metaPreg$statusPreg <- paste0(metaPreg$status_dom_sub,"_",metaPreg$preg)
  metaStatus$statusPreg <- paste0(metaStatus$status_dom_sub,"_",metaStatus$preg)
  
  ### stil using full data's resids and meta
  #geneList <- rownames(emmremlResultsPreg)
  
  ### use the intersection of genes from resample status, resample preg, and original
  geneList <- intersect(intersect(rownames(emmremlResultsPreg),rownames(emmremlResultsStat)), rownames(residsPreg))
  
  repeatSamples <- names(table(metaPreg$animal_id)[table(metaPreg$animal_id) > 1])
  
  ### find the repeat animals preg/not-preg
  s <- repeatSamples[1]
  doublePreg <- vector()
  sampleID <- vector()
  pregTF <- vector()
  status <- vector()
  animalID <- vector()
  
  for (s in repeatSamples) {
    #print(s)
    if (length(unique(subset(metaPreg, animal_id == s)$pregStatusBackDate35)) > 1) {
      #print("duplicate")
      doublePreg <- c(doublePreg,s)
      
      animalMeta <- subset(metaPreg, animal_id == s)
      sampleID <- c(sampleID,animalMeta$sampleID)
      pregTF <- c(pregTF,animalMeta$pregStatusBackDate35)
      status <- c(status,animalMeta$status_dom_sub)
      animalID <- c(animalID,animalMeta$animal_id)
      
    } else {
      #print("nope")
    }
  }
  
  pregResample <- cbind.data.frame(sampleID,pregTF,status,animalID)
  
  ### remove samples where status is mis-matched
  statusMismatch <- c("SSUR1474","SSUR1605","SSUR1781","SSUR2229","SSUR1249","SSUR0749")
  pregResample <- pregResample[!pregResample$sampleID %in% statusMismatch,]
  
  ### Find the repeat animals S/D
  repeatSamples <- names(table(metaPreg$animal_id)[table(metaPreg$animal_id) > 1])
  
  ### find the repeat animals preg/not-preg
  s <- repeatSamples[1]
  doubleStatus <- vector()
  sampleID <- vector()
  pregTF <- vector()
  status <- vector()
  animalID <- vector()
  
  for (s in repeatSamples) {
    #print(s)
    if (length(unique(subset(metaStatus, animal_id == s)$status_dom_sub)) > 1) {
      #print("duplicate")
      doubleStatus <- c(doubleStatus,s)
      
      animalMeta <- subset(metaStatus, animal_id == s)
      sampleID <- c(sampleID,animalMeta$sampleID)
      pregTF <- c(pregTF,animalMeta$pregStatusBackDate35)
      status <- c(status,animalMeta$status_dom_sub)
      animalID <- c(animalID,animalMeta$animal_id)
      
    } else {
      #print("nope")
    }
  }
  
  statusResample <- cbind.data.frame(sampleID,pregTF,status,animalID)
  
  pregMismatch <- c("SSUR1400","SSUR1485","SSUR1474","SSUR1605","SSUR1781","SSUR2229","SSUR1862","SSUR1905")
  statusResample <- statusResample[!statusResample$sampleID %in% pregMismatch,]
  
  #gene <- geneList[223]
  gene <- "PYGB"
  
  for (gene in geneList) {
    #print(gene)
    ### Check that this code is truly making sure the samples match on the un-tested variable
    
    
    #get all repeats
    pregResampTTest <- pregResample
    pregResampTTest$geneResid <- 0

    #randomly pick one resample PREG
    for (a in unique(pregResampTTest$animalID)) {
      for (p in c(TRUE,FALSE)) {
        if (sum(pregResampTTest$animalID == a & pregResampTTest$pregTF == p) > 1) {
          #these are the samples that fit the category, just choose one, toss the rest)
          resamps <- length(which(pregResampTTest$animalID == a & pregResampTTest$pregTF == p))
          drop <- sample(which(pregResampTTest$animalID == a & pregResampTTest$pregTF == p), size = resamps - 1, replace = FALSE)
          pregResampTTest <- pregResampTTest[-drop,]
        }
      }
    }
    
    #get all repeats
    statResampTTest <- statusResample
    statResampTTest$geneResid <- 0

    #randomly pick one resample STATUS
    for (a in unique(statResampTTest$animalID)) {
      for (p in c("S","D")) {
        if (sum(statResampTTest$animalID == a & statResampTTest$status == p) > 1) {
          #these are the samples that fit the category, just choose one, toss the rest)
          resamps <- length(which(statResampTTest$animalID == a & statResampTTest$status == p))
          drop <- sample(which(statResampTTest$animalID == a & statResampTTest$status == p), size = resamps - 1, replace = FALSE)
          statResampTTest <- statResampTTest[-drop,]
        }
      }
    }

    for (s in pregResampTTest$sampleID) {
      pregResampTTest$geneResid[which(pregResampTTest$sampleID == s)] <- residsPreg[which(rownames(residsPreg) == gene),
                                                                              which(colnames(residsPreg) == s)]
    }
    
    for (s in statResampTTest$sampleID) {
      statResampTTest$geneResid[which(statResampTTest$sampleID == s)] <- residsStatus[which(rownames(residsStatus) == gene),
                                                                              which(colnames(residsStatus) == s)]
    }
    
    geneNamePlot <- c(geneNamePlot,gene)
    
    pregPval <- c(pregPval,emmremlResultsPreg$pval_preg[rownames(emmremlResultsPreg) == gene])
    statPval <- c(statPval,emmremlResultsStat$pval_status[rownames(emmremlResultsStat) == gene])
    
    meanGEpreg <- c(meanGEpreg,mean(subset(pregResampTTest, pregTF == TRUE)$geneResid))
    meanGEnonpreg <- c(meanGEnonpreg,mean(subset(pregResampTTest, pregTF == FALSE)$geneResid))
    
    meanGEdom <- c(meanGEdom,mean(subset(statResampTTest, status == "D")$geneResid))
    meanGEsub <- c(meanGEsub,mean(subset(statResampTTest, status == "S")$geneResid))
    
    ### Get emmreml data first, informs T-Test
    #preg emmreml
    emmremlBeta <- emmremlResultsPreg$beta_preg[rownames(emmremlResultsPreg) == gene]
    emmremlVar <- emmremlResultsPreg$var_beta_preg[rownames(emmremlResultsPreg) == gene]
    
    emmremlStdBetaPreg <- c(emmremlStdBetaPreg, emmremlBeta / sqrt(emmremlVar))
    emmremlPValPreg <- c(emmremlPValPreg, emmremlResultsPreg$pval_preg[rownames(emmremlResultsPreg) == gene])
    emmremlQValPreg <- c(emmremlQValPreg, emmremlResultsPreg$qval_preg[rownames(emmremlResultsPreg) == gene])
    
    #ttest preg
    if (emmremlBeta > 0) {
      testDir <- "greater"
    } else {
      testDir <- "less"
    }
    
    #wTestPreg <- wilcox.test(subset(pregResampTTest, pregTF == TRUE)$geneResid,
    #                     subset(pregResampTTest, pregTF == FALSE)$geneResid, paired = TRUE, alternative = testDir)
    tTestPreg <- t.test(subset(pregResampTTest, pregTF == TRUE)$geneResid,
                    subset(pregResampTTest, pregTF == FALSE)$geneResid, paired = TRUE, alternative = testDir)
    
    wTestPValPreg <- c(wTestPValPreg, tTestPreg$p.value)
    tTestStatPreg <- c(tTestStatPreg, tTestPreg$statistic)
    
    #status emmreml
    emmremlBeta <- emmremlResultsStat$beta_status[rownames(emmremlResultsStat) == gene]
    emmremlVar <- emmremlResultsStat$var_beta_status[rownames(emmremlResultsStat) == gene]
    
    emmremlStdBetaStatus <- c(emmremlStdBetaStatus, emmremlBeta / sqrt(emmremlVar))
    emmremlPValStatus <- c(emmremlPValStatus, emmremlResultsStat$pval_status[rownames(emmremlResultsStat) == gene])
    emmremlQValStatus <- c(emmremlQValStatus, emmremlResultsStat$fdr_status[rownames(emmremlResultsStat) == gene])
    
    #ttest status
    if (emmremlBeta > 0) {
      testDir <- "greater"
    } else {
      testDir <- "less"
    }
    
    #wTestStat <- wilcox.test(subset(statResampTTest, status == "D")$geneResid,
    #                     subset(statResampTTest, status == "S")$geneResid, paired = TRUE, alternative = testDir)
    
    tTestStat <- t.test(subset(statResampTTest, status == "D")$geneResid,
                    subset(statResampTTest, status == "S")$geneResid, paired = TRUE, alternative = testDir)
    
    wTestPValStatus <- c(wTestPValStatus, tTestStat$p.value)
    tTestStatStatus <- c(tTestStatStatus, tTestStat$statistic)

    treatment <- c(treatment,varA)
    
    if (geneByGenePlot) {
      
    }
  }
  
}

longSampleResults <- cbind.data.frame(geneNamePlot,
                                      pregPval,
                                      meanGEpreg,
                                      meanGEnonpreg,
                                      wTestPValPreg,
                                      tTestStatPreg,
                                      emmremlStdBetaPreg,
                                      emmremlPValPreg,
                                      emmremlQValPreg,
                                      statPval,
                                      meanGEdom,
                                      meanGEsub,
                                      wTestPValStatus,
                                      tTestStatStatus,
                                      emmremlStdBetaStatus,
                                      emmremlPValStatus,
                                      emmremlQValStatus,
                                      treatment)

longSampleResults$pregGEdiff <- longSampleResults$meanGEpreg - longSampleResults$meanGEnonpreg
longSampleResults$statusGEdiff <- longSampleResults$meanGEdom - longSampleResults$meanGEsub

#preg directional consistent
longSampleResults$pregDirConsistent <- (longSampleResults$pregGEdiff / longSampleResults$emmremlStdBetaPreg) > 0
longSampleResults$pregSigConsistent <- longSampleResults$pregDirConsistent & longSampleResults$wTestPValPreg < .05

#status directional consistent
longSampleResults$statusDirConsistent <- (longSampleResults$statusGEdiff / longSampleResults$emmremlStdBetaStatus) > 0
longSampleResults$statusSigConsistent <- longSampleResults$statusDirConsistent & longSampleResults$wTestPValStatus < .05

### already ran this, drop the final file in here
longSampleResults <- readRDS(file = paste0(outDir,"longitudinalSampleResults.RDS"))

## what amount sub-Qval .1 are directionally consistent
subset(longSampleResults, emmremlQValStatus < .1)$statusDirConsistent
subset(longSampleResults, emmremlQValPreg < .1)$pregDirConsistent

## what amount sub-Qval .1 are statistically consistent
subset(longSampleResults, emmremlQValStatus < .1)$statusSigConsistent



plotSigOnly <- TRUE
if (plotSigOnly) {
  longSampleResultsPlot <- subset(longSampleResults, emmremlQValPreg < .1)
} else {
  longSampleResultsPlot <- longSampleResults
}

fdrThresh <- .2

ggplot(longSampleResultsPlot) +
geom_histogram(aes(x = tTestStatPreg,
                   fill = ifelse(emmremlQValPreg < fdrThresh,
                                 ifelse(emmremlStdBetaPreg > 0,"sigUp","sigDn"),"n.s.")), bins = 50) +
labs(fill = "sig in\nemmreml")

ggplot(longSampleResultsPlot) +
geom_histogram(aes(x = meanGEpreg - meanGEnonpreg,
                   fill = ifelse(emmremlQValPreg < fdrThresh,
                                 ifelse(emmremlStdBetaPreg > 0,"sigUp","sigDn"),"n.s.")), bins = 50) +
labs(fill = "sig in\nemmreml")

if (plotSigOnly) {
  longSampleResultsPlot <- subset(longSampleResults, emmremlQValStatus < .1)
} else {
  longSampleResultsPlot <- longSampleResults
}


ggplot(longSampleResultsPlot) +
  geom_histogram(aes(x = tTestStatStatus,
                   fill = ifelse(emmremlQValStatus < fdrThresh,
                                 ifelse(emmremlStdBetaStatus > 0,"sigUp","sigDn"),"n.s.")), bins = 50) +
labs(fill = "sig in\nemmreml")
ggplot(longSampleResultsPlot) +
  geom_histogram(aes(x = meanGEdom - meanGEsub,
                   fill = ifelse(emmremlQValStatus < fdrThresh,
                                 ifelse(emmremlStdBetaStatus > 0,"sigUp","sigDn"),"n.s.")), bins = 50) +
labs(fill = "sig in\nemmreml")

#genes significant in status & significant t.test result
sigMatchTTest <- subset(longSampleResults, emmremlQValStatus < fdrThresh)[subset(longSampleResults, 
                                                                          emmremlQValStatus < fdrThresh)$statusSigConsistent,c(1,14,15,18)]

sigMatchTTest$trtBetaDir <- ""
sigMatchTTest$statusBetaDir <- ""

gardTrtEmmreml <- read.table(paste0("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_","Gard",
                                       "_10000genes_5lCPM_model","G","_basic+Preg+WeightwKmatrix_emmreml"))
lpsTrtEmmreml <- read.table(paste0("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_","LPS",
                                       "_10000genes_5lCPM_model","F","_basic+Preg+WeightwKmatrix_emmreml"))

for (row in 1:dim(sigMatchTTest)[1]) {
  if (sigMatchTTest$treatment[row] == "Gard") {
    trtCtrlEmreml <- gardTrtEmmreml
  } else {
    trtCtrlEmreml <- lpsTrtEmmreml
  }
  
  if ( trtCtrlEmreml$beta_treatment[which(rownames(trtCtrlEmreml) == sigMatchTTest$geneNamePlot[row])] > 0 ) {
    sigMatchTTest$trtBetaDir[row] <- "Pos"
  } else {
    sigMatchTTest$trtBetaDir[row] <- "Neg"
  }
}

sigMatchTTest$statusBetaDir <- ifelse(sigMatchTTest$emmremlStdBetaStatus > 0, "Pos", "Neg")

sigMatchTTest$groupFig3 <- ifelse(sigMatchTTest$trtBetaDir == "Pos" & sigMatchTTest$statusBetaDir == "Neg", "GroupI",
                                  ifelse(sigMatchTTest$trtBetaDir == "Pos" & sigMatchTTest$statusBetaDir == "Pos", "GroupII",
                                         ifelse(sigMatchTTest$trtBetaDir == "Neg" & sigMatchTTest$statusBetaDir == "Neg", "GroupIII","GroupIV")))

#check multiple hits
sigMatchTTest[sigMatchTTest$geneNamePlot %in% names(table(sigMatchTTest$geneNamePlot)[table(sigMatchTTest$geneNamePlot) > 1]),]

table(sigMatchTTest$treatment)
table(sigMatchTTest$groupFig3)

sigMatchTTestLPS <- subset(sigMatchTTest, treatment %in% c("LPS","Ctrl"))

### run the FET enrichment here as well
### Enrichment analysis, FET for presence/absence in hallmark pathways
fetPVal <- vector()
inPath <- vector()
pathTotal <- vector()
geneGrp <- vector()
oddsRatio <- vector()
pathNames <- vector()

pway <- "HALLMARK_TNFA_SIGNALING_VIA_NFKB"
geneGroup <- "GroupI"

gene.set <- read.table(paste0(dataDir,"hallmarkPathways_wCategories.csv"), sep = ",")
colnames(gene.set) <- c("Associated.Gene.Name","GO.Term.Accession","Category")

for (geneGroup in c("GroupI","GroupII","GroupIII","GroupIV")) {
  #print(geneGroup)
  groupGeneList <- subset(sigMatchTTestLPS, groupFig3 == geneGroup)$geneNamePlot
  nonGroupGeneList <- rownames(lpsTrtEmmreml)[!rownames(lpsTrtEmmreml) %in% groupGeneList]
  
  for (pway in unique(gene.set$GO.Term.Accession)) {
    #check for 2 genes
    pathwayGII <- length(intersect(groupGeneList,
                            unique(gene.set$Associated.Gene.Name[gene.set$GO.Term.Accession == pway])))
    #print(pathwayGII)
    if (pathwayGII >= 2) {
      nPathwayGII <- length(unique(gene.set$Associated.Gene.Name[gene.set$GO.Term.Accession == pway])) - pathwayGII
      nPathwayBck <- length(nonGroupGeneList) - nPathwayGII
      pathwayBck <- length(groupGeneList) - pathwayGII
      
      dat <- data.frame(
        "Non-Pathway" = c(pathwayBck, nPathwayBck),
        "Pathway" = c(pathwayGII, nPathwayGII),
        row.names = c(geneGroup, "Background"),
        stringsAsFactors = FALSE
      )
      colnames(dat) <- c("Non-Pathway", "Pathway")
      
      test <- fisher.test(dat, alternative = "less")
      test
      print(paste0(pway," ",test$p.value))
      if (test$p.value < .1) {
        #print(dat)
      }
      pathNames <- c(pathNames,pway)
      fetPVal <- c(fetPVal,test$p.value)
      oddsRatio <- c(oddsRatio,test$estimate)
      inPath <- c(inPath,pathwayGII)
      pathTotal <- c(pathTotal,nPathwayGII)
      geneGrp <- c(geneGrp,geneGroup)
    }
  }
}

fetSigResampleResults <- cbind.data.frame(geneGrp,pathNames,fetPVal,oddsRatio,inPath,pathTotal)
colnames(fetSigResampleResults)[2] <- "gseaHallmark"
fetSigResampleResults$gseaHallmark <- str_remove_all(fetSigResampleResults$gseaHallmark, pattern = "HALLMARK_")
#fetResults <- fetResults[order(fetResults$fetPVal),]

fetSigResampleResults <- fetSigResampleResults[order(fetSigResampleResults$geneGrp, fetSigResampleResults$fetPVal),]

subset(fetSigResampleResults, inPath > 1 & fetPVal < .1 & oddsRatio < 1)


#gene by gene plot
for (trt in c("LPS","Ctrl","Gard","Dex")) {
  lpsPlot <- subset(longSampleResults, treatment == trt & emmremlQValStatus < .1)
  
  lpsPlot <- lpsPlot[,colnames(lpsPlot) %in% c("emmremlStdBetaStatus","geneNamePlot","meanGEdom","meanGEsub")]
  
  colnames(lpsPlot)[1:3] <- c("geneName","D","S")
  
  lpsPlotUp <- subset(lpsPlot, emmremlStdBetaStatus > 0)
  lpsPlotDown <- subset(lpsPlot, emmremlStdBetaStatus < 0)
  
  lpsPlotUp <- melt(lpsPlotUp[,1:3])
  lpsPlotDown <- melt(lpsPlotDown[,1:3])
  
  lpsPlotUp$variable <- factor(lpsPlotUp$variable, levels = c("S","D"))
  lpsPlotDown$variable <- factor(lpsPlotDown$variable, levels = c("S","D"))
  
  ggplot(data = lpsPlotUp,
         aes(x = variable,
             y = value,
             group = geneName)) +
    geom_line(linewidth = .2, alpha = 80/dim(lpsPlotUp)[1]) +
    theme_minimal() + theme_bw() + 
    ylab("Mean GE in Long. Samples") + xlab("Status") +
    ggtitle("Resampled GE Values, Genes Sig Up in Dom F Meerkats",
            subtitle = paste0(trt,", Treatment, FDR < 0.10"))
  ggsave(paste0(figPaperDir,"3b_Resample_GeneValues_",trt,"_Up.png"), width = 6, height = 4)
  
  ggplot(data = lpsPlotDown,
         aes(x = variable,
             y = value,
             group = geneName)) +
    geom_line(linewidth = .2, alpha = 80/dim(lpsPlotDown)[1]) +
    theme_minimal() + theme_bw() + 
    ylab("Mean GE in Long. Samples") + xlab("Status") +
    ggtitle("Resampled GE Values, Genes Sig Down in Dom F Meerkats",
            subtitle = paste0(trt,", Treatment, FDR < 0.10"))
  ggsave(paste0(figPaperDir,"3b_Resample_GeneValues_",trt,"_Down.png"), width = 6, height = 4)
}


ggplot() + geom_histogram(aes(x = subset(longSampleResults, emmremlQValStatus < .1)$wTestPValStatus), bins = 50)

statusTtestPvals <- subset(longSampleResults, emmremlQValStatus < .1)$wTestPValStatus

qvalue(statusTtestPvals, pi0=1)$qvalues

### try a permutation test - 100x permutations of D-S labels?
#then calculate the p-value for each gene, and then that list of permuted p-values goes into a DF to use perm.fdr as a comparison with the true set

#1) pull out the sig genes from longSampleResults
geneByGenePlot <- TRUE
varA <- "LPS"

nPerms <- 100
nGenes <- dim(subset(longSampleResults, emmremlQValStatus < .1))[1]

permPvals <- matrix(NA, nrow = nGenes, ncol = nPerms)
permNumber <- 1
geneNum <- 1

wTestPValPreg <- vector()
tTestStatPreg <- vector()
wTestPValStatus <- vector()
tTestStatStatus <- vector()
emmremlStdBetaStatus <- vector()
emmremlPValStatus <- vector()
emmremlStdBetaPreg <- vector()
emmremlPValPreg <- vector()
emmremlQValPreg <- vector()
emmremlQValStatus <- vector()

#vectors for gene by gene line plot
geneNamePlot <- vector()
pregPval <- vector()
statPval <- vector()
meanGEpreg <- vector()
meanGEnonpreg <- vector()
meanGEdom <- vector()
meanGEsub <- vector()
treatment <- vector()


  
for (varA in c("LPS","Gard","Dex","Ctrl")) {
  #print(varA)
  if (varA %in% c("LPS","Ctrl")) {
    modelA <- "modelF"
  } else {
    modelA <- "modelG"
  }
  
  ### use the model without the repeated animals for the gene info
  emmremlResultsPreg <- read.table(paste0(outDir,"allBatches_F_",varA,"only_10000genes_5lCPM_pregReSamp_",
                                          modelA,"_basic+Preg+WeightwKmatrix_emmreml_wPermFDR_v1_preg"))
  emmremlResultsPreg$qval_preg <- qvalue(emmremlResultsPreg$pval_preg)$qvalue
  
  emmremlResultsStat <- read.table(paste0(outDir,"allBatches_F_",varA,"only_10000genes_5lCPM_statusReSamp_",
                                          modelA,"_basic+Preg+WeightwKmatrix_emmreml_wPermFDR_v1_status"))
  emmremlResultsStat$qval_status <- qvalue(emmremlResultsStat$pval_status)$qvalue
  
  residsStatus <- readRDS(paste0(outDir,"allBatches_F_",varA,"only_10000genes_5lCPM_",modelA,"_resids.RDS"))
  metaStatus <- readRDS(paste0(outDir,"allBatches_F_",varA,"only_10000genes_5lCPM_",modelA,"_dfMeta.RDS"))
  
  residsPreg <- readRDS(paste0(outDir,"allBatches_F_",varA,"only_10000genes_5lCPM_",modelA,"_resids.RDS"))
  metaPreg <- readRDS(paste0(outDir,"allBatches_F_",varA,"only_10000genes_5lCPM_",modelA,"_dfMeta.RDS"))
  
  metaPreg$preg <- ifelse(metaPreg$pregLact == "preg", TRUE, FALSE)
  metaStatus$preg <- ifelse(metaStatus$pregLact == "preg", TRUE, FALSE)
  
  metaPreg$statusPreg <- paste0(metaPreg$status_dom_sub,"_",metaPreg$preg)
  metaStatus$statusPreg <- paste0(metaStatus$status_dom_sub,"_",metaStatus$preg)
  
  ### stil using full data's resids and meta
  #geneList <- rownames(emmremlResultsPreg)
  
  ### use the intersection of genes from resample status, resample preg, and original
  geneList <- intersect(intersect(rownames(emmremlResultsPreg),rownames(subset(emmremlResultsStat, fdr_status < .1))), rownames(residsPreg))
  
  ### only use genes that are sig at FDR 0.10 emmremlResultsStat
  if (length(geneList) > 0) {
  
  repeatSamples <- names(table(metaPreg$animal_id)[table(metaPreg$animal_id) > 1])
  
  ### find the repeat animals preg/not-preg
  s <- repeatSamples[1]
  doublePreg <- vector()
  sampleID <- vector()
  pregTF <- vector()
  status <- vector()
  animalID <- vector()
  
  for (s in repeatSamples) {
    #print(s)
    if (length(unique(subset(metaPreg, animal_id == s)$pregStatusBackDate35)) > 1) {
      #print("duplicate")
      doublePreg <- c(doublePreg,s)
      
      animalMeta <- subset(metaPreg, animal_id == s)
      sampleID <- c(sampleID,animalMeta$sampleID)
      pregTF <- c(pregTF,animalMeta$pregStatusBackDate35)
      status <- c(status,animalMeta$status_dom_sub)
      animalID <- c(animalID,animalMeta$animal_id)
      
    } else {
      #print("nope")
    }
  }
  
  pregResample <- cbind.data.frame(sampleID,pregTF,status,animalID)
  
  ### remove samples where status is mis-matched
  statusMismatch <- c("SSUR1474","SSUR1605","SSUR1781","SSUR2229","SSUR1249","SSUR0749")
  pregResample <- pregResample[!pregResample$sampleID %in% statusMismatch,]
  
  ### Find the repeat animals S/D
  repeatSamples <- names(table(metaPreg$animal_id)[table(metaPreg$animal_id) > 1])
  
  ### find the repeat animals preg/not-preg
  s <- repeatSamples[1]
  doubleStatus <- vector()
  sampleID <- vector()
  pregTF <- vector()
  status <- vector()
  animalID <- vector()
  
  for (s in repeatSamples) {
    #print(s)
    if (length(unique(subset(metaStatus, animal_id == s)$status_dom_sub)) > 1) {
      #print("duplicate")
      doubleStatus <- c(doubleStatus,s)
      
      animalMeta <- subset(metaStatus, animal_id == s)
      sampleID <- c(sampleID,animalMeta$sampleID)
      pregTF <- c(pregTF,animalMeta$pregStatusBackDate35)
      status <- c(status,animalMeta$status_dom_sub)
      animalID <- c(animalID,animalMeta$animal_id)
      
    } else {
      #print("nope")
    }
  }
  
  statusResample <- cbind.data.frame(sampleID,pregTF,status,animalID)
  
  pregMismatch <- c("SSUR1400","SSUR1485","SSUR1474","SSUR1605","SSUR1781","SSUR2229","SSUR1862","SSUR1905")
  statusResample <- statusResample[!statusResample$sampleID %in% pregMismatch,]
  
  #gene <- geneList[223]
  gene <- "CCR7"
  #geneNum <- 1
  
  for (gene in geneList) {
    #print(gene)
    ### Check that this code is truly making sure the samples match on the un-tested variable
    
    
    #get all repeats
    pregResampTTest <- pregResample
    pregResampTTest$geneResid <- 0

    #randomly pick one resample PREG
    for (a in unique(pregResampTTest$animalID)) {
      for (p in c(TRUE,FALSE)) {
        if (sum(pregResampTTest$animalID == a & pregResampTTest$pregTF == p) > 1) {
          #these are the samples that fit the category, just choose one, toss the rest)
          resamps <- length(which(pregResampTTest$animalID == a & pregResampTTest$pregTF == p))
          drop <- sample(which(pregResampTTest$animalID == a & pregResampTTest$pregTF == p), size = resamps - 1, replace = FALSE)
          pregResampTTest <- pregResampTTest[-drop,]
        }
      }
    }
    
    #get all repeats
    statResampTTest <- statusResample
    statResampTTest$geneResid <- 0

    #randomly pick one resample STATUS
    for (a in unique(statResampTTest$animalID)) {
      for (p in c("S","D")) {
        if (sum(statResampTTest$animalID == a & statResampTTest$status == p) > 1) {
          #these are the samples that fit the category, just choose one, toss the rest)
          resamps <- length(which(statResampTTest$animalID == a & statResampTTest$status == p))
          drop <- sample(which(statResampTTest$animalID == a & statResampTTest$status == p), size = resamps - 1, replace = FALSE)
          statResampTTest <- statResampTTest[-drop,]
        }
      }
    }

    for (s in pregResampTTest$sampleID) {
      pregResampTTest$geneResid[which(pregResampTTest$sampleID == s)] <- residsPreg[which(rownames(residsPreg) == gene),
                                                                              which(colnames(residsPreg) == s)]
    }
    
    for (s in statResampTTest$sampleID) {
      statResampTTest$geneResid[which(statResampTTest$sampleID == s)] <- residsStatus[which(rownames(residsStatus) == gene),
                                                                              which(colnames(residsStatus) == s)]
    }
    
    geneNamePlot <- c(geneNamePlot,gene)
    
    pregPval <- c(pregPval,emmremlResultsPreg$pval_preg[rownames(emmremlResultsPreg) == gene])
    statPval <- c(statPval,emmremlResultsStat$pval_status[rownames(emmremlResultsStat) == gene])
    
    meanGEpreg <- c(meanGEpreg,mean(subset(pregResampTTest, pregTF == TRUE)$geneResid))
    meanGEnonpreg <- c(meanGEnonpreg,mean(subset(pregResampTTest, pregTF == FALSE)$geneResid))
    
    meanGEdom <- c(meanGEdom,mean(subset(statResampTTest, status == "D")$geneResid))
    meanGEsub <- c(meanGEsub,mean(subset(statResampTTest, status == "S")$geneResid))
    
    ### Get emmreml data first, informs T-Test
    #preg emmreml
    emmremlBeta <- emmremlResultsPreg$beta_preg[rownames(emmremlResultsPreg) == gene]
    emmremlVar <- emmremlResultsPreg$var_beta_preg[rownames(emmremlResultsPreg) == gene]
    
    emmremlStdBetaPreg <- c(emmremlStdBetaPreg, emmremlBeta / sqrt(emmremlVar))
    emmremlPValPreg <- c(emmremlPValPreg, emmremlResultsPreg$pval_preg[rownames(emmremlResultsPreg) == gene])
    emmremlQValPreg <- c(emmremlQValPreg, emmremlResultsPreg$qval_preg[rownames(emmremlResultsPreg) == gene])
    
    #ttest preg
    if (emmremlBeta > 0) {
      testDir <- "greater"
    } else {
      testDir <- "less"
    }
    
    #wTestPreg <- wilcox.test(subset(pregResampTTest, pregTF == TRUE)$geneResid,
    #                     subset(pregResampTTest, pregTF == FALSE)$geneResid, paired = TRUE, alternative = testDir)
    tTestPreg <- t.test(subset(pregResampTTest, pregTF == TRUE)$geneResid,
                    subset(pregResampTTest, pregTF == FALSE)$geneResid, paired = TRUE, alternative = testDir)
    
    wTestPValPreg <- c(wTestPValPreg, tTestPreg$p.value)
    tTestStatPreg <- c(tTestStatPreg, tTestPreg$statistic)
    
    #status emmreml
    emmremlBeta <- emmremlResultsStat$beta_status[rownames(emmremlResultsStat) == gene]
    emmremlVar <- emmremlResultsStat$var_beta_status[rownames(emmremlResultsStat) == gene]
    
    emmremlStdBetaStatus <- c(emmremlStdBetaStatus, emmremlBeta / sqrt(emmremlVar))
    emmremlPValStatus <- c(emmremlPValStatus, emmremlResultsStat$pval_status[rownames(emmremlResultsStat) == gene])
    emmremlQValStatus <- c(emmremlQValStatus, emmremlResultsStat$fdr_status[rownames(emmremlResultsStat) == gene])
    
    #ttest status
    if (emmremlBeta > 0) {
      testDir <- "greater"
    } else {
      testDir <- "less"
    }
    
    #wTestStat <- wilcox.test(subset(statResampTTest, status == "D")$geneResid,
    #                     subset(statResampTTest, status == "S")$geneResid, paired = TRUE, alternative = testDir)
    
    tTestStat <- t.test(subset(statResampTTest, status == "D")$geneResid,
                    subset(statResampTTest, status == "S")$geneResid, paired = TRUE, alternative = testDir)
    
    wTestPValStatus <- c(wTestPValStatus, tTestStat$p.value)
    tTestStatStatus <- c(tTestStatStatus, tTestStat$statistic)

    treatment <- c(treatment,varA)
    
    
    for (permNumber in 1:nPerms) {
      ## shuffle and generate new p-value
      if (dim(statResampTTest)[1] == 10) {
        rowOrder <- c(sample(c(1,2)),
                      sample(c(3,4)),
                      sample(c(5,6)),
                      sample(c(7,8)),
                      sample(c(9,10)))
      } else {
                rowOrder <- c(sample(c(1,2)),
                      sample(c(3,4)),
                      sample(c(5,6)),
                      sample(c(7,8)),
                      sample(c(9,10)),
                      sample(c(11,12)))
                    }
      
      statResampTTestPerm <- statResampTTest
      statResampTTestPerm$status <- statResampTTestPerm$status[rowOrder]
      
      tTestStatPerm <- t.test(subset(statResampTTestPerm, status == "D")$geneResid,
                              subset(statResampTTestPerm, status == "S")$geneResid, paired = TRUE, alternative = testDir)
      
      permPvals[geneNum,permNumber] <- tTestStatPerm$p.value
    }
    
    geneNum <- geneNum + 1
  }
  
  }
}

longSampleResults <- cbind.data.frame(geneNamePlot,
                                      pregPval,
                                      meanGEpreg,
                                      meanGEnonpreg,
                                      wTestPValPreg,
                                      tTestStatPreg,
                                      emmremlStdBetaPreg,
                                      emmremlPValPreg,
                                      emmremlQValPreg,
                                      statPval,
                                      meanGEdom,
                                      meanGEsub,
                                      wTestPValStatus,
                                      tTestStatStatus,
                                      emmremlStdBetaStatus,
                                      emmremlPValStatus,
                                      emmremlQValStatus,
                                      treatment)

longSampleResults$pregGEdiff <- longSampleResults$meanGEpreg - longSampleResults$meanGEnonpreg
longSampleResults$statusGEdiff <- longSampleResults$meanGEdom - longSampleResults$meanGEsub

#preg directional consistent
longSampleResults$pregDirConsistent <- (longSampleResults$pregGEdiff / longSampleResults$emmremlStdBetaPreg) > 0
longSampleResults$pregSigConsistent <- longSampleResults$pregDirConsistent & longSampleResults$wTestPValPreg < .05

#status directional consistent
longSampleResults$statusDirConsistent <- (longSampleResults$statusGEdiff / longSampleResults$emmremlStdBetaStatus) > 0
longSampleResults$statusSigConsistent <- longSampleResults$statusDirConsistent & longSampleResults$wTestPValStatus < .05

longSampleResultsStatusSig <- subset(longSampleResults, emmremlQValStatus < .1)

#use perm.fdr
longSampleResultsStatusSigFDR <- perm.fdr(input_df = longSampleResultsStatusSig, Pvals_col_name = "wTestPValStatus", perm_df = permPvals, name = "pairedTTest")

ggplot() + geom_histogram(aes(as.vector(permPvals)), bins = 50) +
  ggtitle("All T-Test Permuted P-values (100x perms)") + xlim(c(-0.05,1.05))

ggplot() + geom_histogram(aes(longSampleResultsStatusSigFDR$wTestPValStatus), bins = 50) +
  ggtitle("Original T-test P-values") + xlim(c(-0.05,1.05))

ggplot() + geom_histogram(aes(longSampleResultsStatusSigFDR$fdr_pairedTTest), bins = 50) +
  ggtitle("perm.fdr results (Sanz script)") + xlim(c(-0.05,1.05))

sum(longSampleResultsStatusSigFDR$fdr_pairedTTest < .1)

#save to output
saveRDS(object = longSampleResults, 
        file = paste0(outDir,"longitudinalSampleResults.RDS"))

ggplot() + 
  geom_histogram(aes(p.adjust(longSampleResultsStatusSig$wTestPValStatus, 
                          method = "BH", n = length(longSampleResultsStatusSig$wTestPValStatus))), bins = 50) +
  ggtitle("Benjamini-Hochberg adj p-values") + xlim(c(-0.05,1.05))



##### FET for upregulated genes - confirmed in resample
upGenes <- subset(longSampleResults, wTestPValStatus < .05 & emmremlStdBetaStatus > 0)$geneNamePlot

dat <- data.frame(
  "Non-Pathway" = c(6018, 5),
  "Pathway" = c(198, 2),
  row.names = c(geneGroup, "Background"),
  stringsAsFactors = FALSE
)
colnames(dat) <- c("Non-Pathway", "Pathway")

test <- fisher.test(dat, alternative = "greater")
test


```

### what does the change in GE in animals who were sampled twice look like?

```{r check status and preg change in sub sub animals}
geneByGenePlot <- TRUE
varA <- "LPS"

wTestPValPreg <- vector()
tTestStatPreg <- vector()
wTestPValStatus <- vector()
tTestStatStatus <- vector()
emmremlStdBetaStatus <- vector()
emmremlPValStatus <- vector()
emmremlStdBetaPreg <- vector()
emmremlPValPreg <- vector()
emmremlQValPreg <- vector()
emmremlQValStatus <- vector()

#vectors for gene by gene line plot
geneNamePlot <- vector()
pregPval <- vector()
statPval <- vector()
meanGEpreg <- vector()
meanGEnonpreg <- vector()
meanGE1st <- vector()
meanGE2nd <- vector()
treatment <- vector()

for (varA in c("LPS","Gard","Dex","Ctrl")) {
  #print(varA)
  if (varA %in% c("LPS","Ctrl")) {
    modelA <- "modelF"
  } else {
    modelA <- "modelG"
  }
  
  ### use the model without the repeated animals for the gene info
  emmremlResultsPreg <- read.table(paste0(outDir,"allBatches_F_",varA,"only_10000genes_5lCPM_pregReSamp_",
                                          modelA,"_basic+Preg+WeightwKmatrix_emmreml_wPermFDR_v1_preg"))
  emmremlResultsPreg$qval_preg <- qvalue(emmremlResultsPreg$pval_preg)$qvalue
  
  emmremlResultsStat <- read.table(paste0(outDir,"allBatches_F_",varA,"only_10000genes_5lCPM_statusReSamp_",
                                          modelA,"_basic+Preg+WeightwKmatrix_emmreml_wPermFDR_v1_status"))
  emmremlResultsStat$qval_status <- qvalue(emmremlResultsStat$pval_status)$qvalue
  
  residsStatus <- readRDS(paste0(outDir,"allBatches_F_",varA,"only_10000genes_5lCPM_",modelA,"_resids.RDS"))
  metaStatus <- readRDS(paste0(outDir,"allBatches_F_",varA,"only_10000genes_5lCPM_",modelA,"_dfMeta.RDS"))
  
  residsPreg <- readRDS(paste0(outDir,"allBatches_F_",varA,"only_10000genes_5lCPM_",modelA,"_resids.RDS"))
  metaPreg <- readRDS(paste0(outDir,"allBatches_F_",varA,"only_10000genes_5lCPM_",modelA,"_dfMeta.RDS"))
  
  metaPreg$preg <- ifelse(metaPreg$pregLact == "preg", TRUE, FALSE)
  metaStatus$preg <- ifelse(metaStatus$pregLact == "preg", TRUE, FALSE)
  
  metaPreg$statusPreg <- paste0(metaPreg$status_dom_sub,"_",metaPreg$preg)
  metaStatus$statusPreg <- paste0(metaStatus$status_dom_sub,"_",metaStatus$preg)
  
  ### stil using full data's resids and meta
  #geneList <- rownames(emmremlResultsPreg)
  
  ### use the intersection of genes from resample status, resample preg, and original
  geneList <- intersect(intersect(rownames(emmremlResultsPreg),rownames(emmremlResultsStat)), rownames(residsPreg))
  
  repeatSamples <- names(table(metaPreg$animal_id)[table(metaPreg$animal_id) > 1])
  
  ### find the repeat animals preg/not-preg
  s <- repeatSamples[1]
  doublePreg <- vector()
  sampleID <- vector()
  pregTF <- vector()
  status <- vector()
  animalID <- vector()
  
  for (s in repeatSamples) {
    #print(s)
    if (length(unique(subset(metaPreg, animal_id == s)$pregStatusBackDate35)) > 1) {
      #print("duplicate")
      doublePreg <- c(doublePreg,s)
      
      animalMeta <- subset(metaPreg, animal_id == s)
      sampleID <- c(sampleID,animalMeta$sampleID)
      pregTF <- c(pregTF,animalMeta$pregStatusBackDate35)
      status <- c(status,animalMeta$status_dom_sub)
      animalID <- c(animalID,animalMeta$animal_id)
      
    } else {
      #print("nope")
    }
  }
  
  pregResample <- cbind.data.frame(sampleID,pregTF,status,animalID)
  
  ### remove samples where status is mis-matched
  statusMismatch <- c("SSUR1474","SSUR1605","SSUR1781","SSUR2229","SSUR1249","SSUR0749")
  pregResample <- pregResample[!pregResample$sampleID %in% statusMismatch,]
  
  ### Find the repeat animals S/D
  repeatSamples <- names(table(metaPreg$animal_id)[table(metaPreg$animal_id) > 1])
  
  ### find the repeat animals preg/not-preg
  s <- repeatSamples[1]
  doubleStatus <- vector()
  sampleID <- vector()
  pregTF <- vector()
  status <- vector()
  animalID <- vector()
  
  for (s in repeatSamples) {
    #if there are more samples than statuses, then we're guaranteed a repeat status
    if (length(subset(metaStatus, animal_id == s)$status_dom_sub) > length(unique(subset(metaStatus, animal_id == s)$status_dom_sub))) {
      #print("duplicate")
      doubleStatus <- c(doubleStatus,s)
      
      animalMeta <- subset(metaStatus, animal_id == s)
      sampleID <- c(sampleID,animalMeta$sampleID)
      pregTF <- c(pregTF,animalMeta$pregStatusBackDate35)
      status <- c(status,animalMeta$status_dom_sub)
      animalID <- c(animalID,animalMeta$animal_id)
      
    } else {
      #print("nope")
    }
  }
  
  statusResample <- cbind.data.frame(sampleID,pregTF,status,animalID)
  
  pregMismatch <- c("SSUR2522","SSUR1284","SSUR1862","SSUR2301","SSUR2372","SSUR1312","SSUR2402","SSUR1474","SSUR1249")
  statusResample <- statusResample[!statusResample$sampleID %in% pregMismatch,]
  
  ## drop animals with only one sample left
  statusResample <- statusResample[!statusResample$animalID %in% names(table(statusResample$animalID)[table(statusResample$animalID) == 1]),]
  
  ## drop the "odd status out" (S, S, D, drop the "D")
  oddStatusOut <- c("SSUR0749","SSUR1407","SSUR1769","SSUR1761","SSUR1897","SSUR2088")
  statusResample <- statusResample[!statusResample$sampleID %in% oddStatusOut,]

  ### limit to only subordinate animals sampled as longitudinal controls
  subAndLongSamples <- subset(metaStatus, sampleID %in% statusResample$sampleID & collectionRationale == "longitudinal" & status_dom_sub == "S")$sampleID
  statusResample <- subset(statusResample, sampleID %in% subAndLongSamples)
  
  #need to have 2 samples remaining
  multiSampleAnimals <- names(table(statusResample$animalID)[table(statusResample$animalID) > 1])
  statusResample <- subset(statusResample, animalID %in% multiSampleAnimals)
  
  ## randomly pick 5 animals
  animalsToUse <- sample(unique(statusResample$animalID), 5)
  statusResample <- subset(statusResample, animalID %in% animalsToUse)
  
  #gene <- geneList[223]
  gene <- "PYGB"
  
  for (gene in geneList) {
    #print(gene)
    ### Check that this code is truly making sure the samples match on the un-tested variable
    
    
    #get all repeats
    pregResampTTest <- pregResample
    pregResampTTest$geneResid <- 0

    #randomly pick one resample PREG
    for (a in unique(pregResampTTest$animalID)) {
      for (p in c(TRUE,FALSE)) {
        if (sum(pregResampTTest$animalID == a & pregResampTTest$pregTF == p) > 1) {
          #these are the samples that fit the category, just choose one, toss the rest)
          resamps <- length(which(pregResampTTest$animalID == a & pregResampTTest$pregTF == p))
          drop <- sample(which(pregResampTTest$animalID == a & pregResampTTest$pregTF == p), size = resamps - 1, replace = FALSE)
          pregResampTTest <- pregResampTTest[-drop,]
        }
      }
    }
    
    #get all repeats
    statResampTTest <- statusResample
    statResampTTest$geneResid <- 0

    #randomly pick one resample STATUS
    for (a in unique(statResampTTest$animalID)) {
      if (sum(statResampTTest$animalID == a) > 2) {
        #these are the samples that fit the category, just choose one, toss the rest)
          resamps <- length(which(statResampTTest$animalID == a))
          drop <- sample(which(statResampTTest$animalID == a), size = resamps - 2, replace = FALSE)
          statResampTTest <- statResampTTest[-drop,]
        }
    }

    for (s in pregResampTTest$sampleID) {
      pregResampTTest$geneResid[which(pregResampTTest$sampleID == s)] <- residsPreg[which(rownames(residsPreg) == gene),
                                                                              which(colnames(residsPreg) == s)]
    }
    
    for (s in statResampTTest$sampleID) {
      statResampTTest$geneResid[which(statResampTTest$sampleID == s)] <- residsStatus[which(rownames(residsStatus) == gene),
                                                                              which(colnames(residsStatus) == s)]
    }
    
    geneNamePlot <- c(geneNamePlot,gene)
    
    pregPval <- c(pregPval,emmremlResultsPreg$pval_preg[rownames(emmremlResultsPreg) == gene])
    statPval <- c(statPval,emmremlResultsStat$pval_status[rownames(emmremlResultsStat) == gene])
    
    meanGEpreg <- c(meanGEpreg,mean(subset(pregResampTTest, pregTF == TRUE)$geneResid))
    meanGEnonpreg <- c(meanGEnonpreg,mean(subset(pregResampTTest, pregTF == FALSE)$geneResid))
    
    meanGE1st <- c(meanGE1st,mean(statResampTTest$geneResid[seq(1,dim(statResampTTest)[1],2)]))
    meanGE2nd <- c(meanGE2nd,mean(statResampTTest$geneResid[seq(2,dim(statResampTTest)[1],2)]))
    
    ### Get emmreml data first, informs T-Test
    #preg emmreml
    emmremlBeta <- emmremlResultsPreg$beta_preg[rownames(emmremlResultsPreg) == gene]
    emmremlVar <- emmremlResultsPreg$var_beta_preg[rownames(emmremlResultsPreg) == gene]
    
    emmremlStdBetaPreg <- c(emmremlStdBetaPreg, emmremlBeta / sqrt(emmremlVar))
    emmremlPValPreg <- c(emmremlPValPreg, emmremlResultsPreg$pval_preg[rownames(emmremlResultsPreg) == gene])
    emmremlQValPreg <- c(emmremlQValPreg, emmremlResultsPreg$qval_preg[rownames(emmremlResultsPreg) == gene])
    
    #ttest preg
    if (emmremlBeta > 0) {
      testDir <- "greater"
    } else {
      testDir <- "less"
    }
    
    #wTestPreg <- wilcox.test(subset(pregResampTTest, pregTF == TRUE)$geneResid,
    #                     subset(pregResampTTest, pregTF == FALSE)$geneResid, paired = TRUE, alternative = testDir)
    tTestPreg <- t.test(subset(pregResampTTest, pregTF == TRUE)$geneResid,
                    subset(pregResampTTest, pregTF == FALSE)$geneResid, paired = TRUE, alternative = testDir)
    
    wTestPValPreg <- c(wTestPValPreg, tTestPreg$p.value)
    tTestStatPreg <- c(tTestStatPreg, tTestPreg$statistic)
    
    #status emmreml
    emmremlBeta <- emmremlResultsStat$beta_status[rownames(emmremlResultsStat) == gene]
    emmremlVar <- emmremlResultsStat$var_beta_status[rownames(emmremlResultsStat) == gene]
    
    emmremlStdBetaStatus <- c(emmremlStdBetaStatus, emmremlBeta / sqrt(emmremlVar))
    emmremlPValStatus <- c(emmremlPValStatus, emmremlResultsStat$pval_status[rownames(emmremlResultsStat) == gene])
    emmremlQValStatus <- c(emmremlQValStatus, emmremlResultsStat$fdr_status[rownames(emmremlResultsStat) == gene])
    
    #ttest status
    if (emmremlBeta > 0) {
      testDir <- "greater"
    } else {
      testDir <- "less"
    }
    
    #tTestStat <- t.test(statResampTTest$geneResid[seq(2,dim(statResampTTest)[1],2)],
    #                    statResampTTest$geneResid[seq(1,dim(statResampTTest)[1],2)], paired = TRUE, alternative = testDir)
    
    ### no assumption here that genes would go up or down with time
    tTestStat <- t.test(statResampTTest$geneResid[seq(2,dim(statResampTTest)[1],2)],
                        statResampTTest$geneResid[seq(1,dim(statResampTTest)[1],2)], paired = TRUE)
    
    wTestPValStatus <- c(wTestPValStatus, tTestStat$p.value)
    tTestStatStatus <- c(tTestStatStatus, tTestStat$statistic)

    treatment <- c(treatment,varA)
    
    if (geneByGenePlot) {
      
    }
  }
}

longSampleResultsCtrl <- cbind.data.frame(geneNamePlot,
                                      pregPval,
                                      meanGEpreg,
                                      meanGEnonpreg,
                                      wTestPValPreg,
                                      tTestStatPreg,
                                      emmremlStdBetaPreg,
                                      emmremlPValPreg,
                                      emmremlQValPreg,
                                      statPval,
                                      meanGE1st,
                                      meanGE2nd,
                                      wTestPValStatus,
                                      tTestStatStatus,
                                      emmremlStdBetaStatus,
                                      emmremlPValStatus,
                                      emmremlQValStatus,
                                      treatment)

longSampleResultsCtrl$statusGEdiff <- longSampleResultsCtrl$meanGE2nd - longSampleResultsCtrl$meanGE1st

#preg directional consistent
#longSampleResultsCtrl$pregDirConsistent <- (longSampleResultsCtrl$pregGEdiff / longSampleResultsCtrl$emmremlStdBetaPreg) > 0
#longSampleResultsCtrl$pregSigConsistent <- longSampleResultsCtrl$pregDirConsistent & longSampleResultsCtrl$wTestPValPreg < .05

#status directional consistent
longSampleResultsCtrl$statusDirConsistent <- (longSampleResultsCtrl$statusGEdiff / longSampleResultsCtrl$emmremlStdBetaStatus) > 0
longSampleResultsCtrl$statusSigConsistent <- longSampleResultsCtrl$statusDirConsistent & longSampleResultsCtrl$wTestPValStatus < .05

## what amount sub-Qval .1 are directionally consistent
subset(longSampleResultsCtrl, emmremlQValStatus < .1)$statusDirConsistent
#subset(longSampleResultsCtrl, emmremlQValPreg < .1)$pregDirConsistent

## what amount sub-Qval .1 are statistically consistent
subset(longSampleResultsCtrl, emmremlQValStatus < .1)$statusSigConsistent


#### Save this specific run of longitudinal sample results (control version)
saveRDS(object = longSampleResultsCtrl, file = paste0(outDir,"longitudinalSampleResultsCtrl.RDS"))



plotSigOnly <- TRUE
if (plotSigOnly) {
  longSampleResultsCtrlPlot <- subset(longSampleResultsCtrl, emmremlQValPreg < .1)
} else {
  longSampleResultsCtrlPlot <- longSampleResultsCtrl
}

fdrThresh <- .2

ggplot(longSampleResultsCtrlPlot) +
geom_histogram(aes(x = tTestStatPreg,
                   fill = ifelse(emmremlQValPreg < fdrThresh,
                                 ifelse(emmremlStdBetaPreg > 0,"sigUp","sigDn"),"n.s.")), bins = 50) +
labs(fill = "sig in\nemmreml")

ggplot(longSampleResultsCtrlPlot) +
geom_histogram(aes(x = meanGEpreg - meanGEnonpreg,
                   fill = ifelse(emmremlQValPreg < fdrThresh,
                                 ifelse(emmremlStdBetaPreg > 0,"sigUp","sigDn"),"n.s.")), bins = 50) +
labs(fill = "sig in\nemmreml")

if (plotSigOnly) {
  longSampleResultsCtrlPlot <- subset(longSampleResultsCtrl, emmremlQValStatus < .1)
} else {
  longSampleResultsCtrlPlot <- longSampleResultsCtrl
}


ggplot(longSampleResultsCtrlPlot) +
  geom_histogram(aes(x = tTestStatStatus,
                   fill = ifelse(emmremlQValStatus < fdrThresh,
                                 ifelse(emmremlStdBetaStatus > 0,"sigUp","sigDn"),"n.s.")), bins = 50) +
labs(fill = "sig in\nemmreml")

# ggplot(longSampleResultsCtrlPlot) +
#   geom_histogram(aes(x = meanGE2nd - meanGE1st,
#                    fill = ifelse(emmremlQValStatus < fdrThresh,
#                                  ifelse(emmremlStdBetaStatus > 0,"sigUp","sigDn"),"n.s.")), bins = 50) +
# labs(fill = "sig in\nemmreml")
# 

ggplot(longSampleResultsCtrlPlot) +
  geom_point(aes(x = emmremlStdBetaStatus,
                 y = tTestStatStatus,
                   col = wTestPValStatus < 0.05)) +
labs(col = "sig in\nT-Test")

ggplot(data = longSampleResultsCtrlPlot,
       aes(x = emmremlStdBetaStatus,
           y = tTestStatStatus)) +
  geom_hline(yintercept = 0) + 
  geom_vline(xintercept = 0) +
  ylim(c(-11,11)) +
  geom_point(size = 2, col = "gray80") + theme_minimal() + theme_bw() +
  theme(axis.text = element_text(size = 16),
        axis.title = element_text(size = 14)) +
  ylab("Longitudinal Sample T-Test") +
  xlab("Population Sample Status Beta")
ggsave(paste0(figPaperDir,"4a_longResample_inSub_ttest_v_emreml_statusEffect.png"), width = 6, height = 4)

##### 

#genes significant in status & significant t.test result
sigMatchTTest <- subset(longSampleResults, emmremlQValStatus < fdrThresh)[subset(longSampleResults, 
                                                                          emmremlQValStatus < fdrThresh)$statusSigConsistent,c(1,14,15,18)]

sigMatchTTest$trtBetaDir <- ""
sigMatchTTest$statusBetaDir <- ""

gardTrtEmmreml <- read.table(paste0("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_","Gard",
                                       "_10000genes_5lCPM_model","G","_basic+Preg+WeightwKmatrix_emmreml"))
lpsTrtEmmreml <- read.table(paste0("~/Dropbox (Personal)/meerkats/RNA/meerkatGE/output/allBatches_F_","LPS",
                                       "_10000genes_5lCPM_model","F","_basic+Preg+WeightwKmatrix_emmreml"))

for (row in 1:dim(sigMatchTTest)[1]) {
  if (sigMatchTTest$treatment[row] == "Gard") {
    trtCtrlEmreml <- gardTrtEmmreml
  } else {
    trtCtrlEmreml <- lpsTrtEmmreml
  }
  
  if ( trtCtrlEmreml$beta_treatment[which(rownames(trtCtrlEmreml) == sigMatchTTest$geneNamePlot[row])] > 0 ) {
    sigMatchTTest$trtBetaDir[row] <- "Pos"
  } else {
    sigMatchTTest$trtBetaDir[row] <- "Neg"
  }
}

sigMatchTTest$statusBetaDir <- ifelse(sigMatchTTest$emmremlStdBetaStatus > 0, "Pos", "Neg")

sigMatchTTest$groupFig3 <- ifelse(sigMatchTTest$trtBetaDir == "Pos" & sigMatchTTest$statusBetaDir == "Neg", "GroupI",
                                  ifelse(sigMatchTTest$trtBetaDir == "Pos" & sigMatchTTest$statusBetaDir == "Pos", "GroupII",
                                         ifelse(sigMatchTTest$trtBetaDir == "Neg" & sigMatchTTest$statusBetaDir == "Neg", "GroupIII","GroupIV")))

#check multiple hits
sigMatchTTest[sigMatchTTest$geneNamePlot %in% names(table(sigMatchTTest$geneNamePlot)[table(sigMatchTTest$geneNamePlot) > 1]),]

table(sigMatchTTest$treatment)
table(sigMatchTTest$groupFig3)

sigMatchTTestLPS <- subset(sigMatchTTest, treatment %in% c("LPS","Ctrl"))

### run the FET enrichment here as well
### Enrichment analysis, FET for presence/absence in hallmark pathways
fetPVal <- vector()
inPath <- vector()
pathTotal <- vector()
geneGrp <- vector()
oddsRatio <- vector()
pathNames <- vector()

pway <- "HALLMARK_TNFA_SIGNALING_VIA_NFKB"
geneGroup <- "GroupI"

gene.set <- read.table(paste0(dataDir,"hallmarkPathways_wCategories.csv"), sep = ",")
colnames(gene.set) <- c("Associated.Gene.Name","GO.Term.Accession","Category")

for (geneGroup in c("GroupI","GroupII","GroupIII","GroupIV")) {
  #print(geneGroup)
  groupGeneList <- subset(sigMatchTTestLPS, groupFig3 == geneGroup)$geneNamePlot
  nonGroupGeneList <- rownames(lpsTrtEmmreml)[!rownames(lpsTrtEmmreml) %in% groupGeneList]
  
  for (pway in unique(gene.set$GO.Term.Accession)) {
    #check for 2 genes
    pathwayGII <- length(intersect(groupGeneList,
                            unique(gene.set$Associated.Gene.Name[gene.set$GO.Term.Accession == pway])))
    #print(pathwayGII)
    if (pathwayGII >= 2) {
      nPathwayGII <- length(unique(gene.set$Associated.Gene.Name[gene.set$GO.Term.Accession == pway])) - pathwayGII
      nPathwayBck <- length(nonGroupGeneList) - nPathwayGII
      pathwayBck <- length(groupGeneList) - pathwayGII
      
      dat <- data.frame(
        "Non-Pathway" = c(pathwayBck, nPathwayBck),
        "Pathway" = c(pathwayGII, nPathwayGII),
        row.names = c(geneGroup, "Background"),
        stringsAsFactors = FALSE
      )
      colnames(dat) <- c("Non-Pathway", "Pathway")
      
      test <- fisher.test(dat, alternative = "less")
      test
      print(paste0(pway," ",test$p.value))
      if (test$p.value < .1) {
        #print(dat)
      }
      pathNames <- c(pathNames,pway)
      fetPVal <- c(fetPVal,test$p.value)
      oddsRatio <- c(oddsRatio,test$estimate)
      inPath <- c(inPath,pathwayGII)
      pathTotal <- c(pathTotal,nPathwayGII)
      geneGrp <- c(geneGrp,geneGroup)
    }
  }
}

fetSigResampleResults <- cbind.data.frame(geneGrp,pathNames,fetPVal,oddsRatio,inPath,pathTotal)
colnames(fetSigResampleResults)[2] <- "gseaHallmark"
fetSigResampleResults$gseaHallmark <- str_remove_all(fetSigResampleResults$gseaHallmark, pattern = "HALLMARK_")
#fetResults <- fetResults[order(fetResults$fetPVal),]

fetSigResampleResults <- fetSigResampleResults[order(fetSigResampleResults$geneGrp, fetSigResampleResults$fetPVal),]

subset(fetSigResampleResults, inPath > 1 & fetPVal < .1 & oddsRatio < 1)


#gene by gene plot
for (trt in c("LPS","Ctrl","Gard","Dex")) {
  lpsPlot <- subset(longSampleResults, treatment == trt & emmremlQValStatus < .1)
  
  lpsPlot <- lpsPlot[,colnames(lpsPlot) %in% c("emmremlStdBetaStatus","geneNamePlot","meanGE2nd","meanGE1st")]
  
  colnames(lpsPlot)[1:3] <- c("geneName","2ndMeasure","1stMeasure")
  
  lpsPlotUp <- subset(lpsPlot, emmremlStdBetaStatus > 0)
  lpsPlotDown <- subset(lpsPlot, emmremlStdBetaStatus < 0)
  
  lpsPlotUp <- melt(lpsPlotUp[,1:3])
  lpsPlotDown <- melt(lpsPlotDown[,1:3])
  
  lpsPlotUp$variable <- factor(lpsPlotUp$variable, levels = c("1stMeasure","2ndMeasure"))
  lpsPlotDown$variable <- factor(lpsPlotDown$variable, levels = c("1stMeasure","2ndMeasure"))
  
  ggplot(data = lpsPlotUp,
         aes(x = variable,
             y = value,
             group = geneName)) +
    geom_line(linewidth = .2, alpha = 80/dim(lpsPlotUp)[1]) +
    theme_minimal() + theme_bw() + 
    ylab("Mean GE in Long. Samples") + xlab("Status") +
    ggtitle("Resampled NoTransition GE Values, Genes Sig Up in Dom F Meerkats",
            subtitle = paste0(trt,", Treatment, FDR < 0.10"))
  ggsave(paste0(figPaperDir,"3b_ResampleNoTransition_GeneValues_",trt,"_Up.png"), width = 6, height = 4)
  
  ggplot(data = lpsPlotDown,
         aes(x = variable,
             y = value,
             group = geneName)) +
    geom_line(linewidth = .2, alpha = 80/dim(lpsPlotDown)[1]) +
    theme_minimal() + theme_bw() + 
    ylab("Mean GE in Long. Samples") + xlab("Status") +
    ggtitle("Resampled NoTransition GE Values, Genes Sig Down in Dom F Meerkats",
            subtitle = paste0(trt,", Treatment, FDR < 0.10"))
  ggsave(paste0(figPaperDir,"3b_ResampleNoTransition_GeneValues_",trt,"_Down.png"), width = 6, height = 4)
}


# text to add to paper:
paste0("In the same set of genes with expression significantly associated with status, only ",
       sum(subset(longSampleResults, emmremlQValStatus < .1)$wTestPValStatus < .05),
       " genes had significantly differences in the 1st and 2nd measurement across meerkats with no change in status, or ",
       mean(subset(longSampleResults, emmremlQValStatus < .1)$wTestPValStatus < .05)*100," %.")

bTest <- binom.test(sum(subset(longSampleResults, emmremlQValStatus < .1 & wTestPValStatus < .05)$statusDirConsistent),
           length(subset(longSampleResults, emmremlQValStatus < .1 & wTestPValStatus < .05)$statusDirConsistent),
           p = 0.5, alternative = c("greater"),conf.level = 0.95)

paste0(sum(subset(longSampleResults, emmremlQValStatus < .1 & wTestPValStatus < .05)$statusDirConsistent),
       " of these genes, or ",
       100*mean(subset(longSampleResults, emmremlQValStatus < .1 & wTestPValStatus < .05)$statusDirConsistent),
       " %, were significant in the same direction as measured by the linear modeling among the population sample (n.s. Binomial Test, p = ",bTest$p.value,").")





ggplot() + geom_histogram(aes(x = subset(longSampleResults, emmremlQValStatus < .1)$wTestPValStatus), bins = 50)

statusTtestPvals <- subset(longSampleResults, emmremlQValStatus < .1)$wTestPValStatus

qvalue(statusTtestPvals, pi0=1)$qvalues

### try a permutation test - 100x permutations of D-S labels?
#then calculate the p-value for each gene, and then that list of permuted p-values goes into a DF to use perm.fdr as a comparison with the true set

#1) pull out the sig genes from longSampleResults
geneByGenePlot <- TRUE
varA <- "LPS"

nPerms <- 100
nGenes <- dim(subset(longSampleResults, emmremlQValStatus < .1))[1]

permPvals <- matrix(NA, nrow = nGenes, ncol = nPerms)
permNumber <- 1
geneNum <- 1

wTestPValPreg <- vector()
tTestStatPreg <- vector()
wTestPValStatus <- vector()
tTestStatStatus <- vector()
emmremlStdBetaStatus <- vector()
emmremlPValStatus <- vector()
emmremlStdBetaPreg <- vector()
emmremlPValPreg <- vector()
emmremlQValPreg <- vector()
emmremlQValStatus <- vector()

#vectors for gene by gene line plot
geneNamePlot <- vector()
pregPval <- vector()
statPval <- vector()
meanGEpreg <- vector()
meanGEnonpreg <- vector()
meanGEdom <- vector()
meanGEsub <- vector()
treatment <- vector()


  
for (varA in c("LPS","Gard","Dex","Ctrl")) {
  #print(varA)
  if (varA %in% c("LPS","Ctrl")) {
    modelA <- "modelF"
  } else {
    modelA <- "modelG"
  }
  
  ### use the model without the repeated animals for the gene info
  emmremlResultsPreg <- read.table(paste0(outDir,"allBatches_F_",varA,"only_10000genes_5lCPM_pregReSamp_",
                                          modelA,"_basic+Preg+WeightwKmatrix_emmreml_wPermFDR_v1_preg"))
  emmremlResultsPreg$qval_preg <- qvalue(emmremlResultsPreg$pval_preg)$qvalue
  
  emmremlResultsStat <- read.table(paste0(outDir,"allBatches_F_",varA,"only_10000genes_5lCPM_statusReSamp_",
                                          modelA,"_basic+Preg+WeightwKmatrix_emmreml_wPermFDR_v1_status"))
  emmremlResultsStat$qval_status <- qvalue(emmremlResultsStat$pval_status)$qvalue
  
  residsStatus <- readRDS(paste0(outDir,"allBatches_F_",varA,"only_10000genes_5lCPM_",modelA,"_resids.RDS"))
  metaStatus <- readRDS(paste0(outDir,"allBatches_F_",varA,"only_10000genes_5lCPM_",modelA,"_dfMeta.RDS"))
  
  residsPreg <- readRDS(paste0(outDir,"allBatches_F_",varA,"only_10000genes_5lCPM_",modelA,"_resids.RDS"))
  metaPreg <- readRDS(paste0(outDir,"allBatches_F_",varA,"only_10000genes_5lCPM_",modelA,"_dfMeta.RDS"))
  
  metaPreg$preg <- ifelse(metaPreg$pregLact == "preg", TRUE, FALSE)
  metaStatus$preg <- ifelse(metaStatus$pregLact == "preg", TRUE, FALSE)
  
  metaPreg$statusPreg <- paste0(metaPreg$status_dom_sub,"_",metaPreg$preg)
  metaStatus$statusPreg <- paste0(metaStatus$status_dom_sub,"_",metaStatus$preg)
  
  ### stil using full data's resids and meta
  #geneList <- rownames(emmremlResultsPreg)
  
  ### use the intersection of genes from resample status, resample preg, and original
  geneList <- intersect(intersect(rownames(emmremlResultsPreg),rownames(subset(emmremlResultsStat, fdr_status < .1))), rownames(residsPreg))
  
  ### only use genes that are sig at FDR 0.10 emmremlResultsStat
  if (length(geneList) > 0) {
  
  repeatSamples <- names(table(metaPreg$animal_id)[table(metaPreg$animal_id) > 1])
  
  ### find the repeat animals preg/not-preg
  s <- repeatSamples[1]
  doublePreg <- vector()
  sampleID <- vector()
  pregTF <- vector()
  status <- vector()
  animalID <- vector()
  
  for (s in repeatSamples) {
    #print(s)
    if (length(unique(subset(metaPreg, animal_id == s)$pregStatusBackDate35)) > 1) {
      #print("duplicate")
      doublePreg <- c(doublePreg,s)
      
      animalMeta <- subset(metaPreg, animal_id == s)
      sampleID <- c(sampleID,animalMeta$sampleID)
      pregTF <- c(pregTF,animalMeta$pregStatusBackDate35)
      status <- c(status,animalMeta$status_dom_sub)
      animalID <- c(animalID,animalMeta$animal_id)
      
    } else {
      #print("nope")
    }
  }
  
  pregResample <- cbind.data.frame(sampleID,pregTF,status,animalID)
  
  ### remove samples where status is mis-matched
  statusMismatch <- c("SSUR1474","SSUR1605","SSUR1781","SSUR2229","SSUR1249","SSUR0749")
  pregResample <- pregResample[!pregResample$sampleID %in% statusMismatch,]
  
  ### Find the repeat animals S/D
  repeatSamples <- names(table(metaPreg$animal_id)[table(metaPreg$animal_id) > 1])
  
  ### find the repeat animals preg/not-preg
  s <- repeatSamples[1]
  doubleStatus <- vector()
  sampleID <- vector()
  pregTF <- vector()
  status <- vector()
  animalID <- vector()
  
  for (s in repeatSamples) {
    #print(s)
    if (length(unique(subset(metaStatus, animal_id == s)$status_dom_sub)) > 1) {
      #print("duplicate")
      doubleStatus <- c(doubleStatus,s)
      
      animalMeta <- subset(metaStatus, animal_id == s)
      sampleID <- c(sampleID,animalMeta$sampleID)
      pregTF <- c(pregTF,animalMeta$pregStatusBackDate35)
      status <- c(status,animalMeta$status_dom_sub)
      animalID <- c(animalID,animalMeta$animal_id)
      
    } else {
      #print("nope")
    }
  }
  
  statusResample <- cbind.data.frame(sampleID,pregTF,status,animalID)
  
  pregMismatch <- c("SSUR1400","SSUR1485","SSUR1474","SSUR1605","SSUR1781","SSUR2229","SSUR1862","SSUR1905")
  statusResample <- statusResample[!statusResample$sampleID %in% pregMismatch,]
  
  #gene <- geneList[223]
  gene <- "CCR7"
  #geneNum <- 1
  
  for (gene in geneList) {
    #print(gene)
    ### Check that this code is truly making sure the samples match on the un-tested variable
    
    
    #get all repeats
    pregResampTTest <- pregResample
    pregResampTTest$geneResid <- 0

    #randomly pick one resample PREG
    for (a in unique(pregResampTTest$animalID)) {
      for (p in c(TRUE,FALSE)) {
        if (sum(pregResampTTest$animalID == a & pregResampTTest$pregTF == p) > 1) {
          #these are the samples that fit the category, just choose one, toss the rest)
          resamps <- length(which(pregResampTTest$animalID == a & pregResampTTest$pregTF == p))
          drop <- sample(which(pregResampTTest$animalID == a & pregResampTTest$pregTF == p), size = resamps - 1, replace = FALSE)
          pregResampTTest <- pregResampTTest[-drop,]
        }
      }
    }
    
    #get all repeats
    statResampTTest <- statusResample
    statResampTTest$geneResid <- 0

    #randomly pick one resample STATUS
    for (a in unique(statResampTTest$animalID)) {
      for (p in c("S","D")) {
        if (sum(statResampTTest$animalID == a & statResampTTest$status == p) > 1) {
          #these are the samples that fit the category, just choose one, toss the rest)
          resamps <- length(which(statResampTTest$animalID == a & statResampTTest$status == p))
          drop <- sample(which(statResampTTest$animalID == a & statResampTTest$status == p), size = resamps - 1, replace = FALSE)
          statResampTTest <- statResampTTest[-drop,]
        }
      }
    }

    for (s in pregResampTTest$sampleID) {
      pregResampTTest$geneResid[which(pregResampTTest$sampleID == s)] <- residsPreg[which(rownames(residsPreg) == gene),
                                                                              which(colnames(residsPreg) == s)]
    }
    
    for (s in statResampTTest$sampleID) {
      statResampTTest$geneResid[which(statResampTTest$sampleID == s)] <- residsStatus[which(rownames(residsStatus) == gene),
                                                                              which(colnames(residsStatus) == s)]
    }
    
    geneNamePlot <- c(geneNamePlot,gene)
    
    pregPval <- c(pregPval,emmremlResultsPreg$pval_preg[rownames(emmremlResultsPreg) == gene])
    statPval <- c(statPval,emmremlResultsStat$pval_status[rownames(emmremlResultsStat) == gene])
    
    meanGEpreg <- c(meanGEpreg,mean(subset(pregResampTTest, pregTF == TRUE)$geneResid))
    meanGEnonpreg <- c(meanGEnonpreg,mean(subset(pregResampTTest, pregTF == FALSE)$geneResid))
    
    meanGEdom <- c(meanGEdom,mean(subset(statResampTTest, status == "D")$geneResid))
    meanGEsub <- c(meanGEsub,mean(subset(statResampTTest, status == "S")$geneResid))
    
    ### Get emmreml data first, informs T-Test
    #preg emmreml
    emmremlBeta <- emmremlResultsPreg$beta_preg[rownames(emmremlResultsPreg) == gene]
    emmremlVar <- emmremlResultsPreg$var_beta_preg[rownames(emmremlResultsPreg) == gene]
    
    emmremlStdBetaPreg <- c(emmremlStdBetaPreg, emmremlBeta / sqrt(emmremlVar))
    emmremlPValPreg <- c(emmremlPValPreg, emmremlResultsPreg$pval_preg[rownames(emmremlResultsPreg) == gene])
    emmremlQValPreg <- c(emmremlQValPreg, emmremlResultsPreg$qval_preg[rownames(emmremlResultsPreg) == gene])
    
    #ttest preg
    if (emmremlBeta > 0) {
      testDir <- "greater"
    } else {
      testDir <- "less"
    }
    
    #wTestPreg <- wilcox.test(subset(pregResampTTest, pregTF == TRUE)$geneResid,
    #                     subset(pregResampTTest, pregTF == FALSE)$geneResid, paired = TRUE, alternative = testDir)
    tTestPreg <- t.test(subset(pregResampTTest, pregTF == TRUE)$geneResid,
                    subset(pregResampTTest, pregTF == FALSE)$geneResid, paired = TRUE, alternative = testDir)
    
    wTestPValPreg <- c(wTestPValPreg, tTestPreg$p.value)
    tTestStatPreg <- c(tTestStatPreg, tTestPreg$statistic)
    
    #status emmreml
    emmremlBeta <- emmremlResultsStat$beta_status[rownames(emmremlResultsStat) == gene]
    emmremlVar <- emmremlResultsStat$var_beta_status[rownames(emmremlResultsStat) == gene]
    
    emmremlStdBetaStatus <- c(emmremlStdBetaStatus, emmremlBeta / sqrt(emmremlVar))
    emmremlPValStatus <- c(emmremlPValStatus, emmremlResultsStat$pval_status[rownames(emmremlResultsStat) == gene])
    emmremlQValStatus <- c(emmremlQValStatus, emmremlResultsStat$fdr_status[rownames(emmremlResultsStat) == gene])
    
    #ttest status
    if (emmremlBeta > 0) {
      testDir <- "greater"
    } else {
      testDir <- "less"
    }
    
    #wTestStat <- wilcox.test(subset(statResampTTest, status == "D")$geneResid,
    #                     subset(statResampTTest, status == "S")$geneResid, paired = TRUE, alternative = testDir)
    
    tTestStat <- t.test(subset(statResampTTest, status == "D")$geneResid,
                    subset(statResampTTest, status == "S")$geneResid, paired = TRUE, alternative = testDir)
    
    wTestPValStatus <- c(wTestPValStatus, tTestStat$p.value)
    tTestStatStatus <- c(tTestStatStatus, tTestStat$statistic)

    treatment <- c(treatment,varA)
    
    
    for (permNumber in 1:nPerms) {
      ## shuffle and generate new p-value
      if (dim(statResampTTest)[1] == 10) {
        rowOrder <- c(sample(c(1,2)),
                      sample(c(3,4)),
                      sample(c(5,6)),
                      sample(c(7,8)),
                      sample(c(9,10)))
      } else {
                rowOrder <- c(sample(c(1,2)),
                      sample(c(3,4)),
                      sample(c(5,6)),
                      sample(c(7,8)),
                      sample(c(9,10)),
                      sample(c(11,12)))
                    }
      
      statResampTTestPerm <- statResampTTest
      statResampTTestPerm$status <- statResampTTestPerm$status[rowOrder]
      
      tTestStatPerm <- t.test(subset(statResampTTestPerm, status == "D")$geneResid,
                              subset(statResampTTestPerm, status == "S")$geneResid, paired = TRUE, alternative = testDir)
      
      permPvals[geneNum,permNumber] <- tTestStatPerm$p.value
    }
    
    geneNum <- geneNum + 1
  }
  
  }
}

longSampleResults <- cbind.data.frame(geneNamePlot,
                                      pregPval,
                                      meanGEpreg,
                                      meanGEnonpreg,
                                      wTestPValPreg,
                                      tTestStatPreg,
                                      emmremlStdBetaPreg,
                                      emmremlPValPreg,
                                      emmremlQValPreg,
                                      statPval,
                                      meanGEdom,
                                      meanGEsub,
                                      wTestPValStatus,
                                      tTestStatStatus,
                                      emmremlStdBetaStatus,
                                      emmremlPValStatus,
                                      emmremlQValStatus,
                                      treatment)

longSampleResults$pregGEdiff <- longSampleResults$meanGEpreg - longSampleResults$meanGEnonpreg
longSampleResults$statusGEdiff <- longSampleResults$meanGEdom - longSampleResults$meanGEsub

#preg directional consistent
longSampleResults$pregDirConsistent <- (longSampleResults$pregGEdiff / longSampleResults$emmremlStdBetaPreg) > 0
longSampleResults$pregSigConsistent <- longSampleResults$pregDirConsistent & longSampleResults$wTestPValPreg < .05

#status directional consistent
longSampleResults$statusDirConsistent <- (longSampleResults$statusGEdiff / longSampleResults$emmremlStdBetaStatus) > 0
longSampleResults$statusSigConsistent <- longSampleResults$statusDirConsistent & longSampleResults$wTestPValStatus < .05

longSampleResultsStatusSig <- subset(longSampleResults, emmremlQValStatus < .1)

#use perm.fdr
longSampleResultsStatusSigFDR <- perm.fdr(input_df = longSampleResultsStatusSig, Pvals_col_name = "wTestPValStatus", perm_df = permPvals, name = "pairedTTest")

ggplot() + geom_histogram(aes(as.vector(permPvals)), bins = 50) +
  ggtitle("All T-Test Permuted P-values (100x perms)") + xlim(c(-0.05,1.05))

ggplot() + geom_histogram(aes(longSampleResultsStatusSigFDR$wTestPValStatus), bins = 50) +
  ggtitle("Original T-test P-values") + xlim(c(-0.05,1.05))

ggplot() + geom_histogram(aes(longSampleResultsStatusSigFDR$fdr_pairedTTest), bins = 50) +
  ggtitle("perm.fdr results (Sanz script)") + xlim(c(-0.05,1.05))

sum(longSampleResultsStatusSigFDR$fdr_pairedTTest < .1)

#save to output
saveRDS(object = longSampleResults, 
        file = paste0(outDir,"longitudinalSampleResultsSubControl.RDS"))

ggplot() + 
  geom_histogram(aes(p.adjust(longSampleResultsStatusSig$wTestPValStatus, 
                          method = "BH", n = length(longSampleResultsStatusSig$wTestPValStatus))), bins = 50) +
  ggtitle("Benjamini-Hochberg adj p-values") + xlim(c(-0.05,1.05))


```


```{r cross species GSEA}

```
